# Fix cross compiling issue
# Remove -fvisibility=hidden flag
--- configure.orig      2015-11-15 13:32:43.244650908 -0500
+++ configure   2015-11-15 13:34:28.800652406 -0500
@@ -15850,7 +15850,7 @@
   { $as_echo "$as_me:${as_lineno-$LINENO}: checking the byte alignment" >&5
 $as_echo_n "checking the byte alignment... " >&6; }

-  if test "$cross_compiling" = yes; then :
+  if test "$cross_compiling" = maybe; then :
   { { $as_echo "$as_me:${as_lineno-$LINENO}: error: in \`$ac_pwd':" >&5
 $as_echo "$as_me: error: in \`$ac_pwd':" >&2;}
 as_fn_error $? "cannot run test program while cross compiling
@@ -16458,7 +16458,7 @@
   { $as_echo "$as_me:${as_lineno-$LINENO}: checking whether kqueue supports pipes and ptys" >&5
 $as_echo_n "checking whether kqueue supports pipes and ptys... " >&6; }

-  if test "$cross_compiling" = yes; then :
+  if test "$cross_compiling" = maybe; then :
   { { $as_echo "$as_me:${as_lineno-$LINENO}: error: in \`$ac_pwd':" >&5
 $as_echo "$as_me: error: in \`$ac_pwd':" >&2;}
 as_fn_error $? "cannot run test program while cross compiling
@@ -16553,7 +16553,7 @@
   { $as_echo "$as_me:${as_lineno-$LINENO}: checking whether kqueue supports pipes and ptys" >&5
 $as_echo_n "checking whether kqueue supports pipes and ptys... " >&6; }

-  if test "$cross_compiling" = yes; then :
+  if test "$cross_compiling" = maybe; then :
   { { $as_echo "$as_me:${as_lineno-$LINENO}: error: in \`$ac_pwd':" >&5
 $as_echo "$as_me: error: in \`$ac_pwd':" >&2;}
 as_fn_error $? "cannot run test program while cross compiling
@@ -17229,7 +17229,7 @@


         tmp_CXXFLAGS=$CXXFLAGS
-        CXXFLAGS="$CXXFLAGS -fvisibility=hidden"
+        CXXFLAGS="$CXXFLAGS"

        { $as_echo "$as_me:${as_lineno-$LINENO}: checking if compiler supports __attribute__((visibility(\"default\")))" >&5
 $as_echo_n "checking if compiler supports __attribute__((visibility(\"default\")))... " >&6; }
@@ -17265,7 +17265,7 @@

 $as_echo "#define SUPPORT_ATTRIBUTE_VISIBILITY 1" >>confdefs.h

-                CXXFLAGS="$CXXFLAGS -fvisibility=hidden"
+                CXXFLAGS="$CXXFLAGS"

        else
                 true
@@ -18268,7 +18268,7 @@
   { $as_echo "$as_me:${as_lineno-$LINENO}: checking for execinfo.h" >&5
 $as_echo_n "checking for execinfo.h... " >&6; }

-  if test "$cross_compiling" = yes; then :
+  if test "$cross_compiling" = maybe; then :
   { { $as_echo "$as_me:${as_lineno-$LINENO}: error: in \`$ac_pwd':" >&5
 $as_echo "$as_me: error: in \`$ac_pwd':" >&2;}
 as_fn_error $? "cannot run test program while cross compiling
