# disable test execution when cross_compile = yes
# 
--- configure.orig	2024-09-28 16:06:43.000000000 +0000
+++ configure	2025-01-18 18:28:08.024342161 +0000
@@ -18375,7 +18375,7 @@
   { printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking the byte alignment" >&5
 printf %s "checking the byte alignment... " >&6; }
 
-  if test "$cross_compiling" = yes
+  if test "$cross_compiling" = maybe
 then :
   { { printf "%s\n" "$as_me:${as_lineno-$LINENO}: error: in '$ac_pwd':" >&5
 printf "%s\n" "$as_me: error: in '$ac_pwd':" >&2;}
