--- configure.orig	2012-04-20 08:56:14.000000000 +0200
+++ configure	2013-08-25 15:20:24.093974100 +0200
@@ -15658,7 +15658,7 @@
   { $as_echo "$as_me:${as_lineno-$LINENO}: checking the byte alignment" >&5
 $as_echo_n "checking the byte alignment... " >&6; }
 
-  if test "$cross_compiling" = yes; then :
+  if test "$cross_compiling" = maybe; then :
   { { $as_echo "$as_me:${as_lineno-$LINENO}: error: in \`$ac_pwd':" >&5
 $as_echo "$as_me: error: in \`$ac_pwd':" >&2;}
 as_fn_error $? "cannot run test program while cross compiling
@@ -16373,7 +16373,7 @@
   { $as_echo "$as_me:${as_lineno-$LINENO}: checking whether kqueue supports pipes and ptys" >&5
 $as_echo_n "checking whether kqueue supports pipes and ptys... " >&6; }
 
-  if test "$cross_compiling" = yes; then :
+  if test "$cross_compiling" = maybe; then :
   { { $as_echo "$as_me:${as_lineno-$LINENO}: error: in \`$ac_pwd':" >&5
 $as_echo "$as_me: error: in \`$ac_pwd':" >&2;}
 as_fn_error $? "cannot run test program while cross compiling
@@ -17049,7 +17049,7 @@
 
 
         tmp_CXXFLAGS=$CXXFLAGS
-        CXXFLAGS="$CXXFLAGS -fvisibility=hidden"
+        CXXFLAGS="$CXXFLAGS"
 
 	{ $as_echo "$as_me:${as_lineno-$LINENO}: checking if compiler supports __attribute__((visibility(\"default\")))" >&5
 $as_echo_n "checking if compiler supports __attribute__((visibility(\"default\")))... " >&6; }
@@ -17085,7 +17085,7 @@
 
 $as_echo "#define SUPPORT_ATTRIBUTE_VISIBILITY 1" >>confdefs.h
 
-                CXXFLAGS="$CXXFLAGS -fvisibility=hidden"
+                CXXFLAGS="$CXXFLAGS"
 
 	else
                 true
