# add ncursesw library to search for ncurses
#
--- configure.ac.orig	2022-01-11 09:10:34.000000000 +0000
+++ configure.ac	2022-04-22 22:05:16.494640022 +0000
@@ -121,7 +121,7 @@
   # Prefer ncurses over curses, if both are available.
   # (On Solaris 11.3, ncurses builds and links for us, but curses does not.)
   [AC_SEARCH_LIBS(
-    [initscr], [ncurses curses],
+    [initscr], [ncursesw ncurses curses],
     [AC_DEFINE([HAVE_CURSES], [1], [Define if a curses library available])],
     [with_ncurses=no])
 ])
