--- Makefile.in.orig	2015-08-30 19:24:26.000000000 +0200
+++ Makefile.in	2016-01-18 21:35:18.757004586 +0100
@@ -263,7 +263,7 @@
 ETAGS = etags
 CTAGS = ctags
 CSCOPE = cscope
-DIST_SUBDIRS = m4 po intl doc contrib imap
+DIST_SUBDIRS = m4 po intl contrib imap
 DISTFILES = $(DIST_COMMON) $(DIST_SOURCES) $(TEXINFOS) $(EXTRA_DIST)
 distdir = $(PACKAGE)-$(VERSION)
 top_distdir = $(distdir)
@@ -458,7 +458,7 @@
 AUTOMAKE_OPTIONS = 1.6 foreign
 @BUILD_IMAP_TRUE@IMAP_SUBDIR = imap
 @BUILD_IMAP_TRUE@IMAP_INCLUDES = -I$(top_srcdir)/imap
-SUBDIRS = m4 po intl doc contrib $(IMAP_SUBDIR)
+SUBDIRS = m4 po intl contrib $(IMAP_SUBDIR)
 bin_SCRIPTS = muttbug flea $(SMIMEAUX_TARGET)
 @BUILD_HCACHE_TRUE@HCVERSION = hcversion.h
 BUILT_SOURCES = keymap_defs.h patchlist.c reldate.h conststrings.c $(HCVERSION)
