--- Makefile.orig	2025-11-17 03:25:46.000000000 +0000
+++ Makefile	2025-12-03 11:53:18.597608979 +0000
@@ -132,13 +132,13 @@
 # also set SANDBOX_INIT_ERROR_IGNORE in their environment to ignore
 # failure.
 # Has no effect unless HAVE_SANDBOX_INIT is defined.
-.ifdef SANDBOX_INIT_ERROR_IGNORE
-.if $(SANDBOX_INIT_ERROR_IGNORE) == "always"
+#ifdef SANDBOX_INIT_ERROR_IGNORE
+ifeq ($(SANDBOX_INIT_ERROR_IGNORE), "always")
 CFLAGS		+= -DSANDBOX_INIT_ERROR_IGNORE=2
-.else
+else
 CFLAGS		+= -DSANDBOX_INIT_ERROR_IGNORE=1
-.endif
-.endif
+endif
+#endif
 
 # Names of shared and static libraries.
 LIB_SO		 = liblowdown.$(LINKER_SOSUFFIX)
@@ -152,20 +152,20 @@
 # object.
 LIB_LOWDOWN 	 = $(LIB_ST)
 MAIN_OBJS	 =
-.ifdef LINK_METHOD
-.if $(LINK_METHOD) == "shared"
+#ifdef LINK_METHOD
+ifeq ($(LINK_METHOD), "shared")
 LIB_LOWDOWN 	 = $(LIB_SO)
 MAIN_OBJS	 = $(COMPAT_OBJS)
-.endif
-.endif
+endif
+#endif
 
 # Mac OS X and other Unix systems use different conventions for
 # indicating shared library versions.
-.if $(LINKER_SOSUFFIX) == "dylib"
+ifeq ($(LINKER_SOSUFFIX), "dylib")
 LIB_SOVER	 = liblowdown.$(LIBVER).$(LINKER_SOSUFFIX)
-.else
+else
 LIB_SOVER	 = liblowdown.$(LINKER_SOSUFFIX).$(LIBVER)
-.endif
+endif
 
 # Because the objects will be compiled into a shared library:
 CFLAGS		+= -fPIC
