--- src/lib/Makefile.am.orig	2013-12-21 11:16:10.000000000 +0100
+++ src/lib/Makefile.am	2014-05-17 22:57:38.419177900 +0200
@@ -86,7 +86,7 @@ AMD64_OBJS = $(AMD64_SRCS:.S=.lo)
 
 EXTRA_DIST = $(MMX_SRCS) $(AMD64_SRCS) asm_loadimmq.S
 
-MY_LIBS = -lXext -lX11 $(FREETYPE_LIBS) $(DLOPEN_LIBS) -lm
+MY_LIBS = $(FREETYPE_LIBS) $(DLOPEN_LIBS) -lm
 
 if BUILD_MMX
 libImlib2_la_LIBADD       = $(MMX_OBJS) $(MY_LIBS)
--- src/lib/Makefile.in.orig	2013-12-21 13:42:42.000000000 +0100
+++ src/lib/Makefile.in	2014-05-17 22:58:07.049160300 +0200
@@ -443,7 +443,7 @@ amd64_blend_cmod.S
 MMX_OBJS = $(MMX_SRCS:.S=.lo)
 AMD64_OBJS = $(AMD64_SRCS:.S=.lo)
 EXTRA_DIST = $(MMX_SRCS) $(AMD64_SRCS) asm_loadimmq.S
-MY_LIBS = -lXext -lX11 $(FREETYPE_LIBS) $(DLOPEN_LIBS) -lm
+MY_LIBS = $(FREETYPE_LIBS) $(DLOPEN_LIBS) -lm
 @BUILD_AMD64_FALSE@@BUILD_MMX_FALSE@libImlib2_la_LIBADD = $(MY_LIBS)
 @BUILD_AMD64_TRUE@@BUILD_MMX_FALSE@libImlib2_la_LIBADD = $(AMD64_OBJS) $(MY_LIBS)
 @BUILD_MMX_TRUE@libImlib2_la_LIBADD = $(MMX_OBJS) $(MY_LIBS)
--- src/lib/Imlib2.h.orig	2013-12-21 11:15:58.000000000 +0100
+++ src/lib/Imlib2.h	2014-05-17 23:37:06.770536000 +0200
@@ -1,3 +1,5 @@
+# define X_DISPLAY_MISSING 1
+
 #ifndef __IMLIB_API_H
 # define __IMLIB_API_H 1
 
