--- Makefile.orig	2012-07-14 00:01:42.000000000 +0200
+++ Makefile	2012-07-14 15:34:33.000000000 +0200
@@ -54,13 +54,15 @@
 CC_OPTS = -O0 -ggdb
 endif
 
-CC = $(CROSS_DIR)$(CROSS)gcc
-AR = $(CROSS_DIR)$(CROSS)ar
-STRIP = $(CROSS_DIR)$(CROSS)strip
-RANLIB = $(CROSS_DIR)$(CROSS)ranlib
+#CC = $(CROSS_DIR)$(CROSS)gcc
+#AR = $(CROSS_DIR)$(CROSS)ar
+#STRIP = $(CROSS_DIR)$(CROSS)strip
+#RANLIB = $(CROSS_DIR)$(CROSS)ranlib
+LIBUSB = "$(STAGING_INSTALL_PREFIX)/$(SPK_NAME)/lib"
+USE_LIBUSB=1
 
-ARFLAGS = -rcsl
-LDFLAGS = -Wl,--gc-sections
+#ARFLAGS = -rcsl
+LDFLAGS += -Wl,--gc-sections
 
 # The linker for powerpc have bug that prevents --gc-sections from working
 # Check for the linker version and if it matches disable --gc-sections
@@ -388,6 +388,6 @@
 $(OSCAM_BIN): $(ALGO_LIB) $(CSCRYPT_LIB) $(CSCTAPI_LIB) $(OSCAM_LIB)
 	$(SAY) "LINK	$@"
-	$(Q)$(CC) $(STD_DEFS) $(CC_OPTS) $(CC_WARN) $(LDFLAGS) $(OSCAM_LIB) $(ALGO_LIB) $(CSCRYPT_LIB) $(CSCTAPI_LIB) $(LIBS) -o $@
+	$(Q)$(CC) $(STD_DEFS) $(CC_OPTS) $(CC_WARN) $(CFLAGS) $(LDFLAGS) $(OSCAM_LIB) $(ALGO_LIB) $(CSCRYPT_LIB) $(CSCTAPI_LIB) $(LIBS) -o $@
 ifndef DEBUG
 	$(SAY) "STRIP	$@"
 	$(Q)$(STRIP) $@
@@ -398,6 +398,6 @@
 $(LIST_SMARGO_BIN): utils/list_smargo.c
 	$(SAY) "LINK	$@"
-	$(Q)$(CC) $(STD_DEFS) $(LDFLAGS) utils/list_smargo.c $(LIBS) -o $@
+	$(Q)$(CC) $(STD_DEFS) $(CFLAGS) $(LDFLAGS) utils/list_smargo.c $(LIBS) -o $@
 ifndef DEBUG
 	$(SAY) "STRIP	$@"
 	$(Q)$(STRIP) $@
