--- setup.py.org	2014-04-01 18:57:54.373039343 +0200
+++ setup.py	2014-04-01 19:03:06.537114908 +0200
@@ -44,15 +44,6 @@
 except AttributeError:
     pass
 
-if isironpython:
-    sys.stderr.write("warning: IronPython detected (no bz2 support)\n")
-else:
-    try:
-        import bz2
-    except ImportError:
-        raise SystemExit(
-            "Couldn't import standard bz2 (incomplete Python install).")
-
 import os, subprocess, time
 import shutil
 import tempfile
@@ -408,6 +399,21 @@
             fp.write(data)
             fp.close()
 
+        for outfile in self.outfiles:
+            fp = open(outfile, 'rb')
+            data1 = fp.read()
+            fp.close()
+
+            # skip binary files
+            if b('\0') in data:
+                continue
+                
+            shebang = '#!python'
+            data1 = data1.replace('#!/usr/local/python/bin/python2.7', shebang.encode('string_escape')) 
+            fp = open(outfile, 'wb')
+            fp.write(data1)
+            fp.close()
+
 cmdclass = {'build': hgbuild,
             'build_mo': hgbuildmo,
             'build_ext': hgbuildext,
