--- setup.py.org	2014-04-01 18:57:54.373039343 +0200
+++ setup.py	2014-04-01 19:03:06.537114908 +0200
@@ -45,24 +45,6 @@
     raise SystemExit(
         "Couldn't import standard zlib (incomplete Python install).")
 
-# The base IronPython distribution (as of 2.7.1) doesn't support bz2
-isironpython = False
-try:
-    isironpython = (platform.python_implementation()
-                    .lower().find("ironpython") != -1)
-except AttributeError:
-    pass
-
-if isironpython:
-    sys.stderr.write("warning: IronPython detected (no bz2 support)\n")
-else:
-    try:
-        import bz2
-        bz2.BZ2Compressor # silence unused import warning
-    except ImportError:
-        raise SystemExit(
-            "Couldn't import standard bz2 (incomplete Python install).")
-
 ispypy = "PyPy" in sys.version
 
 import os, stat, subprocess, time
@@ -465,6 +465,21 @@
             fp.write(data)
             fp.close()
 
+        for outfile in self.outfiles:
+            fp = open(outfile, 'rb')
+            data1 = fp.read()
+            fp.close()
+
+            # skip binary files
+            if b('\0') in data:
+                continue
+
+            shebang = '#!python'
+            data1 = data1.replace('#!/usr/local/mercurial/bin/python2.7', shebang.encode('string_escape'))
+            fp = open(outfile, 'wb')
+            fp.write(data1)
+            fp.close()
+
 cmdclass = {'build': hgbuild,
             'build_mo': hgbuildmo,
             'build_ext': hgbuildext,
