--- src/PCSC/winscard.h.orig	2013-11-24 14:11:00.499544514 +0100
+++ src/PCSC/winscard.h	2013-11-24 14:11:18.567543924 +0100
@@ -17,7 +17,7 @@
 #ifndef __winscard_h__
 #define __winscard_h__
 
-#include <pcsclite.h>
+#include <PCSC/pcsclite.h>
 
 #ifdef __cplusplus
 extern "C"
--- src/PCSC/pcsclite.h.in.orig	2013-11-24 14:10:28.419545563 +0100
+++ src/PCSC/pcsclite.h.in	2013-11-24 14:10:52.587544773 +0100
@@ -21,7 +21,7 @@
 #ifndef __pcsclite_h__
 #define __pcsclite_h__
 
-#include <wintypes.h>
+#include <PCSC/wintypes.h>
 
 #ifdef __cplusplus
 extern "C"
