diff -urN include.orig/asm-generic/Kbuild include/asm-generic/Kbuild
--- include.orig/asm-generic/Kbuild	2013-08-03 10:01:43.000000000 +0200
+++ include/asm-generic/Kbuild	2013-10-02 00:26:09.014194884 +0200
@@ -22,6 +22,7 @@
 header-y += siginfo.h
 header-y += signal-defs.h
 header-y += signal.h
+header-y += sizes.h
 header-y += socket.h
 header-y += sockios.h
 header-y += stat.h
diff -urN include.orig/linux/Kbuild include/linux/Kbuild
--- include.orig/linux/Kbuild	2013-08-03 10:01:43.000000000 +0200
+++ include/linux/Kbuild	2013-10-02 00:24:05.225069475 +0200
@@ -407,6 +407,7 @@
 header-y += synosata.h
 header-y += syno_debug.h
 header-y += synobios.h
+header-y += syno_acl.h
 header-y += syno_acl_xattr_ds.h
 header-y += synotify.h
 header-y += wireless.h
