--- etc/Makefile.in.orig	2012-10-04 11:58:03.000000000 +0200
+++ etc/Makefile.in	2012-10-04 12:02:05.000000000 +0200
@@ -176,6 +176,7 @@ CONFS = squidclamav.conf
 cicapdatadir = @DATADIR@
 CONFIGDIR = @CONFDIR@
 CFGINST = @CFGDIR@
+DESTDIR = $(prefix)
 TMPLS = templates/en/MALWARE_FOUND
 EXTRA_DIST = squidclamav.conf $(TMPLS)
 all: all-am
--- src/Makefile.in.orig	2012-10-04 12:06:28.000000000 +0200
+++ src/Makefile.in	2012-10-04 12:07:15.000000000 +0200
@@ -225,6 +225,7 @@ top_builddir = @top_builddir@
 top_srcdir = @top_srcdir@
 INCLUDES = -I$(top_srcdir)
 CONFIGDIR = @CONFDIR@
+DESTDIR = $(prefix)/../../../
 cicapdatadir = @DATADIR@
 cicapmoduledir = @MODSDIR@
 cicapmodule_LTLIBRARIES = squidclamav.la
