--- etc/Makefile.in.orig	2012-10-04 11:58:03.000000000 +0200
+++ etc/Makefile.in	2012-10-04 12:02:05.000000000 +0200
@@ -178,6 +178,7 @@
 CFGINST = @CFGDIR@
 TMPLS = templates/en/MALWARE_FOUND
 EXTRA_DIST = squidclamav.conf $(TMPLS)
+DESTDIR = $(prefix)
 all: all-am
 
 .SUFFIXES:
--- src/Makefile.in.orig	2012-10-04 12:06:28.000000000 +0200
+++ src/Makefile.in	2012-10-04 12:07:15.000000000 +0200
@@ -226,6 +226,7 @@
 INCLUDES = -I$(top_srcdir)
 CONFIGDIR = @CONFDIR@
 cicapdatadir = @DATADIR@
+DESTDIR = $(prefix)/../../../
 cicapmoduledir = @MODSDIR@
 cicapmodule_LTLIBRARIES = squidclamav.la
 squidclamav_la_LIBADD = @MODULES_LIBADD@
