diff -uprN ../ffmpeg-4.2.1-patch060-NEW/fftools/ffmpeg.h ./fftools/ffmpeg.h
--- ../ffmpeg-4.2.1-patch060-NEW/fftools/ffmpeg.h	2020-01-18 09:42:59.123197281 -0500
+++ ./fftools/ffmpeg.h	2020-01-18 09:44:13.780059416 -0500
@@ -608,6 +608,9 @@ extern int print_stats;
 extern int qp_hist;
 extern int stdin_interaction;
 extern int frame_bits_per_raw_sample;
+// SYNO_EVANSPORT_DTV_SMD_AVSYNC
+extern int enable_dtv_smd;
+// SYNO_EVANSPORT_DTV_SMD_AVSYNC
 extern AVIOContext *progress_avio;
 extern float max_error_rate;
 extern char *videotoolbox_pixfmt;
diff -uprN ../ffmpeg-4.2.1-patch060-NEW/fftools/ffmpeg_opt.c ./fftools/ffmpeg_opt.c
--- ../ffmpeg-4.2.1-patch060-NEW/fftools/ffmpeg_opt.c	2020-01-18 09:42:48.767077108 -0500
+++ ./fftools/ffmpeg_opt.c	2020-01-18 09:44:13.780059416 -0500
@@ -114,6 +114,9 @@ int stdin_interaction = 1;
 int frame_bits_per_raw_sample = 0;
 float max_error_rate  = 2.0/3;
 
+// SYNO_EVANSPORT_DTV_SMD_AVSYNC
+int enable_dtv_smd = 0;
+// SYNO_EVANSPORT_DTV_SMD_AVSYNC
 #ifdef SYNO_DTV_SKIP_UNKNOWN_TRACK
 int dtv_skip_unknown = 0;
 #endif
@@ -3396,6 +3399,10 @@ static int opt_progress(void *optctx, co
 const OptionDef options[] = {
     /* main options */
     CMDUTILS_COMMON_OPTIONS
+#ifdef SYNO_EVANSPORT_DTV_SMD_AVSYNC
+    { "dtv_smd",        OPT_BOOL,                                    {              &enable_dtv_smd},
+        "enable passthough timestamp in hardware pipeline mode" },
+#endif
 #ifdef SYNO_DTV_SKIP_UNKNOWN_TRACK
     { "dtv_skip_unknown",        OPT_BOOL,                                    {              &dtv_skip_unknown},
         "skip unknown codec track during dtv streaming" },
diff -uprN ../ffmpeg-4.2.1-patch060-NEW/synoconfig.h ./synoconfig.h
--- ../ffmpeg-4.2.1-patch060-NEW/synoconfig.h	2020-01-18 09:43:27.451525257 -0500
+++ ./synoconfig.h	2020-01-18 09:44:13.812059786 -0500
@@ -119,3 +119,12 @@
 #if defined(SYNO_EVANSPORT_SMD)
 #define SYNO_EVANSPORT_FLV_EXTRADATA
 #endif
+ 
+/* Fixed DTV recoded file or DTV streaming A/V issue on Evansport.
+ * add flag "-dtv_smd" is used other way to calculate timestamp.
+ * in hardware pipeline mode.
+ * See Video Station #1652, #1716
+ */
+#if defined(SYNO_EVANSPORT_SMD)
+#define SYNO_EVANSPORT_DTV_SMD_AVSYNC
+#endif
