--- support/version.orig	2019-07-29 14:53:41.000000000 -0400
+++ support/version	2019-08-06 21:05:00.135025823 -0400
@@ -7,7 +7,9 @@
 FILE=$1
 
 # Calculate version
-if [ -d ".git" ]; then
+if [ "$TVH_VERS" ]; then
+  VER="$TVH_VERS"
+elif [ -d ".git" ]; then
   VER=$(cd "$(dirname "$0")"/..; git describe --dirty --match "v*" 2> /dev/null)
   if [ $? -ne 0 ]; then
     # Git describe failed, maybe "--dirty" option is not available
