--- configure-ORIG	2019-10-06 14:33:17.000000000 -0400
+++ configure	2019-10-13 09:20:09.081287605 -0400
@@ -646,10 +646,12 @@ fi
 if enabled_or_auto vaapi; then
   if enabled libav; then
     if check_pkg libva ">=0.38.0"; then
-      check_pkg libva-x11 ">=0.38.0" || die "libva-x11 not found"
-      check_pkg libva-drm ">=0.38.0" || die "libva-drm not found"
-      enable vaapi
-      enable hwaccels
+      if check_pkg libva-x11 ">=0.38.0" || check_pkg libva-drm ">=0.38.0"; then
+        enable vaapi
+        enable hwaccels
+      else
+        die "libva-x11 and libva-drm not found"
+      fi
     elif enabled vaapi; then
       die "vaapi (Video Acceleration (VA) API for Linux) not found"
     fi
