--- Makefile.orig	2012-09-30 12:47:23.000000000 +0200
+++ Makefile	2012-09-30 12:50:03.000000000 +0200
@@ -27,7 +27,7 @@
 # Common compiler flags
 #
 
-CFLAGS  += -Wall -Werror -Wwrite-strings -Wno-deprecated-declarations
+CFLAGS  += -Wall -Wwrite-strings -Wno-deprecated-declarations
 CFLAGS  += -Wmissing-prototypes -fms-extensions
 CFLAGS  += -g -funsigned-char -O2 
 CFLAGS  += -D_FILE_OFFSET_BITS=64
