--- data/conf/transcoder/profiles.orig
+++ data/conf/transcoder/profiles
@@ -76,6 +76,13 @@
         "name": "webtv-h264-software-aac-mp4"
     },
     {
+        "comment": "WEBTV profile H264 Software/AAC/MPEG-TS",
+        "container": 2,
+        "pro_vcodec": "webtv-h264",
+        "pro_acodec": "webtv-aac",
+        "name": "webtv-h264-software-aac-mpegts"
+    },
+    {
         "comment": "WEBTV profile HEVC QSV/AAC/Matroska",
         "container": 1,
         "pro_vcodec": "webtv-hevc-qsv",
