diff -Naur configure.orig configure
--- configure.orig	2013-04-18 16:23:19.000000000 +0200
+++ configure		2013-04-25 20:03:24.365353280 +0200
@@ -54,6 +54,11 @@
 check_cc_snippet getloadavg '#include <stdlib.h> 
 void test() { getloadavg(NULL,0); }'
 
+check_cc_snippet atomic32 '#include <stdint.h>
+int test(int *ptr){
+return __sync_fetch_and_add(ptr, 1);
+}'
+
 check_cc_snippet atomic64 '#include <stdint.h>
 uint64_t test(uint64_t *ptr){
 return __sync_fetch_and_add(ptr, 1);
diff -Naur src/atomic.h.orig src/atomic.h
--- src/atomic.h.orig   2016-03-13 18:58:10.242043596 -0400
+++ src/atomic.h        2016-03-13 19:04:28.360549636 -0400
@@ -21,7 +21,16 @@
 static inline int
 atomic_add(volatile int *ptr, int incr)
 {
+#if ENABLE_ATOMIC32
   return __sync_fetch_and_add(ptr, incr);
+#else
+  int ret;
+  pthread_mutex_lock(&atomic_lock);
+  ret = *ptr;
+  *ptr += incr;
+  pthread_mutex_unlock(&atomic_lock);
+  return ret;
+#endif
 }

 static inline int
@@ -33,7 +42,16 @@
 static inline int
 atomic_exchange(volatile int *ptr, int new)
 {
+#if ENABLE_ATOMIC32
   return  __sync_lock_test_and_set(ptr, new);
+#else
+  int ret;
+  pthread_mutex_lock(&atomic_lock);
+  ret = *ptr;
+  *ptr = new;
+  pthread_mutex_unlock(&atomic_lock);
+  return ret;
+#endif
 }

 static inline uint64_t
