--- src/dvr/dvr_db.c.orig	2012-07-14 12:13:35.000000000 +0200
+++ src/dvr/dvr_db.c	2012-07-14 12:10:04.000000000 +0200
@@ -958,6 +958,12 @@
   dvr_config_t *cfg;
 
   dvr_iov_max = sysconf(_SC_IOV_MAX);
+  if( dvr_iov_max == -1 )                        
+#ifdef IOV_MAX                    
+    dvr_iov_max = IOV_MAX;                                           
+#else                                                                    
+    dvr_iov_max = 16;                       
+#endif
 
   /* Default settings */
 
