--- Makefile.orig	2016-03-13 18:48:44.480219027 -0400
+++ Makefile	2016-03-13 18:54:29.004982881 -0400
@@ -28,7 +28,7 @@
 #
 
 CFLAGS  += -g -O2 -Wunused-result
-CFLAGS  += -Wall -Werror -Wwrite-strings -Wno-deprecated-declarations
+CFLAGS  += -Wall -Wwrite-strings -Wno-deprecated-declarations
 CFLAGS  += -Wmissing-prototypes
 CFLAGS  += -fms-extensions -funsigned-char -fno-strict-aliasing
 CFLAGS  += -D_FILE_OFFSET_BITS=64
