Index: b/make/hotspot/gensrc/GensrcJfr.gmk
===================================================================
--- a/make/hotspot/gensrc/GensrcJfr.gmk
+++ b/make/hotspot/gensrc/GensrcJfr.gmk
@@ -41,6 +41,7 @@ $(eval $(call SetupJavaCompilation, BUIL
     SETUP := GENERATE_JFRBYTECODE, \
     SRC := $(JFR_TOOLS_SRCDIR), \
     BIN := $(JFR_TOOLS_OUTPUTDIR), \
+    INCLUDES := build/tools/jfr, \
 ))
 
 TARGETS += $(BUILD_JFR_TOOLS)
