--- HandBrake-1.3.1/libhb/module.defs	2020-01-04 10:28:06.000000000 -0600
+++ HandBrake-1.3.1/libhb/module.defs.new	2020-03-19 17:19:53.134856978 -0500
@@ -48,7 +48,7 @@
 ifneq (,$(filter $(HOST.system),freebsd netbsd))
     LIBHB.GCC.I += $(LOCALBASE)/include
     LIBHB.GCC.I += $(LOCALBASE)/include/libxml2
-else ifneq (,$(filter $(HOST.system),darwin cygwin mingw))
+else ifneq (,$(filter $(HOST.system),darwin cygwin mingw linux))
     LIBHB.GCC.I += $(CONTRIB.build/)include/libxml2
 else
     LIBHB.GCC.I += /usr/include/libxml2
