--- HandBrake-1.3.1/make/include/gcc.defs	2020-01-04 10:28:06.000000000 -0600
+++ HandBrake-1.3.1/make/include/gcc.defs.new	2020-03-19 17:32:26.245982806 -0500
@@ -52,7 +52,7 @@
 endif
 GCC.args.sysroot   = --sysroot=$(1)
 GCC.args.minver    = -mmacosx-version-min=$(1)
-GCC.args.vis       = -fvisibility=hidden
+GCC.args.vis       = -fvisibility=default
 GCC.args.pic       = -fPIC
 GCC.args.c_std     = -std=gnu99
 GCC.args.cxx_std   = -std=c++98
