--- HandBrake-1.3.1/make/include/contrib.defs.orig	2020-03-19 18:20:00.885641478 -0500
+++ HandBrake-1.3.1/make/include/contrib.defs	2020-03-19 18:22:56.776876723 -0500
@@ -1,4 +1,4 @@
-CONTRIB.build/    = $(BUILD/)contrib/
+CONTRIB.build/    = $(DESTDIR)$(PREFIX)/
 CONTRIB.download/ = $(SRC/)download/
 CONTRIB.host      = $(if $(filter 1,$(HOST.cross)),$(HOST.spec))
 CONTRIB.spec      = $(if $(filter 1,$(HOST.cross)),$(BUILD.spec))
@@ -124,7 +124,7 @@
         $(1).CONFIGURE.env.CXXFLAGS = CXXFLAGS="-I$$(call fn.ABSOLUTE,$(CONTRIB.build/))include $$(call fn.ARGS,$(1).GCC,*archs *sysroot *minver ?cxx_std ?extra *D)"
     endif
     $(1).CONFIGURE.env.CPPFLAGS = CPPFLAGS="-I$$(call fn.ABSOLUTE,$(CONTRIB.build/))include $$(call fn.ARGS,$(1).GCC,*archs *sysroot *minver *D)"
-    $(1).CONFIGURE.env.LDFLAGS  = LDFLAGS="-L$$(call fn.ABSOLUTE,$(CONTRIB.build/))lib $$(call fn.ARGS,$(1).GCC,*archs *sysroot *minver ?extra.exe *D)"
+    $(1).CONFIGURE.env.LDFLAGS  = LDFLAGS="-L$$(call fn.ABSOLUTE,$(CONTRIB.build/))lib -Wl,--rpath-link,$$(call fn.ABSOLUTE,$(CONTRIB.build/))lib -Wl,--rpath,$$(call fn.ABSOLUTE,$(CONFIGURE.prefix/))lib $$(call fn.ARGS,$(1).GCC,*archs *sysroot *minver ?extra.exe *D)"
     $(1).CONFIGURE.env.PKG_CONFIG_PATH  = PKG_CONFIG_PATH="$$(call fn.ABSOLUTE,$$(CONTRIB.build/))lib/pkgconfig"
 
     $(1).CONFIGURE.env.args = !CC !CFLAGS !CXX !CXXFLAGS !CPPFLAGS !LD !LDFLAGS !PKG_CONFIG_PATH !LOCAL_PATH !CROSS
