--- HandBrake-1.3.1/test/module.defs	2020-04-02 22:24:56.193050180 -0500
+++ HandBrake-1.3.1/test/module.defs.new	2020-04-02 22:48:12.032426050 -0500
@@ -15,7 +15,7 @@
 
 TEST.GCC.l = \
         ass avformat avfilter avcodec avutil swresample postproc mp3lame dvdnav \
-        dvdread fribidi x265 iconv numa drm uuid png \
+        dvdread fribidi x265 iconv numa drm uuid png fdk-aac \
         swscale vpx theoraenc theoradec vorbis vorbisenc ogg x264 \
         bluray freetype xml2 bz2 z jansson harfbuzz opus speex lzma dav1d
 
