--- php/settings.php.orig	2026-01-10 02:10:53.502183636 +0000
+++ php/settings.php	2026-01-10 02:12:09.684692186 +0000
@@ -230,7 +230,10 @@
 					$this->apiVersion = $req->val[0];
 			}
 
-                        $req = new rXMLRPCRequest( new rXMLRPCCommand("to_kb", floatval(1024)) );
+                        $req = new rXMLRPCRequest(new rXMLRPCCommand(
+                        	"convert.kb",
+                        	$this->iVersion >= 0x1000 ? array('',floatval(1024)) : floatval(1024)
+                        ));
 			if($req->run())
 			{
 				if(!$req->fault)
@@ -305,7 +308,13 @@
 	}
 	public function getRatioGroupCommand($ratio,$cmd,$args)
 	{
-		$prefix = ($this->iVersion >= 0x904) && in_array($cmd,$this->ratioCmds) ? "group2." : "group.";
+		if (($this->iVersion >= 0x904) && ($this->iVersion < 0x1000) && in_array($cmd,$this->ratioCmds)) {
+			$prefix = "group2.";
+		} else {
+			$prefix = "group.";
+			if (!is_array($args))
+				$args = array('', $args);
+		}
 		return( new rXMLRPCCommand( $prefix.$ratio.".".$cmd, $args ) );
 	}
 	public function getEventCommand($cmd1,$cmd2,$args)
