diff -uprN ../libva-utils-2.6.0/videoprocess/vppscaling_n_out_usrptr.cpp ./videoprocess/vppscaling_n_out_usrptr.cpp
--- ../libva-utils-2.6.0/videoprocess/vppscaling_n_out_usrptr.cpp	2019-12-10 09:26:13.000000000 -0500
+++ ./videoprocess/vppscaling_n_out_usrptr.cpp	2020-02-03 18:02:30.177449534 -0500
@@ -38,7 +38,9 @@
 #include <va/va.h>
 #include <va/va_vpp.h>
 #include "va_display.h"
+#if 0
 #include <va/va_x11.h>
+#endif
 
 #define MAX_LEN   1024
 
