diff -uprN ../libva-utils-2.6.0-ORIG/videoprocess/vppscaling_n_out_usrptr.cpp ./videoprocess/vppscaling_n_out_usrptr.cpp
--- ../libva-utils-2.6.0-ORIG/videoprocess/vppscaling_n_out_usrptr.cpp	2019-12-10 09:26:13.000000000 -0500
+++ ./videoprocess/vppscaling_n_out_usrptr.cpp	2020-02-04 03:38:09.983915618 -0500
@@ -38,7 +38,6 @@
 #include <va/va.h>
 #include <va/va_vpp.h>
 #include "va_display.h"
-#include <va/va_x11.h>
 
 #define MAX_LEN   1024
 
