--- deps/lua/src/Makefile.orig	2012-12-03 12:53:09.000000000 +0100
+++ deps/lua/src/Makefile	2012-12-08 11:35:21.000000000 +0100
@@ -5,18 +5,18 @@
 # == CHANGE THE SETTINGS BELOW TO SUIT YOUR ENVIRONMENT =======================
 
 # Your platform. See PLATS for possible values.
-PLAT= none
+PLAT= generic
 
-CC= gcc
-CFLAGS= -O2 -Wall $(MYCFLAGS)
-AR= ar rcu
-RANLIB= ranlib
+#CC= gcc
+#CFLAGS= -O2 -Wall $(MYCFLAGS)
+AR:= $(AR) rcu
+#RANLIB= ranlib
 RM= rm -f
-LIBS= -lm $(MYLIBS)
+#LIBS= -lm $(MYLIBS)
 
-MYCFLAGS=
-MYLDFLAGS=
-MYLIBS=
+#MYCFLAGS=
+#MYLDFLAGS=
+#MYLIBS=
 
 # == END OF USER SETTINGS. NO NEED TO CHANGE ANYTHING BELOW THIS LINE =========
 
