--- deps/lua/src/Makefile.orig	2014-09-01 09:37:09.547519848 +0200
+++ deps/lua/src/Makefile	2014-09-01 09:38:40.703930434 +0200
@@ -5,14 +5,14 @@
 # == CHANGE THE SETTINGS BELOW TO SUIT YOUR ENVIRONMENT =======================
 
 # Your platform. See PLATS for possible values.
-PLAT= none
+PLAT= generic
 
-CC?= gcc
-CFLAGS= -O2 -Wall $(MYCFLAGS)
-AR= ar rcu
-RANLIB= ranlib
+#CC?= gcc
+#CFLAGS= -O2 -Wall $(MYCFLAGS)
+AR:= $(AR) rcu
+#RANLIB= ranlib
 RM= rm -f
-LIBS= -lm $(MYLIBS)
+#LIBS= -lm $(MYLIBS)
 
 MYCFLAGS=
 MYLDFLAGS=
