--- deps/lua/src/Makefile-orig	2019-12-08 18:34:35.068004000 +0100
+++ deps/lua/src/Makefile	2019-12-08 18:35:45.836004000 +0100
@@ -5,12 +5,12 @@
 # == CHANGE THE SETTINGS BELOW TO SUIT YOUR ENVIRONMENT =======================
 
 # Your platform. See PLATS for possible values.
-PLAT= none
+PLAT= generic
 
 CC?= gcc
-CFLAGS= -O2 -Wall $(MYCFLAGS)
-AR= ar rcu
-RANLIB= ranlib
+CFLAGS+= -O2 -Wall $(MYCFLAGS)
+AR:= $(AR) rcu
+#RANLIB= ranlib
 RM= rm -f
 LIBS= -lm $(MYLIBS)
 
