--- external/boringssl/crypto/internal.h.orig	2017-11-30 12:17:41.844390358 +0000
+++ external/boringssl/crypto/internal.h	2017-11-30 12:17:48.627723651 +0000
@@ -117,7 +117,11 @@
 #define alignas(x) __declspec(align(x))
 #define alignof __alignof
 #endif
+#elif defined __GNUC__ && __STDC_VERSION <= 199901L
+// add support for alignas() for GCC versions below 4.8 (no C11 support)
+#define alignas(x) __attribute__ ((aligned (x)))
 #else
+// available with the C11 standard
 #include <stdalign.h>
 #endif
 
