# Fix implicit declaration errors for old PPC architectures
--- src/dvr/dvr_db.c.org	2017-03-17 20:55:43.784617822 +0100
+++ src/dvr/dvr_db.c	2017-03-17 21:03:35.314171581 +0100
@@ -17,6 +17,8 @@
  *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+#define _ISOC9X_SOURCE
+
 #include <pthread.h>
 #include <sys/stat.h>
 #include <unistd.h>
--- src/tvhtime.c.orig	2017-04-10 15:30:56.664970057 +0200
+++ src/tvhtime.c	2017-04-10 00:21:12.363948013 +0200
@@ -1,3 +1,5 @@
+#define _ISOC9X_SOURCE
+
 #include <time.h>
 #include <sys/ipc.h>
 #include <sys/time.h>
