--- cgi-bin/squidguardmgr.cgi.old	2012-12-19 17:49:18.134925914 +0100
+++ cgi-bin/squidguardmgr.cgi	2012-12-19 17:48:57.043191077 +0100
@@ -14,6 +14,19 @@
 
 use strict qw{vars subs};
 
+my $user;
+if (open (IN,"/usr/syno/synoman/webman/modules/authenticate.cgi|")) {
+	$user=<IN>;
+	chop($user);
+	close(IN);
+}
+# if not admin or no user at all...no authentication...so, bye-bye
+if ($user ne 'admin') {
+	print "Content-type: text/html\n\n";
+	print "<HTML><HEAD><TITLE>Login Required</TITLE></HEAD><BODY>Please login as admin first, before using this webpage</BODY></HTML>\n";
+	die;
+}
+
 $VERSION     = '1.11',
 $AUTHOR      = 'Gilles DAROLD <gilles AT darold DOT net>';
 $COPYRIGHT   = 'Copyright &copy; 2010-2012 Gilles DAROLD, all rights reserved';
