--- cgi-bin/squidguardmgr.cgi.orig	2014-12-05 19:10:41.477258783 +0100
+++ cgi-bin/squidguardmgr.cgi	2014-12-05 19:35:06.081331023 +0100
@@ -14,6 +14,62 @@
 
 use strict qw{vars subs};
 
+my $user;
+my $admin=0;
+if (open (IN,"/usr/syno/synoman/webman/modules/authenticate.cgi|")) {
+	$user=<IN>;
+	chop($user);
+	close(IN);
+}
+open GROUP, "</etc/group"
+   or die "Could not find /etc/group: $!\n";
+while (<GROUP>) {
+  my ($group,$x,$gid,$list)=split(':',$_);
+  chop($list);
+  if ($group eq "administrators") {
+	my (@users)=split(',',$list);
+	$admin = 1 if (grep {$_ eq $user} @users);
+  }
+}
+
+if ( `synoldapclient --get-user $user | grep " administrators@"` ){
+        $admin = 1;
+}
+
+# if not admin or no user at all...no authentication...so, bye-bye
+if ($admin == 0) {
+	print "Content-type: text/html\n\n";
+	print "<HTML><HEAD><TITLE>Login Required</TITLE></HEAD><BODY>Please login as admin first, before using this webpage</BODY></HTML>\n";
+	die;
+}
+
+our $LANG       = 'en_US';
+#
+# RETURN THE FIRST SUPPORTED LANGUAGE OF THE BROWSERS PREFERRED OR THE
+# DEFAULT:
+#
+sub getpreferedlang(@) {
+  my @supported = @_;
+  my @languages = split(/\s*,\s*/,$ENV{"HTTP_ACCEPT_LANGUAGE"}) if(defined($ENV{"HTTP_ACCEPT_LANGUAGE"}));
+  my $lang;
+  my $supp;
+  push(@languages,$supported[0]);
+  for $lang (@languages) {
+    $lang =~ s/-/_/;
+    if ($lang =~ /[a-z]{2}_[A-Z]{2}/ ) {
+    	for $supp (@supported) {
+      		$supp =~ s/\s.*//;
+      		return($lang) if ($lang eq $supp);
+      	}
+    }
+  }
+}
+our @supported   = (
+		"en_US",
+		"fr_FR"
+		);
+$LANG = getpreferedlang(@supported);
+
 $VERSION     = '1.14',
 $AUTHOR      = 'Gilles DAROLD <gilles AT darold DOT net>';
 $COPYRIGHT   = 'Copyright &copy; 2010-2015 Gilles DAROLD, all rights reserved';
@@ -24,8 +75,8 @@
 my $SC_PROGRAM  = 'SquidClamav Manager';
 
 # Default SquidGuard installation path
-my $DBHOME = "/var/lib/squidguard/db";
-my $LOGDIR = "/var/log/squidguard";
+my $DBHOME = "/usr/local/squidguard/var/db";
+my $LOGDIR = "/usr/local/squidguard/var/logs";
 
 # Variables that can be overidden into the squidguardmgr.conf file
 our $DIFF       = '/usr/bin/diff';
@@ -37,7 +88,6 @@
 our $BLDESC     = 'description.dat';
 our $CONF_FILE  = '/usr/local/squidGuard/squidguard.conf';
 our $LANGDIR    = 'lang';
-our $LANG       = 'en_US';
 our $IMG_DIR    = 'images';
 our $CSS_FILE   = 'squidguardmgr.css';
 our $JS_FILE    = 'squidguardmgr.js';
@@ -161,7 +211,7 @@
 	if ($VIEW eq 'squid') {
 		if ($SQUID_WRAPPER) {
 			print "Running: $SQUID_WRAPPER ... \n";
-			$ERROR = `$SQUID_WRAPPER 2>&1`;
+			$ERROR = `su squid -c $SQUID_WRAPPER 2>&1`;
 			&error($ERROR) if ($ERROR);
 			print "Ok." if (!$ERROR);
 		}
@@ -276,7 +326,7 @@
 
 if ($ACTION eq 'squid') {
 	if ($SQUID_WRAPPER) {
-		$ERROR = `$SQUID_WRAPPER`;
+		$ERROR = `su squid -c $SQUID_WRAPPER 2>&1`;
 	}
 	$ACTION = '';
 }
