# make prefix configurable
# 
--- Makefile.orig	2020-07-06 08:00:00.000000000 +0000
+++ Makefile	2021-05-16 23:32:38.309124941 +0000
@@ -30,10 +30,11 @@
 TARGETS = $(PROJECT) $(SHAREDLIB) $(STATICLIB)
 
 INSTALL = install
-libdir = /usr/lib
-libexecdir = /usr/lib
-includedir = /usr/include
-mandir = /usr/share/man
+prefix ?= /usr
+libdir = $(prefix)/lib
+libexecdir = $(prefix)/lib
+includedir = $(prefix)/include
+mandir = $(prefix)/share/man
 man3dir = $(mandir)/man3
 DESTDIR =
 
