--- arm-marvell-linux-gnueabi/libc/include/linux/ioctl.h.orig	2013-05-08 23:19:29.000000000 +0200
+++ arm-marvell-linux-gnueabi/libc/include/linux/ioctl.h	2013-05-08 23:19:40.000000000 +0200
@@ -1,7 +1,7 @@
 #ifndef _LINUX_IOCTL_H
 #define _LINUX_IOCTL_H
 
-#include <asm/ioctl.h>
+#include <asm-generic/ioctl.h>
 
 #endif /* _LINUX_IOCTL_H */
 
--- arm-marvell-linux-gnueabi/libc/include/linux/errno.h.orig	2013-05-08 23:20:13.000000000 +0200
+++ arm-marvell-linux-gnueabi/libc/include/linux/errno.h	2013-05-08 23:20:27.000000000 +0200
@@ -1,7 +1,7 @@
 #ifndef _LINUX_ERRNO_H
 #define _LINUX_ERRNO_H
 
-#include <asm/errno.h>
+#include <asm-generic/errno.h>
 
 #ifdef __KERNEL__
 
--- arm-marvell-linux-gnueabi/libc/include/asm-generic/int-ll64.h.orig	2013-05-02 14:24:09.000000000 +0200
+++ arm-marvell-linux-gnueabi/libc/include/asm-generic/int-ll64.h	2013-05-09 00:00:41.000000000 +0200
@@ -8,7 +8,7 @@
 #ifndef _ASM_GENERIC_INT_LL64_H
 #define _ASM_GENERIC_INT_LL64_H
 
-#include <asm/bitsperlong.h>
+#include <asm-generic/bitsperlong.h>
 
 #ifndef __ASSEMBLY__
 /*
