--- configure.orig	2013-04-24 18:41:28.729391574 +0200
+++ configure	2013-04-24 18:42:00.093391013 +0200
@@ -205,7 +205,7 @@
 
 if [ "$pie" = "1" ]; then
 	echo 'CFLAGS_BITLBEE=-fPIE' >> Makefile.settings
-	echo 'LDFLAGS_BITLBEE=-pie' >> Makefile.settings
+	echo 'LDFLAGS_BITLBEE=-pie -lz -lnettle -lgmp -lhogweed -lgcrypt -lgpg-error' >> Makefile.settings
 fi
 
 echo CFLAGS=$CFLAGS $CPPFLAGS >> Makefile.settings
