Fix undefined reference to 'rpl_malloc' when cross-compiling.

diff -wpurN pixz-1.0.7/configure.ac pixz-1.0.7.patched/configure.ac
--- pixz-1.0.7/configure.ac	2024-04-17 12:41:01.040203638 -0700
+++ pixz-1.0.7.patched/configure.ac	2024-04-17 12:41:49.224434589 -0700
@@ -65,10 +65,8 @@ AC_TYPE_UINT8_T
 AC_SYS_LARGEFILE
 
 # Checks for library functions.
-AC_FUNC_MALLOC
-AC_FUNC_REALLOC
 AC_FUNC_STRTOD
-AC_CHECK_FUNCS([memchr memmove memset strerror strtol])
+AC_CHECK_FUNCS([malloc realloc memchr memmove memset strerror strtol])
 AC_CHECK_HEADER([sys/endian.h],
                [
                  AC_CHECK_DECLS([htole64, le64toh], [], [], [
