--- configure.ac.orig	2014-11-27 03:45:33.000000000 +0100
+++ configure.ac	2015-12-06 20:31:52.179144837 +0100
@@ -348,11 +348,12 @@
 
 if test "x$debug" = xtrue; then
 	CPPFLAGS="-DDEBUG $CPPFLAGS"
-	CFLAGS=$(echo "$CFLAGS" | sed 's/-g//')
+#	CFLAGS=$(echo "$CFLAGS" | sed 's/-g//')    # PATCH flac-1.3.1. 01/12/2015
 	CFLAGS="-g $CFLAGS"
 else
 	CPPFLAGS="-DNDEBUG $CPPFLAGS"
-	CFLAGS=$(echo "$CFLAGS" | sed 's/-O2//;s/-g//')
+#	CFLAGS=$(echo "$CFLAGS" | sed 's/-O2//;s/-g//')       # PATCH flac-1.3.1. 01/12/2015
+	CFLAGS=$(echo "$CFLAGS" | sed 's/-O2//')              # PATCH flac-1.3.1. 01/12/2015
 	CFLAGS="-O3 -funroll-loops $CFLAGS"
 fi
 
