--- include/mpd/sticker.h.orig	2025-08-23 16:15:34
+++ include/mpd/sticker.h	2026-02-02 17:05:29
@@ -23,7 +23,8 @@
  * Comparison operators for sticker search api
  */
 enum mpd_sticker_operator {
-	MPD_STICKER_OP_UNKOWN = -1,
+	MPD_STICKER_OP_UNKOWN = -1,   // Keep typo, else we break the abi
+	MPD_STICKER_OP_UNKNOWN = -1,
 	MPD_STICKER_OP_EQ,
 	MPD_STICKER_OP_GT,
 	MPD_STICKER_OP_LT,
@@ -38,7 +39,8 @@
  * Sort by settings for sticker search api
  */
 enum mpd_sticker_sort {
-	MPD_STICKER_SORT_UNKOWN = -1,
+	MPD_STICKER_SORT_UNKOWN = -1,   // Keep typo, else we break the abi
+	MPD_STICKER_SORT_UNKNOWN = -1,
 	MPD_STICKER_SORT_URI,
 	MPD_STICKER_SORT_VALUE,
 	MPD_STICKER_SORT_VALUE_INT,
--- src/sticker.c.orig	2025-08-23 16:15:34
+++ src/sticker.c	2026-02-02 17:06:39
@@ -251,7 +251,7 @@
 	case MPD_STICKER_OP_LT_INT:      return "lt";
 	case MPD_STICKER_OP_CONTAINS:    return "contains";
 	case MPD_STICKER_OP_STARTS_WITH: return "starts_with";
-	case MPD_STICKER_OP_UNKOWN:      return NULL;
+	case MPD_STICKER_OP_UNKNOWN:     return NULL;
 	}
 	return NULL;
 }
@@ -293,7 +293,7 @@
 	case MPD_STICKER_SORT_URI:       return "uri";
 	case MPD_STICKER_SORT_VALUE:     return "value";
 	case MPD_STICKER_SORT_VALUE_INT: return "value_int";
-	case MPD_STICKER_SORT_UNKOWN:    return NULL;
+	case MPD_STICKER_SORT_UNKNOWN:   return NULL;
 	}
 	return NULL;
 }
