Description: Fix and FTBFS on ia64, s390x, alpha, and ppc64.
Author: Jackson Doak <noskcaj@ubuntu.com>
Bug-Debian: http://bugs.debian.org/726186
Last-Update: 2013-10-15
Index: zvbi-0.2.35/test/test-packet-830.cc
===================================================================
--- zvbi-0.2.35.orig/test/test-packet-830.cc
+++ zvbi-0.2.35/test/test-packet-830.cc
@@ -115,7 +115,7 @@ assert_decode_teletext_8301_cni	(unsigne
 	unsigned int cni2;
 
 	memcpy (buffer2, buffer, sizeof (buffer2));
-	memset_rand (cni, sizeof (cni));
+	memset_rand (cni, sizeof(*cni));
 	cni2 = *cni;
 
 	assert (TRUE == vbi_decode_teletext_8301_cni (cni, buffer));
