# Fix lib dir
--- configure.ac.orig	2014-11-12 12:56:51.000000000 +0100
+++ configure.ac	2014-12-08 00:30:26.440532517 +0100
@@ -601,12 +601,6 @@
     toolexecdir="${libdir}"/gcc-lib/'$(target_alias)'
     toolexeclibdir="${libdir}"
   fi
-  multi_os_directory=`$CC $CFLAGS -print-multi-os-directory`
-  case $multi_os_directory in
-    .) ;; # Avoid trailing /.
-    ../*) toolexeclibdir=$toolexeclibdir/$multi_os_directory ;;
-  esac
-  AC_SUBST(toolexecdir)
 else
   toolexeclibdir="${libdir}"
 fi
