--- configure.ac.orig	2014-09-01 23:57:51.126855074 +0200
+++ configure.ac	2014-09-02 00:05:53.512790663 +0200
@@ -596,12 +596,6 @@
     toolexecdir='$(libdir)/gcc-lib/$(target_alias)'
     toolexeclibdir='$(libdir)'
   fi
-  multi_os_directory=`$CC $CFLAGS -print-multi-os-directory`
-  case $multi_os_directory in
-    .) ;; # Avoid trailing /.
-    ../*) toolexeclibdir=$toolexeclibdir/$multi_os_directory ;;
-  esac
-  AC_SUBST(toolexecdir)
 else
   toolexeclibdir='$(libdir)'
 fi
