--- src/powerpc/ffi.c.orig	2012-04-12 04:46:06.000000000 +0200
+++ src/powerpc/ffi.c	2012-09-30 10:30:25.000000000 +0200
@@ -664,8 +664,10 @@
   switch (type)
     {
 #ifndef __NO_FPRS__
+#if FFI_TYPE_LONGDOUBLE != FFI_TYPE_DOUBLE
     case FFI_TYPE_LONGDOUBLE:
       flags |= FLAG_RETURNS_128BITS;
+#endif
       /* Fall through.  */
     case FFI_TYPE_DOUBLE:
       flags |= FLAG_RETURNS_64BITS;
