--- install.sh.orig	2016-06-12 10:42:37.409345181 +0200
+++ install.sh	2016-06-12 10:42:29.305345113 +0200
@@ -180,8 +180,6 @@
   echo "- installing HTML help in ${DEST_DIR}${DEST_SHARE_DOC}/DOC"
   mkdir -p "${DEST_DIR}${DEST_SHARE_DOC}"
   cp -r DOC "${DEST_DIR}${DEST_SHARE_DOC}/DOC"
-  find "${DEST_DIR}${DEST_SHARE_DOC}/DOC" -type d -exec chmod 755 {} \;
-  find "${DEST_DIR}${DEST_SHARE_DOC}/DOC" -type f -exec chmod 644 {} \;
 fi
 
 use_lang="n"
@@ -202,7 +200,5 @@
 then
   echo "- installing Lang in ${DEST_DIR}${DEST_SHARE}"
   cp -r GUI/Lang "${DEST_DIR}${DEST_SHARE}/"
-  find "${DEST_DIR}${DEST_SHARE}/Lang" -type d -exec chmod 755 {} \;
-  find "${DEST_DIR}${DEST_SHARE}/Lang" -type f -exec chmod 644 {} \;
 fi
 
