--- Makefile.orig	2019-05-11 14:29:01.000000000 +0200
+++ Makefile	2021-02-10 13:07:34.283425013 +0100
@@ -1,4 +1,4 @@
-PREFIX ?= /usr/local
+PREFIX ?= /usr/local/elixir
 SHARE_PREFIX ?= $(PREFIX)/share
 MAN_PREFIX ?= $(SHARE_PREFIX)/man
 CANONICAL := v1.8/ # master/ or vMAJOR.MINOR/
@@ -109,7 +109,7 @@
 	$(Q) $(INSTALL_PROGRAM) $(filter-out %.ps1, $(filter-out %.bat, $(wildcard bin/*))) "$(DESTDIR)$(PREFIX)/$(LIBDIR)/elixir/bin"
 	$(Q) $(INSTALL_DIR) "$(DESTDIR)$(PREFIX)/$(BINDIR)"
 	$(Q) for file in "$(DESTDIR)$(PREFIX)"/$(LIBDIR)/elixir/bin/*; do \
-		ln -sf "../$(LIBDIR)/elixir/bin/$${file##*/}" "$(DESTDIR)$(PREFIX)/$(BINDIR)/"; \
+		ln -sf "$(DESTDIR)$(PREFIX)/$(LIBDIR)/elixir/bin/$${file##*/}" "$(DESTDIR)$(PREFIX)/$(BINDIR)/"; \
 	done
 	$(MAKE) install_man
 
