# add arm64 (aarch64) configuration

--- Makefile.org	2020-01-18 00:44:37.000000000 +0000
+++ Makefile	2020-05-03 18:25:09.557775228 +0000
@@ -129,6 +129,12 @@
 	prefetch = yes
 endif
 
+ifeq ($(ARCH),arm64)
+	arch = aarch64
+	bits = 64
+	prefetch = yes
+endif
+
 ifeq ($(ARCH),ppc-32)
 	arch = ppc
 endif
@@ -166,8 +172,10 @@
 			LDFLAGS += -m$(bits)
 		endif
 	else
-		CXXFLAGS += -m$(bits)
-		LDFLAGS += -m$(bits)
+		ifneq ($(ARCH),arm64)
+			CXXFLAGS += -m$(bits)
+			LDFLAGS += -m$(bits)
+		endif
 	endif
 
 	ifneq ($(KERNEL),Darwin)
@@ -387,6 +395,7 @@
 	@echo "ppc-64                  > PPC 64-bit"
 	@echo "ppc-32                  > PPC 32-bit"
 	@echo "armv7                   > ARMv7 32-bit"
+	@echo "arm64                   > ARM 64-bit"
 	@echo "general-64              > unspecified 64-bit"
 	@echo "general-32              > unspecified 32-bit"
 	@echo ""
@@ -488,7 +497,8 @@
 	@test "$(sanitize)" = "undefined" || test "$(sanitize)" = "thread" || test "$(sanitize)" = "address" || test "$(sanitize)" = "no"
 	@test "$(optimize)" = "yes" || test "$(optimize)" = "no"
 	@test "$(arch)" = "any" || test "$(arch)" = "x86_64" || test "$(arch)" = "i386" || \
-	 test "$(arch)" = "ppc64" || test "$(arch)" = "ppc" || test "$(arch)" = "armv7"
+	 test "$(arch)" = "ppc64" || test "$(arch)" = "ppc" || test "$(arch)" = "armv7" || \
+	 test "$(arch)" = "aarch64"
 	@test "$(bits)" = "32" || test "$(bits)" = "64"
 	@test "$(prefetch)" = "yes" || test "$(prefetch)" = "no"
 	@test "$(popcnt)" = "yes" || test "$(popcnt)" = "no"
