--- media_driver/agnostic/common/os/mos_utilities.h.orig	2019-12-26 10:28:03.000000000 -0500
+++ media_driver/agnostic/common/os/mos_utilities.h	2019-12-31 11:33:00.012370485 -0500
@@ -74,8 +74,8 @@ public:
     void setupFilePath(char *perfFilePath);
     void setupFilePath();
     bool bPerfUtilityKey;
-    char sSummaryFileName[MOS_MAX_PERF_FILENAME_LEN + 1] = "";
-    char sDetailsFileName[MOS_MAX_PERF_FILENAME_LEN + 1] = "";
+    char sSummaryFileName[MOS_MAX_PERF_FILENAME_LEN + 1] = {};
+    char sDetailsFileName[MOS_MAX_PERF_FILENAME_LEN + 1] = {};
     int32_t dwPerfUtilityIsEnabled;
 
 private:
