--- cmrtlib/linux/CMakeLists.txt	2019-02-03 03:49:50.000000000 -0600
+++ ../media-driver/cmrtlib/linux/CMakeLists.txt	2019-09-02 13:19:16.500788424 -0500
@@ -18,7 +18,7 @@
 # ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 # OTHER DEALINGS IN THE SOFTWARE.
 
-cmake_minimum_required(VERSION 2.8)
+cmake_minimum_required(VERSION 3.5)
 include (${CMAKE_CURRENT_LIST_DIR}/cmrt_utils.cmake)
 
 set(EXECUTABLE_OUTPUT_PATH ${MDF_OPTION__OUTPUT_DIR})
