--- arm-marvell-linux-gnu/lib/gcc/arm-marvell-linux-gnu/3.4.3/include/limits.h.orig	2012-11-18 12:01:59.000000000 +0100
+++ arm-marvell-linux-gnu/lib/gcc/arm-marvell-linux-gnu/3.4.3/include/limits.h	2012-11-18 12:02:39.000000000 +0100
@@ -1,3 +1,15 @@
+/* This administrivia gets added to the beginning of limits.h
+   if the system has its own version of limits.h.  */
+
+/* We use _GCC_LIMITS_H_ because we want this not to match
+   any macros that the system's limits.h uses for its own purposes.  */
+#ifndef _GCC_LIMITS_H_  /* Terminated in limity.h.  */
+#define _GCC_LIMITS_H_
+
+#ifndef _LIBC_LIMITS_H_
+/* Use "..." so that we find syslimits.h only in this same directory.  */
+#include "syslimits.h"
+#endif
 #ifndef _LIMITS_H___
 #define _LIMITS_H___
 
@@ -101,3 +113,13 @@
 #endif
 
 #endif /* _LIMITS_H___ */
+/* This administrivia gets added to the end of limits.h
+   if the system has its own version of limits.h.  */
+
+#else /* not _GCC_LIMITS_H_ */
+
+#ifdef _GCC_NEXT_LIMITS_H
+#include_next <limits.h>		/* recurse down to the real one */
+#endif
+
+#endif /* not _GCC_LIMITS_H_ */
