--- crypto/Makefile.orig	2023-07-01 11:45:34.201947974 +1200
+++ crypto/Makefile	2023-07-01 11:45:37.641947937 +1200
@@ -134,6 +134,5 @@
 obj-$(CONFIG_XOR_BLOCKS) += xor.o
 obj-$(CONFIG_ASYNC_CORE) += async_tx/
 obj-$(CONFIG_ASYMMETRIC_KEY_TYPE) += asymmetric_keys/
-obj-$(CONFIG_SYNO_RAMDISK_INTEGRITY_CHECK) += hydrogen/
 obj-$(CONFIG_CRYPTO_HASH_INFO) += hash_info.o
 obj-$(CONFIG_CRYPTO_ABLK_HELPER) += ablk_helper.o
