diff -uprN ../ffmpeg-4.3-033-evansport-002/fftools/ffmpeg.c ./fftools/ffmpeg.c
--- ../ffmpeg-4.3-033-evansport-002/fftools/ffmpeg.c	2020-06-28 07:55:28.264003640 -0400
+++ ./fftools/ffmpeg.c	2020-06-28 08:10:39.816594350 -0400
@@ -172,6 +172,9 @@ typedef struct X264Context {
     int nal_hrd;
     int avcintra_class;
     char *x264_params;
+#ifdef SYNO_EVANSPORT_FLV_EXTRADATA
+	bool dont_free_extradata;
+#endif
 } X264Context;
 
 static viddec_struct_t vid_dec;
@@ -5224,7 +5226,14 @@ discard_packet:
 			if (ost->encoding_needed)
 			{
 				//close H.264 smd encoder because transcoding will be done through pipeline
+#ifdef SYNO_EVANSPORT_FLV_EXTRADATA
+				X264Context *x4 = ost->enc_ctx->priv_data;
+				x4->dont_free_extradata = true;
+#endif
 				ost->enc_ctx->codec->close(ost->enc_ctx);
+#ifdef SYNO_EVANSPORT_FLV_EXTRADATA
+				x4->dont_free_extradata = false;
+#endif
 				ost->enc_ctx->coded_frame = av_frame_alloc();
 				if (!ost->enc_ctx->coded_frame)
 					return AVERROR(ENOMEM);
diff -uprN ../ffmpeg-4.3-033-evansport-002/libavcodec/libx264.c ./libavcodec/libx264.c
--- ../ffmpeg-4.3-033-evansport-002/libavcodec/libx264.c	2020-06-28 07:55:28.268003913 -0400
+++ ./libavcodec/libx264.c	2020-06-28 08:12:17.332141843 -0400
@@ -125,6 +125,11 @@ typedef struct X264Context {
      * encounter a frame with ROI side data.
      */
     int roi_warned;
+#ifdef SYNO_EVANSPORT_SMD
+#ifdef SYNO_EVANSPORT_FLV_EXTRADATA
+    bool dont_free_extradata;
+#endif
+#endif
 } X264Context;
 
 static void X264_log(void *p, int level, const char *fmt, va_list args)
@@ -652,7 +657,9 @@ static av_cold int X264_close(AVCodecCon
     X264Context *x4 = avctx->priv_data;
 
 #ifdef SYNO_EVANSPORT_SMD
+#ifdef SYNO_EVANSPORT_FLV_EXTRADATA
 	if (!x264_closed && !x4->dont_free_extradata)
+#endif
 	{
 #endif
     av_freep(&avctx->extradata);
diff -uprN ../ffmpeg-4.3-033-evansport-002/synoconfig.h ./synoconfig.h
--- ../ffmpeg-4.3-033-evansport-002/synoconfig.h	2020-06-28 08:07:33.445068259 -0400
+++ ./synoconfig.h	2020-06-28 08:10:39.817594417 -0400
@@ -74,3 +74,10 @@
 #if defined(SYNO_EVANSPORT_SMD)
 #define SYNO_EVANSPORT_AUTO_HW_DECODER
 #endif
+ 
+/* Fixed evansport coredump transcode to flv on pipeline mode
+ * See Video Station #2132
+ */
+#if defined(SYNO_EVANSPORT_SMD)
+#define SYNO_EVANSPORT_FLV_EXTRADATA
+#endif
