--- SimpleModule-orig.h	2021-03-16 20:16:01.805939179 +0100
+++ libaudiofile/modules/SimpleModule.h	2021-03-16 20:18:22.617932020 +0100
@@ -122,7 +122,7 @@
 	typedef typename IntTypes<Format>::UnsignedType UnsignedType;
 
 	static const int kScaleBits = (Format + 1) * CHAR_BIT - 1;
-	static const int kMinSignedValue = -1 << kScaleBits;
+	static const int kMinSignedValue = static_cast<signed>(static_cast<unsigned>(-1) << kScaleBits);
 
 	struct signedToUnsigned : public std::unary_function<SignedType, UnsignedType>
 	{
