diff -uprN ../ffmpeg-5.1.3-0110/libavcodec/srtenc.c ./libavcodec/srtenc.c
--- ../ffmpeg-5.1.3-0110/libavcodec/srtenc.c	2023-03-26 14:21:00.000000000 +0000
+++ ./libavcodec/srtenc.c	2023-04-24 23:43:36.351930175 +0000
@@ -29,6 +29,7 @@
 #include "ass.h"
 #include "codec_internal.h"
 
+#include "synoconfig.h"
 
 #define SRT_STACK_SIZE 64
 
@@ -196,7 +197,9 @@ static void srt_alignment_cb(void *priv,
 static void srt_cancel_overrides_cb(void *priv, const char *style)
 {
     srt_stack_push_pop(priv, 0, 1);
+#ifndef SYNO_FILTER_SRT_TAGS
     srt_style_apply(priv, style);
+#endif
 }
 
 static void srt_move_cb(void *priv, int x1, int y1, int x2, int y2,
@@ -224,6 +227,21 @@ static const ASSCodesCallbacks srt_callb
     .end              = srt_end_cb,
 };
 
+#ifdef SYNO_FILTER_SRT_TAGS
+static const ASSCodesCallbacks srt_callbacks_syno_notag = {
+    .text             = srt_text_cb,
+    .new_line         = srt_new_line_cb,
+    .style            = NULL,
+    .color            = NULL,
+    .font_name        = NULL,
+    .font_size        = NULL,
+    .alignment        = NULL,
+    .cancel_overrides = srt_cancel_overrides_cb,
+    .move             = NULL,
+    .end              = srt_end_cb,
+};
+#endif
+
 static const ASSCodesCallbacks text_callbacks = {
     .text             = srt_text_cb,
     .new_line         = srt_new_line_cb,
diff -uprN ../ffmpeg-5.1.3-0110/synoconfig.h ./synoconfig.h
--- ../ffmpeg-5.1.3-0110/synoconfig.h	2023-04-24 23:41:03.000000000 +0000
+++ ./synoconfig.h	2023-04-24 23:43:36.351930175 +0000
@@ -45,3 +45,11 @@
 #if defined(SYNO_VIDEOSTATION)
 #define SYNO_SMOOTH_STREAMING
 #endif
+ 
+/* filter out some tag that not needed
+ * those tags producted at *.ass convert *.srt
+ * See Video Station #1031
+ */
+#if defined(SYNO_VIDEOSTATION)
+#define SYNO_FILTER_SRT_TAGS
+#endif
