Index: jellyfin-ffmpeg/libavutil/vulkan.h
===================================================================
--- libavutil/vulkan.h
+++ libavutil/vulkan.h
@@ -26,9 +26,10 @@
 #include "hwcontext_vulkan.h"
 #include "vulkan_loader.h"
 
-#define FF_VK_DEFAULT_USAGE_FLAGS (VK_IMAGE_USAGE_SAMPLED_BIT      |           \
-                                   VK_IMAGE_USAGE_STORAGE_BIT      |           \
-                                   VK_IMAGE_USAGE_TRANSFER_SRC_BIT |           \
+#define FF_VK_DEFAULT_USAGE_FLAGS (VK_IMAGE_USAGE_SAMPLED_BIT          |       \
+                                   VK_IMAGE_USAGE_STORAGE_BIT          |       \
+                                   VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT |       \
+                                   VK_IMAGE_USAGE_TRANSFER_SRC_BIT     |       \
                                    VK_IMAGE_USAGE_TRANSFER_DST_BIT)
 
 /* GLSL management macros */
