Index: jellyfin-ffmpeg/libavutil/hwcontext_vaapi.c
===================================================================
--- libavutil/hwcontext_vaapi.c
+++ libavutil/hwcontext_vaapi.c
@@ -1692,7 +1692,7 @@ static int vaapi_device_create(AVHWDevic
                 if (priv->drm_fd < 0) {
                     av_log(ctx, AV_LOG_VERBOSE, "Cannot open "
                            "DRM render node for device %d.\n", n);
-                    break;
+                    continue;
                 }
 #if CONFIG_LIBDRM
                 if (kernel_driver) {
