Index: jellyfin-ffmpeg/libavcodec/qsv.c
===================================================================
--- libavcodec/qsv.c
+++ libavcodec/qsv.c
@@ -369,9 +369,17 @@ enum AVPictureType ff_qsv_map_pictype(in
 static int qsv_load_plugins(mfxSession session, const char *load_plugins,
                             void *logctx)
 {
+    mfxVersion ver;
     if (!load_plugins || !*load_plugins)
         return 0;
 
+    // Plugins have been removed in VPL runtime, don't load them
+    // if using the VPL runtime with MSDK loader to avoid failure.
+    if (MFXQueryVersion(session, &ver) == MFX_ERR_NONE &&
+        QSV_RUNTIME_VERSION_ATLEAST(ver, 1, 255)) {
+        return 0;
+    }
+
     while (*load_plugins) {
         mfxPluginUID uid;
         mfxStatus ret;
