--- deluge/plugins/scheduler/scheduler/core.py.org	2015-11-09 00:00:00.000000000 +0100
+++ deluge/plugins/scheduler/scheduler/core.py	2015-11-09 00:00:00.000000000 +0100
@@ -127,7 +127,7 @@
         """
         core_config = deluge.configmanager.ConfigManager("core.conf")
         for setting in CONTROLLED_SETTINGS:
-            component.get("PreferencesManager").do_config_set_func(setting, core_config[setting])
+            core_config.apply_set_functions(setting)
         # Resume the session if necessary
         component.get("Core").session.resume()
 
