--- Cargo.toml.orig	2025-03-28 06:50:41.000000000 +0000
+++ Cargo.toml	2025-07-27 11:05:19.323644906 +0000
@@ -53,6 +53,13 @@
 toml          = "0.8"
 unicode-width = "0.2"
 
+# Dependencies pinned for Rust 1.77.2
+litemap = "=0.7.4"
+potential_utf = "=0.1.0"
+tinystr = "=0.8.0"
+zerofrom = "=0.1.5"
+
+
 [build-dependencies]
 anyhow        = "1.0"
 clap          = {version = "4.4", features = ["derive"]}
