# Disable error for configure curl
--- configure.orig	2015-05-08 17:52:24.000000000 +0200
+++ configure	2015-05-08 18:33:21.860927072 +0200
@@ -14004,7 +14004,7 @@
     if test "$cross_compiling" = yes; then :
   { { $as_echo "$as_me:${as_lineno-$LINENO}: error: in \`$ac_pwd':" >&5
 $as_echo "$as_me: error: in \`$ac_pwd':" >&2;}
-as_fn_error $? "cannot run test program while cross compiling
+$as_echo $? "cannot run test program while cross compiling
 See \`config.log' for more details" "$LINENO" 5; }
 else
   cat confdefs.h - <<_ACEOF >conftest.$ac_ext
