--- include/ac/string.h.orig	2015-02-22 22:07:35.318501814 +0100
+++ include/ac/string.h	2015-02-22 22:07:58.983082002 +0100
@@ -89,11 +89,6 @@
 		else AC_MEMCPY( (d), (s), (n) ); \
 	} while(0)
 
-#ifdef NEED_MEMCMP_REPLACEMENT
-	int (lutil_memcmp)(const void *b1, const void *b2, size_t len);
-#define memcmp lutil_memcmp
-#endif
-
 void *(lutil_memrchr)(const void *b, int c, size_t n);
 /* GNU extension (glibc >= 2.1.91), only declared when defined(_GNU_SOURCE) */
 #if defined(HAVE_MEMRCHR) && defined(_GNU_SOURCE)
