diff -uprN ../ffmpeg-7.1.3/configure ./configure
--- ../ffmpeg-7.1.3/configure
+++ ./configure
@@ -8074,7 +8074,7 @@
 SRC_PATH:=\$(SRC_PATH:.%=..%)
 endif
 CC_IDENT=$cc_ident
-ARCH=$arch
+FFMPEG_ARCH=$arch
 INTRINSICS=$intrinsics
 EXTERN_PREFIX=$extern_prefix
 CC=$cc
diff -uprN ../ffmpeg-7.1.3/ffbuild/common.mak ./ffbuild/common.mak
--- ../ffmpeg-7.1.3/ffbuild/common.mak
+++ ./ffbuild/common.mak
@@ -186,8 +186,8 @@
 STATIC_DEP_LIBS := $(foreach lib,$(FFLIBS),$(call PATH_LIBNAME,$(lib)))
 
 SRC_DIR    := $(SRC_PATH)/lib$(NAME)
-ALLHEADERS := $(subst $(SRC_DIR)/,$(SUBDIR),$(wildcard $(SRC_DIR)/*.h $(SRC_DIR)/$(ARCH)/*.h))
-SKIPHEADERS += $(ARCH_HEADERS:%=$(ARCH)/%) $(SKIPHEADERS-)
+ALLHEADERS := $(subst $(SRC_DIR)/,$(SUBDIR),$(wildcard $(SRC_DIR)/*.h $(SRC_DIR)/$(FFMPEG_ARCH)/*.h))
+SKIPHEADERS += $(ARCH_HEADERS:%=$(FFMPEG_ARCH)/%) $(SKIPHEADERS-)
 SKIPHEADERS := $(SKIPHEADERS:%=$(SUBDIR)%)
 HOBJS        = $(filter-out $(SKIPHEADERS:.h=.h.o),$(ALLHEADERS:.h=.h.o))
 PTXOBJS      = $(filter %.ptx.o,$(OBJS))
diff -uprN ../ffmpeg-7.1.3/ffbuild/library.mak ./ffbuild/library.mak
--- ../ffmpeg-7.1.3/ffbuild/library.mak
+++ ./ffbuild/library.mak
@@ -75,7 +75,7 @@
 
 clean::
 	$(RM) $(addprefix $(SUBDIR),$(CLEANFILES) $(CLEANSUFFIXES) $(LIBSUFFIXES)) \
-	    $(CLEANSUFFIXES:%=$(SUBDIR)$(ARCH)/%) $(CLEANSUFFIXES:%=$(SUBDIR)tests/%)
+	    $(CLEANSUFFIXES:%=$(SUBDIR)$(FFMPEG_ARCH)/%) $(CLEANSUFFIXES:%=$(SUBDIR)tests/%)
 
 install-lib$(NAME)-shared: $(SUBDIR)$(SLIBNAME)
 	$(Q)mkdir -p "$(SHLIBDIR)"
diff -uprN ../ffmpeg-7.1.3/libavcodec/Makefile ./libavcodec/Makefile
--- ../ffmpeg-7.1.3/libavcodec/Makefile
+++ ./libavcodec/Makefile
@@ -69,7 +69,7 @@
 include $(SRC_PATH)/libavcodec/hevc/Makefile
 include $(SRC_PATH)/libavcodec/opus/Makefile
 include $(SRC_PATH)/libavcodec/vvc/Makefile
--include $(SRC_PATH)/libavcodec/$(ARCH)/vvc/Makefile
+-include $(SRC_PATH)/libavcodec/$(FFMPEG_ARCH)/vvc/Makefile
 OBJS-$(CONFIG_AANDCTTABLES)            += aandcttab.o
 OBJS-$(CONFIG_AC3DSP)                  += ac3dsp.o ac3.o ac3tab.o
 OBJS-$(CONFIG_ADTS_HEADER)             += adts_header.o mpeg4audio_sample_rates.o
diff -uprN ../ffmpeg-7.1.3/Makefile ./Makefile
--- ../ffmpeg-7.1.3/Makefile
+++ ./Makefile
@@ -115,7 +115,7 @@
 $(foreach V,$(SUBDIR_VARS),$(eval $(call RESET,$(V))))
 SUBDIR := $(1)/
 include $(SRC_PATH)/$(1)/Makefile
--include $(SRC_PATH)/$(1)/$(ARCH)/Makefile
+-include $(SRC_PATH)/$(1)/$(FFMPEG_ARCH)/Makefile
 -include $(SRC_PATH)/$(1)/$(INTRINSICS)/Makefile
 include $(SRC_PATH)/ffbuild/library.mak
 endef
diff -uprN ../ffmpeg-7.1.3/tests/checkasm/Makefile ./tests/checkasm/Makefile
--- ../ffmpeg-7.1.3/tests/checkasm/Makefile
+++ ./tests/checkasm/Makefile
@@ -106,6 +106,6 @@
 testclean:: checkasmclean
 
 checkasmclean:
-	$(RM) $(CHECKASM) $(CLEANSUFFIXES:%=tests/checkasm/%) $(CLEANSUFFIXES:%=tests/checkasm/$(ARCH)/%)
+	$(RM) $(CHECKASM) $(CLEANSUFFIXES:%=tests/checkasm/%) $(CLEANSUFFIXES:%=tests/checkasm/$(FFMPEG_ARCH)/%)
 
 .PHONY: checkasm
