diff -uprN ../ffmpeg-4.3-020/libavformat/segment.c ./libavformat/segment.c
--- ../ffmpeg-4.3-020/libavformat/segment.c	2020-06-16 19:54:41.000000000 -0400
+++ ./libavformat/segment.c	2020-06-16 19:57:31.891853716 -0400
@@ -237,9 +237,15 @@ static int set_segment_filename(AVFormat
 
     if ((ret = av_reallocp(&seg->cur_entry.filename, size)) < 0)
         return ret;
+#ifdef SYNO_VIDEOSTATION_HLS_COSTUMIZED_TS_NAME
+    snprintf(seg->cur_entry.filename, size, "%s%03d",
+             seg->entry_prefix ? seg->entry_prefix : "",
+             seg->segment_idx);
+#else
     snprintf(seg->cur_entry.filename, size, "%s%s",
              seg->entry_prefix ? seg->entry_prefix : "",
              av_basename(oc->url));
+#endif
 
     return 0;
 }
diff -uprN ../ffmpeg-4.3-020/synoconfig.h ./synoconfig.h
--- ../ffmpeg-4.3-020/synoconfig.h	2020-06-16 19:53:14.000000000 -0400
+++ ./synoconfig.h	2020-06-16 19:57:31.891853716 -0400
@@ -10,3 +10,7 @@
 #if defined(SYNO_VIDEOSTATION)
 #define SYNO_VIDEOSTATION_HLS_SEEK_TIME
 #endif
+
+#if defined(SYNO_VIDEOSTATION)
+#define SYNO_VIDEOSTATION_HLS_COSTUMIZED_TS_NAME
+#endif
