--- ../ffmpeg-7.0.2/libavformat/movenc.c	2024-08-02 22:55:25.000000000 +0000
+++ ./libavformat/movenc.c	2024-09-24 23:27:44.198262220 +0000
@@ -124,6 +124,9 @@ static const AVOption options[] = {
       { "pts", NULL, 0, AV_OPT_TYPE_CONST, {.i64 = MOV_PRFT_SRC_PTS}, 0, 0, AV_OPT_FLAG_ENCODING_PARAM, .unit = "prft"},
       { "wallclock", NULL, 0, AV_OPT_TYPE_CONST, {.i64 = MOV_PRFT_SRC_WALLCLOCK}, 0, 0, AV_OPT_FLAG_ENCODING_PARAM, .unit = "prft"},
     { "write_tmcd", "force or disable writing tmcd", offsetof(MOVMuxContext, write_tmcd), AV_OPT_TYPE_BOOL, {.i64 = -1}, -1, 1, AV_OPT_FLAG_ENCODING_PARAM},
+#ifdef SYNO_VIDEOSTATION_SKIP_DISPLAYMATRIX
+    { "skip_displaymatrix", "Skip writing displaymatrix metadata.", offsetof(MOVMuxContext, skip_displaymatrix), AV_OPT_TYPE_BOOL, {.i64 = 0}, 0, 1, AV_OPT_FLAG_ENCODING_PARAM},
+#endif
     { NULL },
 };
 
@@ -3505,7 +3508,13 @@ static int mov_write_tkhd_tag(AVIOContex
                                      st->codecpar->nb_coded_side_data,
                                      AV_PKT_DATA_DISPLAYMATRIX);
         if (sd && sd->size == 9 * sizeof(*display_matrix))
+#ifdef SYNO_VIDEOSTATION_SKIP_DISPLAYMATRIX
+            if (mov->skip_displaymatrix) {
+               display_matrix = NULL;
+            }
+#else
             display_matrix = (uint32_t *)sd->data;
+#endif
     }
 
     if (track->flags & MOV_TRACK_ENABLED)
--- ../ffmpeg-7.0.2/libavformat/movenc.h	2024-08-02 22:55:22.000000000 +0000
+++ ./libavformat/movenc.h	2024-09-24 23:19:32.992183584 +0000
@@ -28,6 +28,8 @@
 #include "movenccenc.h"
 #include "libavcodec/packet_internal.h"
 
+#include "synoconfig.h"
+
 #define MOV_FRAG_INFO_ALLOC_INCREMENT 64
 #define MOV_INDEX_CLUSTER_SIZE 1024
 #define MOV_TIMESCALE 1000
@@ -205,6 +207,9 @@ typedef struct MOVMuxContext {
     int rtp_flags;
 
     int iods_skip;
+#ifdef SYNO_VIDEOSTATION_SKIP_DISPLAYMATRIX
+    int skip_displaymatrix;
+#endif
     int iods_video_profile;
     int iods_audio_profile;
 
