Index: FFmpeg/libavutil/hwcontext_opencl.c
===================================================================
--- libavutil/hwcontext_opencl.c
+++ libavutil/hwcontext_opencl.c
@@ -62,6 +62,9 @@
 #endif
 
 #if HAVE_OPENCL_D3D11
+#if CONFIG_LIBMFX
+#include "hwcontext_qsv.h"
+#endif
 #include <CL/cl_d3d11.h>
 #include "hwcontext_d3d11va.h"
 
@@ -139,6 +142,7 @@ typedef struct OpenCLDeviceContext {
 
 #if HAVE_OPENCL_D3D11
     int d3d11_mapping_usable;
+    int d3d11_qsv_mapping_usable;
     int d3d11_map_amd;
     int d3d11_map_intel;
     clCreateFromD3D11Texture2DKHR_fn
@@ -906,6 +910,11 @@ static int opencl_device_init(AVHWDevice
             priv->d3d11_mapping_usable = 0;
         } else {
             priv->d3d11_mapping_usable = 1;
+
+            if (priv->d3d11_map_intel)
+                priv->d3d11_qsv_mapping_usable = 1;
+            else
+                priv->d3d11_qsv_mapping_usable = 0;
         }
     }
 #endif
@@ -1785,18 +1794,20 @@ static void opencl_frames_uninit(AVHWFra
 
 #if HAVE_OPENCL_DXVA2 || HAVE_OPENCL_D3D11
     int i, p;
-    for (i = 0; i < priv->nb_mapped_frames; i++) {
-        AVOpenCLFrameDescriptor *desc = &priv->mapped_frames[i];
-        for (p = 0; p < desc->nb_planes; p++) {
-            cle = clReleaseMemObject(desc->planes[p]);
-            if (cle != CL_SUCCESS) {
-                av_log(hwfc, AV_LOG_ERROR, "Failed to release mapped "
-                       "frame object (frame %d plane %d): %d.\n",
-                       i, p, cle);
+    if (priv->nb_mapped_frames && priv->mapped_frames) {
+        for (i = 0; i < priv->nb_mapped_frames; i++) {
+            AVOpenCLFrameDescriptor *desc = &priv->mapped_frames[i];
+            for (p = 0; p < desc->nb_planes; p++) {
+                cle = clReleaseMemObject(desc->planes[p]);
+                if (cle != CL_SUCCESS) {
+                    av_log(hwfc, AV_LOG_ERROR, "Failed to release mapped "
+                           "frame object (frame %d plane %d): %d.\n",
+                           i, p, cle);
+                }
             }
         }
+        av_freep(&priv->mapped_frames);
     }
-    av_freep(&priv->mapped_frames);
 #endif
 
     if (priv->command_queue) {
@@ -2572,6 +2583,233 @@ fail:
 
 #if HAVE_OPENCL_D3D11
 
+#if CONFIG_LIBMFX
+
+static void opencl_unmap_from_d3d11_qsv(AVHWFramesContext *dst_fc,
+                                        HWMapDescriptor *hwmap)
+{
+    AVOpenCLFrameDescriptor    *desc = hwmap->priv;
+    OpenCLDeviceContext *device_priv = dst_fc->device_ctx->hwctx;
+    OpenCLFramesContext *frames_priv = dst_fc->hwctx;
+    cl_event event;
+    cl_int cle;
+    int p;
+
+    av_log(dst_fc, AV_LOG_DEBUG, "Unmap QSV surface from OpenCL.\n");
+
+    cle = device_priv->clEnqueueReleaseD3D11ObjectsKHR(
+        frames_priv->command_queue, desc->nb_planes, desc->planes,
+        0, NULL, &event);
+    if (cle != CL_SUCCESS) {
+        av_log(dst_fc, AV_LOG_ERROR, "Failed to release texture "
+               "handle: %d.\n", cle);
+    }
+
+    opencl_wait_events(dst_fc, &event, 1);
+
+    if (!frames_priv->nb_mapped_frames && !frames_priv->mapped_frames) {
+        for (p = 0; p < desc->nb_planes; p++) {
+            cle = clReleaseMemObject(desc->planes[p]);
+            if (cle != CL_SUCCESS) {
+                av_log(dst_fc, AV_LOG_ERROR, "Failed to release CL "
+                       "image of plane %d of D3D11 texture: %d\n",
+                       p, cle);
+            }
+        }
+        av_freep(&desc);
+    }
+}
+
+static int opencl_map_from_d3d11_qsv(AVHWFramesContext *dst_fc, AVFrame *dst,
+                                     const AVFrame *src, int flags)
+{
+    OpenCLDeviceContext  *device_priv = dst_fc->device_ctx->hwctx;
+    OpenCLFramesContext  *frames_priv = dst_fc->hwctx;
+    AVOpenCLDeviceContext    *dst_dev = &device_priv->p;
+    mfxFrameSurface1 *mfx_surface = (mfxFrameSurface1*)src->data[3];
+    mfxHDLPair *pair = (mfxHDLPair*)mfx_surface->Data.MemId;
+    ID3D11Texture2D *tex = (ID3D11Texture2D*)pair->first;
+    AVOpenCLFrameDescriptor *desc;
+    cl_mem_flags cl_flags;
+    cl_event event;
+    cl_int cle;
+    int err, p, index, derived_frames;
+
+    cl_flags = opencl_mem_flags_for_mapping(flags);
+    if (!cl_flags)
+        return AVERROR(EINVAL);
+
+    av_log(dst_fc, AV_LOG_DEBUG, "Map QSV surface %#llx to OpenCL.\n", (uintptr_t)pair);
+
+    index = (intptr_t)pair->second;
+    derived_frames = frames_priv->nb_mapped_frames > 0;
+    if (derived_frames) {
+        av_assert0(index >= 0 && index != MFX_INFINITE);
+        if (index >= frames_priv->nb_mapped_frames) {
+            av_log(dst_fc, AV_LOG_ERROR, "Texture array index out of range for "
+                   "mapping: %d >= %d.\n", index, frames_priv->nb_mapped_frames);
+            return AVERROR(EINVAL);
+        }
+    }
+
+    if (derived_frames) {
+        desc = &frames_priv->mapped_frames[index];
+    } else {
+        desc = av_mallocz(sizeof(*desc));
+        if (!desc)
+            return AVERROR(ENOMEM);
+
+        desc->nb_planes = 2;
+        for (p = 0; p < desc->nb_planes; p++) {
+            desc->planes[p] =
+                device_priv->clCreateFromD3D11Texture2DKHR(
+                    dst_dev->context, cl_flags, tex,
+                    p, &cle);
+            if (!desc->planes[p]) {
+                av_log(dst_fc, AV_LOG_ERROR, "Failed to create CL "
+                       "image from plane %d of D3D11 texture: %d.\n",
+                       p, cle);
+                err = AVERROR(EIO);
+                goto fail2;
+            }
+        }
+    }
+
+    cle = device_priv->clEnqueueAcquireD3D11ObjectsKHR(
+        frames_priv->command_queue, desc->nb_planes, desc->planes,
+        0, NULL, &event);
+    if (cle != CL_SUCCESS) {
+        av_log(dst_fc, AV_LOG_ERROR, "Failed to acquire texture "
+               "handle: %d.\n", cle);
+        err = AVERROR(EIO);
+        goto fail;
+    }
+
+    err = opencl_wait_events(dst_fc, &event, 1);
+    if (err < 0)
+        goto fail;
+
+    for (p = 0; p < desc->nb_planes; p++)
+        dst->data[p] = (uint8_t*)desc->planes[p];
+
+    err = ff_hwframe_map_create(dst->hw_frames_ctx, dst, src,
+                                &opencl_unmap_from_d3d11_qsv, desc);
+    if (err < 0)
+        goto fail;
+
+    dst->width  = src->width;
+    dst->height = src->height;
+
+    return 0;
+
+fail:
+    cle = device_priv->clEnqueueReleaseD3D11ObjectsKHR(
+        frames_priv->command_queue, desc->nb_planes, desc->planes,
+        0, NULL, &event);
+    if (cle == CL_SUCCESS)
+        opencl_wait_events(dst_fc, &event, 1);
+fail2:
+    if (!derived_frames) {
+        for (p = 0; p < desc->nb_planes; p++) {
+            if (desc->planes[p])
+                clReleaseMemObject(desc->planes[p]);
+        }
+        av_freep(&desc);
+    }
+    memset(dst->data, 0, sizeof(dst->data));
+    return err;
+}
+
+static int opencl_frames_derive_from_d3d11_qsv(AVHWFramesContext *dst_fc,
+                                               AVHWFramesContext *src_fc, int flags)
+{
+    AVQSVFramesContext     *src_hwctx = src_fc->hwctx;
+    OpenCLDeviceContext  *device_priv = dst_fc->device_ctx->hwctx;
+    AVOpenCLDeviceContext    *dst_dev = &device_priv->p;
+    OpenCLFramesContext  *frames_priv = dst_fc->hwctx;
+    cl_mem_flags cl_flags;
+    cl_int cle;
+    int err, i, p, nb_planes = 2;
+    mfxHDLPair *pair = NULL;
+    ID3D11Texture2D *tex = NULL;
+
+    if (src_fc->sw_format != AV_PIX_FMT_NV12 &&
+        src_fc->sw_format != AV_PIX_FMT_P010) {
+        av_log(dst_fc, AV_LOG_ERROR, "Only NV12 and P010 textures are "
+               "supported for QSV with D3D11 to OpenCL mapping.\n");
+        return AVERROR(EINVAL);
+    }
+
+    if (src_fc->initial_pool_size == 0) {
+        av_log(dst_fc, AV_LOG_DEBUG, "Non fixed-size pools input for QSV "
+               "with D3D11 to OpenCL mapping.\n");
+        return 0;
+    }
+
+    if ((src_hwctx->frame_type & MFX_MEMTYPE_VIDEO_MEMORY_PROCESSOR_TARGET) ||
+        (src_hwctx->frame_type & MFX_MEMTYPE_FROM_VPPOUT)) {
+        av_log(dst_fc, AV_LOG_DEBUG, "MFX memtype VPP input for QSV "
+               "with D3D11 to OpenCL mapping.\n");
+        return 0;
+    }
+
+    if (!src_hwctx->surfaces)
+        return AVERROR(ENOMEM);
+    pair = (mfxHDLPair*)src_hwctx->surfaces[0].Data.MemId;
+    if (!pair)
+        return AVERROR(ENOMEM);
+    tex = (ID3D11Texture2D*)pair->first;
+
+    cl_flags = opencl_mem_flags_for_mapping(flags);
+    if (!cl_flags)
+        return AVERROR(EINVAL);
+
+    frames_priv->nb_mapped_frames = src_fc->initial_pool_size;
+
+    frames_priv->mapped_frames =
+        av_calloc(frames_priv->nb_mapped_frames,
+                  sizeof(*frames_priv->mapped_frames));
+    if (!frames_priv->mapped_frames)
+        return AVERROR(ENOMEM);
+
+    for (i = 0; i < frames_priv->nb_mapped_frames; i++) {
+        AVOpenCLFrameDescriptor *desc = &frames_priv->mapped_frames[i];
+        desc->nb_planes = nb_planes;
+
+        for (p = 0; p < nb_planes; p++) {
+            UINT subresource = 2 * i + p;
+            desc->planes[p] =
+                device_priv->clCreateFromD3D11Texture2DKHR(
+                    dst_dev->context, cl_flags, tex,
+                    subresource, &cle);
+            if (!desc->planes[p]) {
+                av_log(dst_fc, AV_LOG_ERROR, "Failed to create CL "
+                       "image from plane %d of D3D11 texture "
+                       "index %d (subresource %u): %d.\n",
+                       p, i, (unsigned int)subresource, cle);
+                err = AVERROR(EIO);
+                goto fail;
+            }
+        }
+    }
+
+    return 0;
+
+fail:
+    for (i = 0; i < frames_priv->nb_mapped_frames; i++) {
+        AVOpenCLFrameDescriptor *desc = &frames_priv->mapped_frames[i];
+        for (p = 0; p < desc->nb_planes; p++) {
+            if (desc->planes[p])
+                clReleaseMemObject(desc->planes[p]);
+        }
+    }
+    av_freep(&frames_priv->mapped_frames);
+    frames_priv->nb_mapped_frames = 0;
+    return err;
+}
+
+#endif
+
 static void opencl_unmap_from_d3d11(AVHWFramesContext *dst_fc,
                                     HWMapDescriptor *hwmap)
 {
@@ -3096,6 +3334,11 @@ static int opencl_map_to(AVHWFramesConte
             return opencl_map_from_dxva2(hwfc, dst, src, flags);
 #endif
 #if HAVE_OPENCL_D3D11
+#if CONFIG_LIBMFX
+    case AV_PIX_FMT_QSV:
+        if (priv->d3d11_qsv_mapping_usable)
+            return opencl_map_from_d3d11_qsv(hwfc, dst, src, flags);
+#endif
     case AV_PIX_FMT_D3D11:
         if (priv->d3d11_mapping_usable)
             return opencl_map_from_d3d11(hwfc, dst, src, flags);
@@ -3150,6 +3393,18 @@ static int opencl_frames_derive_to(AVHWF
         break;
 #endif
 #if HAVE_OPENCL_D3D11
+#if CONFIG_LIBMFX
+    case AV_HWDEVICE_TYPE_QSV:
+        if (!priv->d3d11_qsv_mapping_usable)
+            return AVERROR(ENOSYS);
+        {
+            int err;
+            err = opencl_frames_derive_from_d3d11_qsv(dst_fc, src_fc, flags);
+            if (err < 0)
+                return err;
+        }
+        break;
+#endif
     case AV_HWDEVICE_TYPE_D3D11VA:
         if (!priv->d3d11_mapping_usable)
             return AVERROR(ENOSYS);
