Index: FFmpeg/libavutil/hwcontext_d3d11va.c
===================================================================
--- libavutil/hwcontext_d3d11va.c
+++ libavutil/hwcontext_d3d11va.c
@@ -638,6 +638,8 @@ static int d3d11va_device_create(AVHWDev
         adapter = atoi(device);
     } else {
         AVDictionaryEntry *e = av_dict_get(opts, "vendor_id", NULL, 0);
+        if (!e || !e->value)
+            e = av_dict_get(opts, "vendor", NULL, 0); // for backward compatibility
         if (e && e->value) {
             adapter = d3d11va_device_find_adapter_by_vendor_id(ctx, creationFlags, e->value);
             if (adapter < 0) {
