Index: FFmpeg/libavformat/id3v2.c
===================================================================
--- libavformat/id3v2.c
+++ libavformat/id3v2.c
@@ -136,6 +136,7 @@ const CodecMime ff_id3v2_mime_tags[] = {
     { "image/png",  AV_CODEC_ID_PNG   },
     { "image/tiff", AV_CODEC_ID_TIFF  },
     { "image/bmp",  AV_CODEC_ID_BMP   },
+    { "image/webp", AV_CODEC_ID_WEBP  },
     { "JPG",        AV_CODEC_ID_MJPEG }, /* ID3v2.2  */
     { "PNG",        AV_CODEC_ID_PNG   }, /* ID3v2.2  */
     { "",           AV_CODEC_ID_NONE  },
@@ -325,39 +326,54 @@ static void read_ttag(AVFormatContext *s
                       AVDictionary **metadata, const char *key)
 {
     uint8_t *dst;
-    int encoding, dict_flags = AV_DICT_DONT_OVERWRITE | AV_DICT_DONT_STRDUP_VAL;
+    int encoding, nb_values = 0;
     unsigned genre;
+    AVDictionaryEntry *tag = NULL;
 
     if (taglen < 1)
         return;
 
+    tag = av_dict_get(*metadata, key, NULL, 0);
+    if (tag)
+        return;
+
     encoding = avio_r8(pb);
     taglen--; /* account for encoding type byte */
 
-    if (decode_str(s, pb, encoding, &dst, &taglen) < 0) {
-        av_log(s, AV_LOG_ERROR, "Error reading frame %s, skipped\n", key);
-        return;
-    }
+    /* loop through all available tags */
+    while (taglen) {
+        int dict_flags = AV_DICT_APPEND | AV_DICT_DONT_STRDUP_VAL;
 
-    if (!(strcmp(key, "TCON") && strcmp(key, "TCO"))                         &&
-        (sscanf(dst, "(%d)", &genre) == 1 || sscanf(dst, "%d", &genre) == 1) &&
-        genre <= ID3v1_GENRE_MAX) {
-        av_freep(&dst);
-        dst = av_strdup(ff_id3v1_genre_str[genre]);
-    } else if (!(strcmp(key, "TXXX") && strcmp(key, "TXX"))) {
-        /* dst now contains the key, need to get value */
-        key = dst;
         if (decode_str(s, pb, encoding, &dst, &taglen) < 0) {
             av_log(s, AV_LOG_ERROR, "Error reading frame %s, skipped\n", key);
-            av_freep(&key);
             return;
         }
-        dict_flags |= AV_DICT_DONT_STRDUP_KEY;
-    } else if (!*dst)
-        av_freep(&dst);
 
-    if (dst)
-        av_dict_set(metadata, key, dst, dict_flags);
+        if (!(strcmp(key, "TCON") && strcmp(key, "TCO"))                         &&
+            (sscanf(dst, "(%d)", &genre) == 1 || sscanf(dst, "%d", &genre) == 1) &&
+            genre <= ID3v1_GENRE_MAX) {
+            av_freep(&dst);
+            dst = av_strdup(ff_id3v1_genre_str[genre]);
+        } else if (!(strcmp(key, "TXXX") && strcmp(key, "TXX"))) {
+            /* dst now contains the key, need to get value */
+            key = dst;
+            if (decode_str(s, pb, encoding, &dst, &taglen) < 0) {
+                av_log(s, AV_LOG_ERROR, "Error reading frame %s, skipped\n", key);
+                av_freep(&key);
+                return;
+            }
+            dict_flags |= AV_DICT_DONT_STRDUP_KEY;
+        } else if (!*dst) {
+            av_freep(&dst);
+            return;
+        }
+
+        if (dst) {
+            if (nb_values++)
+                av_dict_set(metadata, key, ";", dict_flags & ~AV_DICT_DONT_STRDUP_VAL);
+            av_dict_set(metadata, key, dst, dict_flags);
+        }
+    }
 }
 
 static void read_uslt(AVFormatContext *s, AVIOContext *pb, int taglen,
