Index: FFmpeg/libavcodec/qsvenc_av1.c
===================================================================
--- libavcodec/qsvenc_av1.c
+++ libavcodec/qsvenc_av1.c
@@ -203,8 +203,8 @@ static const AVClass class = {
 
 static const FFCodecDefault qsv_enc_defaults[] = {
     { "b",         "0"    },
-    { "g",         "-1"   },
-    { "bf",        "-1"   },
+    { "g",         "250"  },
+    { "bf",        "3"    },
     { "refs",      "0"    },
     { NULL },
 };
Index: FFmpeg/libavcodec/qsvenc_h264.c
===================================================================
--- libavcodec/qsvenc_h264.c
+++ libavcodec/qsvenc_h264.c
@@ -180,8 +180,8 @@ static const AVClass class = {
 static const FFCodecDefault qsv_enc_defaults[] = {
     { "b",         "0"     },
     { "refs",      "0"     },
-    { "g",         "-1"    },
-    { "bf",        "-1"    },
+    { "g",         "250"   },
+    { "bf",        "3"     },
     { "qmin",      "-1"    },
     { "qmax",      "-1"    },
     { "trellis",   "-1"    },
Index: FFmpeg/libavcodec/qsvenc_hevc.c
===================================================================
--- libavcodec/qsvenc_hevc.c
+++ libavcodec/qsvenc_hevc.c
@@ -379,8 +379,8 @@ static const AVClass class = {
 static const FFCodecDefault qsv_enc_defaults[] = {
     { "b",         "0"     },
     { "refs",      "0"     },
-    { "g",         "248"   },
-    { "bf",        "-1"    },
+    { "g",         "250"   },
+    { "bf",        "4"     },
     { "qmin",      "-1"    },
     { "qmax",      "-1"    },
     { "trellis",   "-1"    },
