Index: FFmpeg/libavcodec/vaapi_encode_h265.c
===================================================================
--- libavcodec/vaapi_encode_h265.c
+++ libavcodec/vaapi_encode_h265.c
@@ -264,6 +264,7 @@ static int vaapi_encode_h265_init_sequen
     H265RawVPS                        *vps = &priv->units.raw_vps;
     H265RawSPS                        *sps = &priv->units.raw_sps;
     H265RawPPS                        *pps = &priv->units.raw_pps;
+    H265RawVUI                        *vui = &sps->vui;
     VAEncSequenceParameterBufferHEVC *vseq = ctx->codec_sequence_params;
     VAEncPictureParameterBufferHEVC  *vpic = ctx->codec_picture_params;
     int i, err;
@@ -391,7 +392,25 @@ static int vaapi_encode_h265_init_sequen
             sps->log2_min_pcm_luma_coding_block_size_minus3 +
             sps->log2_diff_max_min_pcm_luma_coding_block_size,
 
-        .vui_parameters_present_flag = 0,
+        .vui_parameters_present_flag = sps->vui_parameters_present_flag,
+        .vui_fields.bits = {
+            .aspect_ratio_info_present_flag = vui->aspect_ratio_info_present_flag,
+            .vui_timing_info_present_flag = vui->vui_timing_info_present_flag,
+            .bitstream_restriction_flag = vui->bitstream_restriction_flag,
+            .motion_vectors_over_pic_boundaries_flag =
+                vui->motion_vectors_over_pic_boundaries_flag,
+            .restricted_ref_pic_lists_flag = vui->restricted_ref_pic_lists_flag,
+            .log2_max_mv_length_horizontal = vui->log2_max_mv_length_horizontal,
+            .log2_max_mv_length_vertical = vui->log2_max_mv_length_vertical,
+        },
+
+        .aspect_ratio_idc = vui->aspect_ratio_idc,
+        .sar_width = vui->sar_width,
+        .sar_height = vui->sar_height,
+        .vui_num_units_in_tick = vui->vui_num_units_in_tick,
+        .vui_time_scale = vui->vui_time_scale,
+        .max_bytes_per_pic_denom = vui->max_bytes_per_pic_denom,
+        .max_bits_per_min_cu_denom = vui->max_bits_per_min_cu_denom,
     };
 
     *vpic = (VAEncPictureParameterBufferHEVC) {
