Index: FFmpeg/libavcodec/libopusenc.c
===================================================================
--- libavcodec/libopusenc.c
+++ libavcodec/libopusenc.c
@@ -196,6 +196,10 @@ static int libopus_check_vorbis_layout(A
         av_log(avctx, AV_LOG_WARNING,
                "No channel layout specified. Opus encoder will use Vorbis "
                "channel layout for %d channels.\n", avctx->ch_layout.nb_channels);
+    } else if (av_channel_layout_compare(&avctx->ch_layout, &(AVChannelLayout)AV_CHANNEL_LAYOUT_5POINT1) == 0) {
+        av_log(avctx, AV_LOG_WARNING,
+               "Input channel layout 5.1(side) detected, side channels will be mapped to back channels.\n");
+        return 0;
     } else if (av_channel_layout_compare(&avctx->ch_layout, &ff_vorbis_ch_layouts[avctx->ch_layout.nb_channels - 1])) {
         char name[32];
 
