Index: FFmpeg/fftools/ffmpeg_filter.c
===================================================================
--- fftools/ffmpeg_filter.c
+++ fftools/ffmpeg_filter.c
@@ -2458,7 +2458,8 @@ static void sub2video_heartbeat(InputFil
            or if we need to initialize the system, update the
            overlayed subpicture and its start/end times */
         sub2video_update(ifp, pts2 + 1, NULL);
-    else
+
+    if (av_buffersrc_get_nb_failed_requests(ifp->filter))
         sub2video_push_ref(ifp, pts2);
 }
 
Index: FFmpeg/tests/ref/fate/filter-overlay-dvdsub-2397
===================================================================
--- tests/ref/fate/filter-overlay-dvdsub-2397
+++ tests/ref/fate/filter-overlay-dvdsub-2397
@@ -489,12 +489,12 @@
 1,       3877,       3877,       10,     2013, 0x95a39f9c
 1,       3887,       3887,       10,     2013, 0x4f7ea123
 1,       3897,       3897,       10,     2013, 0x9efb9ba1
-0,        117,        117,        1,   518400, 0x61e0f688
+0,        117,        117,        1,   518400, 0xbf8523da
 1,       3907,       3907,       10,     2013, 0xf395b2cd
 1,       3917,       3917,       10,     2013, 0x261a881e
 1,       3927,       3927,       10,     2013, 0x7f2d9f72
 1,       3937,       3937,       10,     2013, 0x0105b38d
-0,        118,        118,        1,   518400, 0xa47de755
+0,        118,        118,        1,   518400, 0x41890ed6
 1,       3952,       3952,       10,     2013, 0x0e5db67e
 1,       3962,       3962,       10,     2013, 0xfc9baf97
 0,        119,        119,        1,   518400, 0x588534fc
Index: FFmpeg/tests/ref/fate/sub2video
===================================================================
--- tests/ref/fate/sub2video
+++ tests/ref/fate/sub2video
@@ -68,8 +68,7 @@
 0,        258,        258,        1,   518400, 0x34cdddee
 0,        269,        269,        1,   518400, 0xbab197ea
 1,   53910000,   53910000,  2696000,     2095, 0x61bb15ed
-0,        270,        270,        1,   518400, 0xbab197ea
-0,        271,        271,        1,   518400, 0x4db4ce51
+0,        270,        270,        1,   518400, 0x4db4ce51
 0,        283,        283,        1,   518400, 0xbab197ea
 1,   56663000,   56663000,  1262000,     1013, 0xc9ae89b7
 0,        284,        284,        1,   518400, 0xe6bc0ea9
