Index: FFmpeg/libavformat/assenc.c
===================================================================
--- libavformat/assenc.c
+++ libavformat/assenc.c
@@ -68,7 +68,8 @@ static int write_header(AVFormatContext
                 ass->trailer = trailer;
         }
 
-        ffio_write_lines(s->pb, par->extradata, header_size, NULL);
+        header_size = av_strnlen(par->extradata, header_size);
+        ffio_write_lines(s->pb, par->extradata, (int)header_size, NULL);
 
         ass->ssa_mode = !strstr(par->extradata, "\n[V4+ Styles]");
         if (!strstr(par->extradata, "\n[Events]"))
