--- libavcodec/libsvtav1.c
+++ libavcodec/libsvtav1.c
@@ -238,7 +238,11 @@ static int config_enc_params(EbSvtAv1Enc
     } else if (svt_enc->qp > 0) {
         param->qp                   = svt_enc->qp;
         param->rate_control_mode    = 0;
+#if SVT_AV1_CHECK_VERSION(4, 0, 0)
+        param->aq_mode = 0;
+#else
         param->enable_adaptive_quantization = 0;
+#endif
     }
 
     desc = av_pix_fmt_desc_get(avctx->pix_fmt);
