--- CMakeLists.txt.orig	2015-03-25 08:59:06.695422175 +0100
+++ CMakeLists.txt	2015-03-25 08:59:32.740438770 +0100
@@ -405,7 +405,7 @@
   message(STATUS "Not building on Raspberry Pi. GPIO support disabled.")
 ENDIF(RaspberryPi)
 
-target_link_libraries(domoticz ${Boost_LIBRARIES} ${SQLITE3_LIBRARY} ${ZLIB_LIBRARIES} ${CURL_LIBRARIES} pthread lua mqtt ${CMAKE_DL_LIBS})
+target_link_libraries(domoticz ${Boost_LIBRARIES} ${SQLITE3_LIBRARY} ${ZLIB_LIBRARIES} ${CURL_LIBRARIES} pthread lua mqtt ${CMAKE_DL_LIBS} -lrt)
 
 ADD_PRECOMPILED_HEADER(domoticz "main/stdafx.h")
 
