diff -uprN ../tvheadend-git221c29b40b1e53ae09a69d9458442dd4fea665f5.ORIG/Makefile ./Makefile
--- ../tvheadend-git221c29b40b1e53ae09a69d9458442dd4fea665f5.ORIG/Makefile	2019-11-28 13:11:58.000000000 -0500
+++ ./Makefile	2020-03-02 20:33:50.811069069 -0500
@@ -36,7 +36,7 @@ CFLAGS  += -g
 ifeq ($(CONFIG_CCDEBUG),yes)
 CFLAGS  += -O0
 else
-CFLAGS  += -O2 -D_FORTIFY_SOURCE=2
+CFLAGS  += -O2
 endif
 ifeq ($(CONFIG_PIE),yes)
 CFLAGS  += -fPIE
@@ -52,9 +52,6 @@ endif
 CFLAGS  += -Wall -Wwrite-strings -Wno-deprecated-declarations
 CFLAGS  += -Wmissing-prototypes
 CFLAGS  += -fms-extensions -funsigned-char -fno-strict-aliasing
-ifeq ($(COMPILER), gcc)
-CFLAGS  += -Wno-stringop-truncation -Wno-stringop-overflow
-endif
 CFLAGS  += -D_FILE_OFFSET_BITS=64
 CFLAGS  += -I${BUILDDIR} -I${ROOTDIR}/src -I${ROOTDIR}
 ifeq ($(CONFIG_ANDROID),yes)
