diff -uprN ../linuxtv-gitea2766f182b3a4e03543be2ded0845fca4d4fa80-ORIG/v4l/Makefile ./v4l/Makefile
--- ../linuxtv-gitea2766f182b3a4e03543be2ded0845fca4d4fa80-ORIG/v4l/Makefile	2020-07-20 04:26:22.000000000 -0400
+++ ./v4l/Makefile	2020-07-28 20:26:51.043908729 -0400
@@ -51,13 +51,13 @@ default:: prepare
 	@echo Kernel build directory is $(OUTDIR)
 	$(MAKE) -C ../linux apply_patches
 	$(MAKE) -C $(OUTDIR) M=$(PWD) $(MYCFLAGS) modules
-	./scripts/rmmod.pl check
+#	./scripts/rmmod.pl check
 #	$(MAKE) checkpatch
 
 mismatch:: prepare
 	@echo Kernel build directory is $(OUTDIR)
 	$(MAKE) -Wfatal-errors -C $(OUTDIR) M=$(PWD) $(MYCFLAGS) CONFIG_DEBUG_SECTION_MISMATCH=y modules
-	./scripts/rmmod.pl check
+#	./scripts/rmmod.pl check
 
 # Anything in this target shouldn't be build in parallel.
 prepare:: config-compat.h Makefile.media links
