# DP: Untested patch suggested to fix #983878

--- a/src/java.base/unix/native/libnio/fs/UnixNativeDispatcher.c
+++ b/src/java.base/unix/native/libnio/fs/UnixNativeDispatcher.c
@@ -156,7 +156,7 @@ static fdopendir_func* my_fdopendir_func
 /**
  * fstatat missing from glibc on Linux.
  */
-#if defined(__linux__) && (defined(__i386) || defined(__arm__))
+#if defined(__linux__) && (defined(__i386) || defined(__arm__) || (defined(__mips__) && !defined(__mips64)))
 #define FSTATAT64_SYSCALL_AVAILABLE
 static int fstatat64_wrapper(int dfd, const char *path,
                              struct stat64 *statbuf, int flag)
