--- module/ConfigParser.py	2014-05-24 01:55:31.000000000 +0200
+++ module/ConfigParser.py	2014-05-24 01:54:57.000000000 +0200
@@ -140,7 +140,7 @@
             try:
                 if line == "":
                     continue
-                elif line.endswith(":"):
+                elif line.endswith(":") and not line.endswith("= ::"):
                     section, none, desc = line[:-1].partition('-')
                     section = section.strip()
                     desc = desc.replace('"', "").strip()
