--- module/network/HTTPChunk.py	2014-05-24 02:07:40.000000000 +0200
+++ module/network/HTTPChunk.py	2014-05-24 02:06:47.000000000 +0200
@@ -258,7 +258,7 @@
 
             if line.startswith("content-disposition") and "filename=" in line:
                 name = orgline.partition("filename=")[2]
-                name = name.replace('"', "").replace("'", "").replace(";", "").strip()
+                name = name.replace('"', "").replace("'", "").replace(";", "").replace("/", "_").strip()
                 self.p.nameDisposition = name
                 self.log.debug("Content-Disposition: %s" % name)
 
