--- module/plugins/hooks/UpdateManager.py.orig	2012-10-13 10:31:30.000000000 +0200
+++ module/plugins/hooks/UpdateManager.py	2015-12-29 12:44:20.653525885 +0100
@@ -1,44 +1,28 @@
 # -*- coding: utf-8 -*-
 
-"""
-    This program is free software; you can redistribute it and/or modify
-    it under the terms of the GNU General Public License as published by
-    the Free Software Foundation; either version 3 of the License,
-    or (at your option) any later version.
-
-    This program is distributed in the hope that it will be useful,
-    but WITHOUT ANY WARRANTY; without even the implied warranty of
-    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
-    See the GNU General Public License for more details.
-
-    You should have received a copy of the GNU General Public License
-    along with this program; if not, see <http://www.gnu.org/licenses/>.
-
-    @author: RaNaN
-"""
-
 import sys
 import re
-from os import stat
-from os.path import join, exists
+from os import remove, stat
+from os.path import join, isfile
 from time import time
 
 from module.ConfigParser import IGNORE
 from module.network.RequestFactory import getURL
 from module.plugins.Hook import threaded, Expose, Hook
 
+
 class UpdateManager(Hook):
     __name__ = "UpdateManager"
-    __version__ = "0.13"
-    __description__ = """checks for updates"""
-    __config__ = [("activated", "bool", "Activated", "True"),
-        ("interval", "int", "Check interval in minutes", "480"),
-        ("debug", "bool", "Check for plugin changes when in debug mode", False)]
-    __author_name__ = ("RaNaN")
-    __author_mail__ = ("ranan@pyload.org")
+    __version__ = "0.16"
+    __description__ = """Checks for updates"""
+    __config__ = [("activated", "bool", "Activated", True),
+                  ("interval", "int", "Check interval in minutes", 480),
+                  ("debug", "bool", "Check for plugin changes when in debug mode", False)]
+    __author_name__ = ("RaNaN", "stickell")
+    __author_mail__ = ("ranan@pyload.org", "l.stickell@yahoo.it")
 
-    URL = "http://get.pyload.org/check2/%s/"
-    MIN_TIME = 3 * 60 * 60 # 3h minimum check interval
+    URL = "http://updatemanager-spyload.rhcloud.com"
+    MIN_TIME = 3 * 60 * 60  # 3h minimum check interval
 
     @property
     def debug(self):
@@ -48,10 +32,10 @@
         if self.debug:
             self.logDebug("Monitoring file changes")
             self.interval = 4
-            self.last_check = 0 #timestamp of updatecheck
+            self.last_check = 0  # timestamp of updatecheck
             self.old_periodical = self.periodical
             self.periodical = self.checkChanges
-            self.mtimes = {}  #recordes times
+            self.mtimes = {}  # recordes times
         else:
             self.interval = max(self.getConfig("interval") * 60, self.MIN_TIME)
 
@@ -63,19 +47,18 @@
 
     @threaded
     def periodical(self):
-
         updates = self.checkForUpdate()
         if updates:
             self.checkPlugins(updates)
 
         if self.updated and not self.reloaded:
             self.info["plugins"] = True
-            self.log.info(_("*** Plugins have been updated, please restart pyLoad ***"))
+            self.logInfo(_("*** Plugins have been updated, please restart pyLoad ***"))
         elif self.updated and self.reloaded:
-            self.log.info(_("Plugins updated and reloaded"))
+            self.logInfo(_("Plugins updated and reloaded"))
             self.updated = False
         elif self.version == "None":
-            self.log.info(_("No plugin updates available"))
+            self.logInfo(_("No plugin updates available"))
 
     @Expose
     def recheckForUpdates(self):
@@ -84,40 +67,42 @@
 
     def checkForUpdate(self):
         """checks if an update is available, return result"""
-
         try:
-            if self.version == "None": # No updated known
-                version_check = getURL(self.URL % self.core.api.getServerVersion()).splitlines()
+            if self.version == "None":  # No updated known
+                version_check = getURL(self.URL, get={'v': self.core.api.getServerVersion()}).splitlines()
                 self.version = version_check[0]
 
                 # Still no updates, plugins will be checked
                 if self.version == "None":
-                    self.log.info(_("No Updates for pyLoad"))
+                    self.logInfo(_("No Updates for pyLoad"))
                     return version_check[1:]
 
-
             self.info["pyload"] = True
-            self.log.info(_("***  New pyLoad Version %s available  ***") % self.version)
-            self.log.info(_("***  Get it here: http://pyload.org/download  ***"))
-
+            self.logInfo(_("***  New pyLoad Version %s available  ***") % self.version)
+            self.logInfo(_("***  Get it here: http://pyload.org/download  ***"))
         except:
-            self.log.warning(_("Not able to connect server for updates"))
-
-        return None # Nothing will be done
+            self.logWarning(_("Not able to connect server for updates"))
 
+        return None  # Nothing will be done
 
     def checkPlugins(self, updates):
         """ checks for plugins updates"""
 
         # plugins were already updated
-        if self.info["plugins"]: return
+        if self.info["plugins"]:
+            return
 
         reloads = []
 
         vre = re.compile(r'__version__.*=.*("|\')([0-9.]+)')
         url = updates[0]
         schema = updates[1].split("|")
-        updates = updates[2:]
+        if 'BLACKLIST' in updates:
+            blacklist = updates[updates.index('BLACKLIST') + 1:]
+            updates = updates[2:updates.index('BLACKLIST')]
+        else:
+            blacklist = None
+            updates = updates[2:]
 
         for plugin in updates:
             info = dict(zip(schema, plugin.split("|")))
@@ -145,7 +130,7 @@
             if name in IGNORE or (type, name) in IGNORE:
                 continue
 
-            self.log.info(_("New version of %(type)s|%(name)s : %(version).2f") % {
+            self.logInfo(_("New version of %(type)s|%(name)s : %(version).2f") % {
                 "type": type,
                 "name": name,
                 "version": float(version)
@@ -169,17 +154,30 @@
 
             reloads.append((prefix, name))
 
+        self.executeBlacklist(blacklist)
+
         self.reloaded = self.core.pluginManager.reloadPlugins(reloads)
 
-    def checkChanges(self):
+    def executeBlacklist(self, blacklist):
+        for b in blacklist:
+            type, name = b.split('|')
+            if isfile(join("userplugins", type, name)):
+                self.logInfo(_("Removing blacklisted plugin %(type)s|%(name)s") % {
+                    "type": type,
+                    "name": name
+                })
+                remove(join("userplugins", type, name))
+            if isfile(join("userplugins", type, name.replace('.py', '.pyc'))):
+                remove(join("userplugins", type, name.replace('.py', '.pyc')))
 
+    def checkChanges(self):
         if self.last_check + max(self.getConfig("interval") * 60, self.MIN_TIME) < time():
             self.old_periodical()
             self.last_check = time()
 
         modules = filter(
-            lambda m: m and (m.__name__.startswith("module.plugins.") or m.__name__.startswith("userplugins.")) and m.__name__.count(".") >= 2,
-            sys.modules.itervalues())
+            lambda m: m and (m.__name__.startswith("module.plugins.") or m.__name__.startswith(
+                "userplugins.")) and m.__name__.count(".") >= 2, sys.modules.itervalues())
 
         reloads = []
 
@@ -188,7 +186,8 @@
             id = (type, name)
             if type in self.core.pluginManager.plugins:
                 f = m.__file__.replace(".pyc", ".py")
-                if not exists(f): continue
+                if not isfile(f):
+                    continue
 
                 mtime = stat(f).st_mtime
 
