Index: jellyfin-ffmpeg/configure
===================================================================
--- configure
+++ configure
@@ -1730,7 +1730,6 @@ EXTERNAL_LIBRARY_GPL_LIST="
 
 EXTERNAL_LIBRARY_NONFREE_LIST="
     decklink
-    libfdk_aac
     openssl
     libtls
 "
@@ -1771,6 +1770,7 @@ EXTERNAL_LIBRARY_LIST="
     libdav1d
     libdc1394
     libdrm
+    libfdk_aac
     libflite
     libfontconfig
     libfreetype
