--- setup.py.orig	2016-07-22 14:43:17.000000000 +0200
+++ setup.py	2016-12-03 11:22:50.358848097 +0100
@@ -71,7 +71,7 @@
             self.openssl = 'c:\\pkg'
         else:
             self.libraries = ['ssl', 'crypto']
-            self.openssl = '/usr'
+            self.openssl = os.environ.get('STAGING_INSTALL_PREFIX')
 
     def finalize_options(self):
         '''Overloaded build_ext implementation to append custom openssl
