--- common/cpu.cpp.orig	2019-01-11 13:49:59.314684910 +0100
+++ common/cpu.cpp	2019-01-11 13:51:37.011275586 +0100
@@ -110,6 +110,11 @@
     { "", 0 },
 };
 
+bool detect512()
+{
+    return(enable512);
+}
+
 #if X265_ARCH_X86
 
 extern "C" {
@@ -123,10 +128,6 @@
 #pragma warning(disable: 4309) // truncation of constant value
 #endif
 
-bool detect512()
-{
-    return(enable512);
-}
 uint32_t cpu_detect(bool benableavx512 )
 {
 
