--- source/common/cpu.cpp.orig	2024-11-22 13:07:34.000000000 +0100
+++ source/common/cpu.cpp	2025-09-07 16:26:25.649990913 +0200
@@ -128,6 +128,11 @@
     { "", 0 },
 };
 
+bool detect512()
+{
+    return(enable512);
+}
+
 #if X265_ARCH_X86
 
 extern "C" {
@@ -141,11 +146,6 @@
 #pragma warning(disable: 4309) // truncation of constant value
 #endif
 
-bool detect512()
-{
-    return(enable512);
-}
-
 uint32_t cpu_detect(bool benableavx512 )
 {
 
