--- source/common/cpu.cpp.orig	2024-04-04 09:39:50.000000000 +0000
+++ source/common/cpu.cpp	2024-06-15 07:20:37.644785266 +0000
@@ -122,6 +122,11 @@
     { "", 0 },
 };
 
+bool detect512()
+{
+    return(enable512);
+}
+
 #if X265_ARCH_X86
 
 extern "C" {
@@ -135,11 +140,6 @@
 #pragma warning(disable: 4309) // truncation of constant value
 #endif
 
-bool detect512()
-{
-    return(enable512);
-}
-
 uint32_t cpu_detect(bool benableavx512 )
 {
 
