--- Cargo.toml.orig	2024-10-08 06:56:26.000000000 +0000
+++ Cargo.toml	2025-07-27 10:42:16.544994628 +0000
@@ -48,6 +48,15 @@
 unicode-width = "0.2.0"
 strum = { version = "0.26.3", features = ["derive"] }
 
+# Dependencies pinned for Rust 1.77.2
+backtrace = "=0.3.74"
+icu_normalizer_data = "=1.5.0"
+icu_properties_data = "=1.5.1"
+idna_adapter = "=1.2.0"
+litemap = "=0.7.4"
+zerofrom = "=0.1.5"
+
+
 [target.'cfg(any(target_os = "android", target_os = "linux"))'.dependencies]
 procfs = "0.17.0"
 
