--- daemon/connect/WebDownloader.cpp.ori	2014-07-03 22:48:10.000000000 +0200
+++ daemon/connect/WebDownloader.cpp	2014-07-23 15:00:31.282548900 +0200
@@ -367,7 +367,7 @@ WebDownloader::EStatus WebDownloader::Do
 		{
 			if (m_iContentLen == -1 && !m_bGZip)
 			{
-				warn("URL %s: Content-Length is not submitted by server, cannot verify whether the file is complete", m_szInfoName);
+				debug("URL %s: Content-Length is not submitted by server, cannot verify whether the file is complete", m_szInfoName);
 			}
 			break;
 		}
