diff -uprN ../ffmpeg-4.2.1-patch1-5-Update3/libavcodec/h264_parser.c ./libavcodec/h264_parser.c
--- ../ffmpeg-4.2.1-patch1-5-Update3/libavcodec/h264_parser.c	2019-07-08 13:45:25.000000000 -0400
+++ ./libavcodec/h264_parser.c	2020-01-17 07:33:31.000000000 -0500
@@ -47,6 +47,8 @@
 #include "mpegutils.h"
 #include "parser.h"
 
+#include "synoconfig.h"
+
 typedef struct H264ParseContext {
     ParseContext pc;
     H264ParamSets ps;
@@ -108,8 +110,13 @@ static int h264_find_frame_end(H264Parse
                 state >>= 1;           // 2->1, 1->0, 0->0
         } else if (state <= 5) {
             int nalu_type = buf[i] & 0x1F;
+#ifdef SYNO_FIX_MP4_REMUX_H264PARSER
+            if (nalu_type == H264_NAL_SEI || nalu_type == H264_NAL_SPS ||
+                nalu_type == H264_NAL_AUD) {
+#else
             if (nalu_type == H264_NAL_SEI || nalu_type == H264_NAL_SPS ||
                 nalu_type == H264_NAL_PPS || nalu_type == H264_NAL_AUD) {
+#endif
                 if (pc->frame_start_found) {
                     i++;
                     goto found;
diff -uprN ../ffmpeg-4.2.1-patch1-5-Update3/synoconfig.h ./synoconfig.h
--- ../ffmpeg-4.2.1-patch1-5-Update3/synoconfig.h	2020-01-17 09:15:23.724603559 -0500
+++ ./synoconfig.h	2020-01-17 09:20:03.490853172 -0500
@@ -53,3 +53,11 @@
 #if defined(SYNO_VIDEOSTATION)
 #define SYNO_FILTER_SRT_TAGS
 #endif
+ 
+/* Fixed h264_parser error in special MP4 remux
+ * some MP4 file could parser error with remain PPS information
+ * See Video Station #944
+ */
+#if defined(SYNO_VIDEOSTATION)
+#define SYNO_FIX_MP4_REMUX_H264PARSER
+#endif
