--- src/Makefile.in.orig	2012-04-20 09:00:08.000000000 +0200
+++ src/Makefile.in	2013-08-12 15:08:04.000000000 +0200
@@ -307,6 +307,7 @@
 	control.h \
 	globals.cc \
 	globals.h \
+	linux-atomic.c \
 	option_parser.cc \
 	option_parser.h \
 	signal_handler.cc \
@@ -324,6 +325,7 @@
 	input/libsub_input.a \
 	rpc/libsub_rpc.a \
 	utils/libsub_utils.a \
+	.libs/liblinux-atomic.a \
 	@PTHREAD_LIBS@
 
 rtorrent_SOURCES = \
@@ -453,6 +455,9 @@
 @AMDEP_TRUE@@am__fastdepCXX_FALSE@	DEPDIR=$(DEPDIR) $(CXXDEPMODE) $(depcomp) @AMDEPBACKSLASH@
 @am__fastdepCXX_FALSE@	$(CXXCOMPILE) -c -o $@ $<
 
+.c.o:
+@am__fastdepCXX_TRUE@	$(LTCOMPILE) -c -o $@ $<
+
 .cc.obj:
 @am__fastdepCXX_TRUE@	$(CXXCOMPILE) -MT $@ -MD -MP -MF $(DEPDIR)/$*.Tpo -c -o $@ `$(CYGPATH_W) '$<'`
 @am__fastdepCXX_TRUE@	$(am__mv) $(DEPDIR)/$*.Tpo $(DEPDIR)/$*.Po
--- src/Makefile.am.orig	2012-03-13 12:10:49.000000000 +0100
+++ src/Makefile.am	2013-08-12 09:20:03.000000000 +0200
@@ -28,6 +28,7 @@
 	control.h \
 	globals.cc \
 	globals.h \
+	linux-atomic.c \
 	option_parser.cc \
 	option_parser.h \
 	signal_handler.cc \
