--- src/command_ui.cc.orig
+++ src/command_ui.cc
@@ -862,13 +862,13 @@
   CMD2_ANY_VALUE("convert.xb",           std::bind(&apply_to_xb, std::placeholders::_2));
   CMD2_ANY_VALUE("convert.throttle",     std::bind(&apply_to_throttle, std::placeholders::_2));
 
-  CMD2_ANY_LIST("math.add",              [](auto, const auto& args) { return apply_math_basic("math.add", std::plus(), args); });
-  CMD2_ANY_LIST("math.sub",              [](auto, const auto& args) { return apply_math_basic("math.sub", std::minus(), args); });
-  CMD2_ANY_LIST("math.mul",              [](auto, const auto& args) { return apply_math_basic("math.mul", std::multiplies(), args); });
-  CMD2_ANY_LIST("math.div",              [](auto, const auto& args) { return apply_math_basic("math.div", std::divides(), args); });
-  CMD2_ANY_LIST("math.mod",              [](auto, const auto& args) { return apply_math_basic("math.mod", std::modulus(), args); });
-  CMD2_ANY_LIST("math.min",              [](auto, const auto& args) { return apply_arith_basic(std::less(), args); });
-  CMD2_ANY_LIST("math.max",              [](auto, const auto& args) { return apply_arith_basic(std::greater(), args); });
+  CMD2_ANY_LIST("math.add",              [](auto, const auto& args) { return apply_math_basic("math.add", std::plus<>(), args); });
+  CMD2_ANY_LIST("math.sub",              [](auto, const auto& args) { return apply_math_basic("math.sub", std::minus<>(), args); });
+  CMD2_ANY_LIST("math.mul",              [](auto, const auto& args) { return apply_math_basic("math.mul", std::multiplies<>(), args); });
+  CMD2_ANY_LIST("math.div",              [](auto, const auto& args) { return apply_math_basic("math.div", std::divides<>(), args); });
+  CMD2_ANY_LIST("math.mod",              [](auto, const auto& args) { return apply_math_basic("math.mod", std::modulus<>(), args); });
+  CMD2_ANY_LIST("math.min",              [](auto, const auto& args) { return apply_arith_basic(std::less<>(), args); });
+  CMD2_ANY_LIST("math.max",              [](auto, const auto& args) { return apply_arith_basic(std::greater<>(), args); });
   CMD2_ANY_LIST("math.cnt",              std::bind(&apply_arith_count, std::placeholders::_2));
   CMD2_ANY_LIST("math.avg",              std::bind(&apply_arith_other, "average", std::placeholders::_2));
   CMD2_ANY_LIST("math.med",              std::bind(&apply_arith_other, "median", std::placeholders::_2));
