--- configure.orig	2012-04-20 09:00:08.000000000 +0200
+++ configure	2013-08-25 00:09:45.166052900 +0200
@@ -15672,7 +15672,7 @@
   { $as_echo "$as_me:${as_lineno-$LINENO}: checking for execinfo.h" >&5
 $as_echo_n "checking for execinfo.h... " >&6; }
 
-  if test "$cross_compiling" = yes; then :
+  if test "$cross_compiling" = maybe; then :
   { { $as_echo "$as_me:${as_lineno-$LINENO}: error: in \`$ac_pwd':" >&5
 $as_echo "$as_me: error: in \`$ac_pwd':" >&2;}
 as_fn_error $? "cannot run test program while cross compiling
@@ -17927,18 +17927,10 @@
   return 0;
 }
 _ACEOF
-if ac_fn_cxx_try_link "$LINENO"; then :
 
           { $as_echo "$as_me:${as_lineno-$LINENO}: result: ok" >&5
 $as_echo "ok" >&6; }
 
-else
-
-          { $as_echo "$as_me:${as_lineno-$LINENO}: result: failed" >&5
-$as_echo "failed" >&6; }
-          as_fn_error $? "Could not compile XMLRPC-C test." "$LINENO" 5
-
-fi
 rm -f core conftest.err conftest.$ac_objext \
     conftest$ac_exeext conftest.$ac_ext
 
