--- ../rtorrent-0.9.8-old/configure	2020-10-07 00:30:13.113972337 +0000
+++ configure	2020-10-07 00:34:03.499327179 +0000
@@ -19938,18 +19938,10 @@ main ()
   return 0;
 }
 _ACEOF
-if ac_fn_cxx_try_link "$LINENO"; then :
 
           { $as_echo "$as_me:${as_lineno-$LINENO}: result: ok" >&5
 $as_echo "ok" >&6; }
 
-else
-
-          { $as_echo "$as_me:${as_lineno-$LINENO}: result: failed" >&5
-$as_echo "failed" >&6; }
-          as_fn_error $? "Could not compile XMLRPC-C test." "$LINENO" 5
-
-fi
 rm -f core conftest.err conftest.$ac_objext \
     conftest$ac_exeext conftest.$ac_ext
 
