--- ./src/sg.h.in.old	2012-07-19 14:16:05.899697937 +0200
+++ ./src/sg.h.in	2012-07-19 14:17:19.935763168 +0200
@@ -87,17 +87,17 @@
 
 #ifdef ACCONFIG
 #undef DEFAULT_CONFIGFILE
-#define DEFAULT_CONFIGFILE "@sg_config@"
+#define DEFAULT_CONFIGFILE "@prefix@/etc/squidguard.conf"
 #endif
 
 #ifdef ACLOGDIR
 #undef DEFAULT_LOGDIR
-#define DEFAULT_LOGDIR "@sg_logdir@"
+#define DEFAULT_LOGDIR "@prefix@/var/logs"
 #endif
 
 #ifdef ACDBHOME
 #undef DEFAULT_DBHOME
-#define DEFAULT_DBHOME "@sg_dbhome@"
+#define DEFAULT_DBHOME "@prefix@/var/db"
 #endif
 
 #define INVALID_IP_ADDR 1
