Index: screen-4.0.3/pty.c
===================================================================
--- screen-4.0.3.orig/pty.c	2009-09-11 15:44:44.881250800 -0700
+++ screen-4.0.3/pty.c	2009-09-11 15:45:28.077862097 -0700
@@ -34,7 +34,7 @@
 #endif
 
 /* for solaris 2.1, Unixware (SVR4.2) and possibly others */
-#ifdef HAVE_SVR4_PTYS
+#if defined HAVE_SVR4_PTYS && !defined __UCLIBC__
 # include <sys/stropts.h>
 #endif
 
