Index: screen-4.0.3/configure.in
===================================================================
--- screen-4.0.3.orig/configure.in	2009-09-11 18:42:08.338231443 -0700
+++ screen-4.0.3/configure.in	2009-09-11 18:43:39.378663533 -0700
@@ -1243,12 +1243,6 @@
 #include <utmp.h>
 #endif
 ],[struct utmp u; u.ut_host[0] = 0;], AC_DEFINE(UTHOST))
-AH_TEMPLATE([HAVE_UTEMPTER])
-AC_CHECK_HEADER(utempter.h, have_utempter=yes, have_utempter=no)
-if test "$have_utempter" = yes; then
-  AC_DEFINE(HAVE_UTEMPTER)
-  LIBS="$LIBS -lutempter"
-fi
 
 dnl
 dnl    ****  loadav  ****
