diff -urN screen-4.0.3-orig/sched.h screen-4.0.3/sched.h
--- screen-4.0.3-orig/sched.h	2002-01-08 16:42:43.000000000 +0100
+++ screen-4.0.3/sched.h	2009-01-13 14:53:15.000000000 +0100
@@ -22,6 +22,10 @@
  * $Id: sched.h,v 1.1.1.1 1993/06/16 23:51:13 jnweiger Exp $ FAU
  */
 
+#ifndef _SCHED_H_
+#define _SCHED_H_ 1
+#include <sys/types.h>
+
 struct event
 {
   struct event *next;
@@ -41,3 +45,4 @@
 #define EV_READ		1
 #define EV_WRITE	2
 #define EV_ALWAYS	3
+#endif

