--- v4l/Makefile-orig	2022-04-05 06:22:11.000000000 +0000
+++ v4l/Makefile	2022-06-13 11:54:23.086968858 +0000
@@ -130,9 +130,6 @@ EXTRA_CFLAGS += $(if $(wildcard $(srctre
 # Make sure to use the backport header for common stuff
 EXTRA_CFLAGS += -include $(obj)/compat.h
 
-# gives too much warnings (false positives) -> disable it
-EXTRA_CFLAGS += -Wno-format-truncation
-
 #################################################
 # Kernel 2.6/3.x specific rules
 
