--- nDPI/example/ndpiReader.c	2014-09-06 13:27:11.000000000 +0200
+++ nDPI/example/ndpiReader.c	2014-11-10 12:24:19.835650831 +0100
@@ -1505,19 +1505,6 @@
 void *processing_thread(void *_thread_id) {
   long thread_id = (long) _thread_id;
 
-#ifdef linux
-  if(core_affinity[thread_id] >= 0) {
-    cpu_set_t cpuset;
-    CPU_ZERO(&cpuset);
-    CPU_SET(core_affinity[thread_id], &cpuset);
-
-    if(pthread_setaffinity_np(pthread_self(), sizeof(cpu_set_t), &cpuset) != 0)
-      fprintf(stderr, "Error while binding thread %ld to core %d\n", thread_id, core_affinity[thread_id]);
-    else {
-      if(!json_flag) printf("Running thread %ld on core %d...\n", thread_id, core_affinity[thread_id]);
-    }
-  } else
-#endif
     if(!json_flag) printf("Running thread %ld...\n", thread_id);
 
  pcap_loop:
