--- Makefile.in.orig	2014-09-06 13:27:07.000000000 +0200
+++ Makefile.in	2014-11-10 22:52:23.647259276 +0100
@@ -5,6 +5,9 @@
 GPP=@GPP@
 INSTALL_DIR=@INSTALL_DIR@
 MAN_DIR=@MAN_DIR@
+
+include ../tc_vars.mk
+
 ######
 HAS_NDPI=$(shell pkg-config --exists libndpi; echo $$?)
 ifeq ($(HAS_NDPI), 0)
@@ -130,7 +133,7 @@
 	$(GPP) $(OBJECTS) -Wall $(LIBS) -o $@
 
 $(NDPI_LIB): nDPI
-	cd nDPI; if test ! -f Makefile; then ./autogen.sh ; ./configure; fi; make
+	cd nDPI; if test ! -f Makefile; then ./autogen.sh ; env $(TC_ENV) ./configure $(TC_CONFIGURE_ARGS); fi; env $(TC_ENV) make
 
 nDPI:
 	svn co https://svn.ntop.org/svn/ntop/trunk/nDPI/
@@ -235,11 +238,6 @@
 	cd ntopng-@NTOPNG_VERSION@; svn co https://svn.ntop.org/svn/ntop/trunk/ntopng/; cd ntopng; svn co https://svn.ntop.org/svn/ntop/trunk/nDPI/; cd ..; find ntopng -name .svn | xargs rm -rf ; mv ntopng ntopng-@NTOPNG_VERSION@; tar cvfz ../ntopng-@NTOPNG_VERSION@.tgz ntopng-@NTOPNG_VERSION@
 
 install: ntopng
-	@echo "Make sure you have already run 'make geoip' to also install geoip dat files"
-	@echo "While we provide you an install make target, we encourage you"
-	@echo "to create a package and install that"
-	@echo "rpm - do 'make build-rpm'"
-	@echo "deb - do 'cd packages/ubuntu;./configure;make"
 	mkdir -p $(INSTALL_DIR)/share/ntopng $(MAN_DIR)/man/man8 $(INSTALL_DIR)/bin
 	cp ntopng $(INSTALL_DIR)/bin
 	cp ./ntopng.8 $(MAN_DIR)/man/man8
