--- Cargo.toml.orig	2024-08-06 00:28:17.000000000 +0000
+++ Cargo.toml	2025-08-25 13:39:32.812252177 +0000
@@ -80,7 +80,6 @@
 
 [dependencies]
 anyhow = "1.0.86"
-backtrace = "0.3.73"
 cfg-if = "1.0.0"
 clap = { version = "4.5.13", features = ["default", "cargo", "wrap_help", "derive"] }
 concat-string = "1.0.1"
@@ -103,6 +102,11 @@
 unicode-segmentation = "1.11.0"
 unicode-width = "0.1.13"
 
+# Dependencies pinned for Rust 1.77.2
+backtrace = "=0.3.74"
+rayon = "=1.10.0"
+rayon-core = "=1.12.1"
+
 # Used for logging.
 fern = { version = "0.6.2", optional = true }
 log = { version = "0.4.22", optional = true }
