Description: zvbi-chains now load libzvbi-chains.so.0.0.0. (Closes: 283004)
Author: Christian Marillat <marillat@debian.org>
Reviewed-by: Joao Eriberto Mota Filho <eriberto@debian.org>
Last-Update: 2015-05-20
Index: zvbi-0.2.35/daemon/Makefile.am
===================================================================
--- zvbi-0.2.35.orig/daemon/Makefile.am
+++ zvbi-0.2.35/daemon/Makefile.am
@@ -16,7 +16,7 @@ EXTRA_DIST = \
 
 AM_CPPFLAGS = \
 	-I$(top_srcdir) \
-	-DLIBZVBI_CHAINS_PATH=\"libzvbi-chains.so\" \
+	-DLIBZVBI_CHAINS_PATH=\"libzvbi-chains.so.0.0.0\" \
 	-D_REENTRANT \
 	-D_GNU_SOURCE
 
