Description: fix write output to a file. (Closes: #603128)
Author: Christian Marillat <marillat@debian.org>
Last-Update: 2010-11-26
Index: zvbi-0.2.35/contrib/atsc-cc.c
===================================================================
--- zvbi-0.2.35.orig/contrib/atsc-cc.c
+++ zvbi-0.2.35/contrib/atsc-cc.c
@@ -1848,6 +1848,7 @@ cr_new_line			(struct caption_recorder *
 				      cr->ucs_buffer,
 				      cr->ucs_buffer_length,
 				      /* repl_char */ '?');
+		fflush(fp);
 	}
 }
 
