Description: fix some spelling errors found in final binaries.
Author: Joao Eriberto Mota Filho <eriberto@debian.org>
Last-Update: 2015-04-12
Index: zvbi-0.2.35/src/proxy-client.c
===================================================================
--- zvbi-0.2.35.orig/src/proxy-client.c
+++ zvbi-0.2.35/src/proxy-client.c
@@ -763,7 +763,7 @@ static vbi_bool proxy_client_start_acq(
       }
       else if (vpc->endianSwap)
       {  /* endian swapping currently unsupported */
-         asprintf(&vpc->p_errorstr, _("Incompatible server architecture (endianess mismatch)."));
+         asprintf(&vpc->p_errorstr, _("Incompatible server architecture (endianness mismatch)."));
          goto failure;
       }
       else
Index: zvbi-0.2.35/daemon/proxyd.c
===================================================================
--- zvbi-0.2.35.orig/daemon/proxyd.c
+++ zvbi-0.2.35/daemon/proxyd.c
@@ -852,7 +852,7 @@ static void vbi_proxyd_stop_acq_thread(
       {
          ret = pthread_join(p_proxy_dev->thread_id, NULL);
          if (ret == 0)
-            dprintf(DBG_MSG, "stop_acq_thread: acq thread killed sucessfully\n");
+            dprintf(DBG_MSG, "stop_acq_thread: acq thread killed successfully\n");
          else
             dprintf(DBG_MSG, "stop_acq_thread: pthread_join failed: %d (%s)\n", errno, strerror(errno));
       }
