Description: fix a FTBFS with GCC-6 (narrowing conversion)
Author: Breno Leitao <brenohl@br.ibm.com>
Last-Update: 2016-07-17
Index: zvbi-0.2.35/test/test-dvb_mux.cc
===================================================================
--- zvbi-0.2.35.orig/test/test-dvb_mux.cc
+++ zvbi-0.2.35/test/test-dvb_mux.cc
@@ -137,7 +137,7 @@ is_good_service			(vbi_service_set	servi
 static const vbi_service_set
 all_services [] = {
 	0,
-	-1,
+	UINT_MAX,
 	VBI_SLICED_2xCAPTION_525,
 	VBI_SLICED_CAPTION_525,
 	VBI_SLICED_CAPTION_525_F1,
@@ -1279,7 +1279,7 @@ test_multiplex_sliced_service_checks
 
 	/* Verify the service filter. */
 
-	if (-1u == service
+	if (UINT_MAX == service
 	    || (VBI_SLICED_TELETEXT_B_625
 		== (VBI_SLICED_TELETEXT_B_625 & service))) {
 		assert_multiplex_sliced (buffer, buffer_size,
@@ -3237,7 +3237,7 @@ static void
 test_dvb_mux_cor_pts (void)
 {
 	static const int64_t ptss [] = {
-		0x8000000000000000ll, -1, 0, 0x7FFFFFFFFFFFFFFFll,
+		0, -1, 0, 0x7FFFFFFFFFFFFFFFll,
 	};
 	DVBPESMuxTest mx;
 	unsigned int i;
