diff -uprN ../ffmpeg-4.3/configure ./configure
--- ../ffmpeg-4.3/configure	2020-06-15 14:54:23.000000000 -0400
+++ ./configure	2020-06-16 17:31:32.432716874 -0400
@@ -7382,7 +7382,7 @@ ifndef MAIN_MAKEFILE
 SRC_PATH:=\$(SRC_PATH:.%=..%)
 endif
 CC_IDENT=$cc_ident
-ARCH=$arch
+FFMPEG_ARCH=$arch
 INTRINSICS=$intrinsics
 EXTERN_PREFIX=$extern_prefix
 CC=$cc
diff -uprN ../ffmpeg-4.3/ffbuild/common.mak ./ffbuild/common.mak
--- ../ffmpeg-4.3/ffbuild/common.mak	2020-06-15 14:54:23.000000000 -0400
+++ ./ffbuild/common.mak	2020-06-16 17:31:32.433716941 -0400
@@ -136,8 +136,8 @@ DEP_LIBS := $(foreach lib,$(FFLIBS),$(ca
 STATIC_DEP_LIBS := $(foreach lib,$(FFLIBS),$(call PATH_LIBNAME,$(lib)))
 
 SRC_DIR    := $(SRC_PATH)/lib$(NAME)
-ALLHEADERS := $(subst $(SRC_DIR)/,$(SUBDIR),$(wildcard $(SRC_DIR)/*.h $(SRC_DIR)/$(ARCH)/*.h))
-SKIPHEADERS += $(ARCH_HEADERS:%=$(ARCH)/%) $(SKIPHEADERS-)
+ALLHEADERS := $(subst $(SRC_DIR)/,$(SUBDIR),$(wildcard $(SRC_DIR)/*.h $(SRC_DIR)/$(FFMPEG_ARCH)/*.h))
+SKIPHEADERS += $(ARCH_HEADERS:%=$(FFMPEG_ARCH)/%) $(SKIPHEADERS-)
 SKIPHEADERS := $(SKIPHEADERS:%=$(SUBDIR)%)
 HOBJS        = $(filter-out $(SKIPHEADERS:.h=.h.o),$(ALLHEADERS:.h=.h.o))
 PTXOBJS      = $(filter %.ptx.o,$(OBJS))
diff -uprN ../ffmpeg-4.3/Makefile ./Makefile
--- ../ffmpeg-4.3/Makefile	2020-06-15 14:54:23.000000000 -0400
+++ ./Makefile	2020-06-16 17:31:32.433716941 -0400
@@ -89,7 +89,7 @@ define DOSUBDIR
 $(foreach V,$(SUBDIR_VARS),$(eval $(call RESET,$(V))))
 SUBDIR := $(1)/
 include $(SRC_PATH)/$(1)/Makefile
--include $(SRC_PATH)/$(1)/$(ARCH)/Makefile
+-include $(SRC_PATH)/$(1)/$(FFMPEG_ARCH)/Makefile
 -include $(SRC_PATH)/$(1)/$(INTRINSICS)/Makefile
 include $(SRC_PATH)/ffbuild/library.mak
 endef
