diff -uprN ../ffmpeg-4.3-021/libavformat/movenc.c ./libavformat/movenc.c
--- ../ffmpeg-4.3-021/libavformat/movenc.c	2020-06-15 14:54:24.000000000 -0400
+++ ./libavformat/movenc.c	2020-06-16 20:00:07.562359224 -0400
@@ -103,6 +103,9 @@ static const AVOption options[] = {
     { "encryption_kid", "The media encryption key identifier (hex)", offsetof(MOVMuxContext, encryption_kid), AV_OPT_TYPE_BINARY, .flags = AV_OPT_FLAG_ENCODING_PARAM },
     { "use_stream_ids_as_track_ids", "use stream ids as track ids", offsetof(MOVMuxContext, use_stream_ids_as_track_ids), AV_OPT_TYPE_BOOL, {.i64 = 0}, 0, 1, AV_OPT_FLAG_ENCODING_PARAM},
     { "write_tmcd", "force or disable writing tmcd", offsetof(MOVMuxContext, write_tmcd), AV_OPT_TYPE_BOOL, {.i64 = -1}, -1, 1, AV_OPT_FLAG_ENCODING_PARAM},
+#ifdef SYNO_VIDEOSTATION_SKIP_DISPLAYMATRIX
+    { "skip_displaymatrix", "Skip writing displaymatrix metadata.", offsetof(MOVMuxContext, skip_displaymatrix), AV_OPT_TYPE_BOOL, {.i64 = 0}, 0, 1, AV_OPT_FLAG_ENCODING_PARAM},
+#endif
     { "write_prft", "Write producer reference time box with specified time source", offsetof(MOVMuxContext, write_prft), AV_OPT_TYPE_INT, {.i64 = MOV_PRFT_NONE}, 0, MOV_PRFT_NB-1, AV_OPT_FLAG_ENCODING_PARAM, "prft"},
     { "wallclock", NULL, 0, AV_OPT_TYPE_CONST, {.i64 = MOV_PRFT_SRC_WALLCLOCK}, 0, 0, AV_OPT_FLAG_ENCODING_PARAM, "prft"},
     { "pts", NULL, 0, AV_OPT_TYPE_CONST, {.i64 = MOV_PRFT_SRC_PTS}, 0, 0, AV_OPT_FLAG_ENCODING_PARAM, "prft"},
@@ -2933,6 +2936,11 @@ static int mov_write_tkhd_tag(AVIOContex
                                                             &display_matrix_size);
         if (display_matrix && display_matrix_size < 9 * sizeof(*display_matrix))
             display_matrix = NULL;
+#ifdef SYNO_VIDEOSTATION_SKIP_DISPLAYMATRIX
+        if (mov->skip_displaymatrix) {
+            display_matrix = NULL;
+        }
+#endif
     }
 
     if (track->flags & MOV_TRACK_ENABLED)
diff -uprN ../ffmpeg-4.3-021/libavformat/movenc.h ./libavformat/movenc.h
--- ../ffmpeg-4.3-021/libavformat/movenc.h	2020-06-15 14:54:24.000000000 -0400
+++ ./libavformat/movenc.h	2020-06-16 20:00:07.563359295 -0400
@@ -27,6 +27,8 @@
 #include "avformat.h"
 #include "movenccenc.h"
 
+#include "synoconfig.h"
+
 #define MOV_FRAG_INFO_ALLOC_INCREMENT 64
 #define MOV_INDEX_CLUSTER_SIZE 1024
 #define MOV_TIMESCALE 1000
@@ -193,6 +195,9 @@ typedef struct MOVMuxContext {
     int rtp_flags;
 
     int iods_skip;
+#ifdef SYNO_VIDEOSTATION_SKIP_DISPLAYMATRIX
+    int skip_displaymatrix;
+#endif
     int iods_video_profile;
     int iods_audio_profile;
 
diff -uprN ../ffmpeg-4.3-021/synoconfig.h ./synoconfig.h
--- ../ffmpeg-4.3-021/synoconfig.h	2020-06-16 19:57:31.891853716 -0400
+++ ./synoconfig.h	2020-06-16 20:00:07.563359295 -0400
@@ -14,3 +14,11 @@
 #if defined(SYNO_VIDEOSTATION)
 #define SYNO_VIDEOSTATION_HLS_COSTUMIZED_TS_NAME
 #endif
+
+/* Add skip displaymatrix config to movenc for video station offline transcoding.
+ * add flag "-skip_displaymatrix 1" to mov/mp4 format for skiping writing displaymatrix metadata.
+ * See Video Station #4464
+ */
+#if defined(SYNO_VIDEOSTATION)
+#define SYNO_VIDEOSTATION_SKIP_DISPLAYMATRIX
+#endif
