Index: jellyfin-ffmpeg/libavfilter/vf_overlay_vaapi.c
===================================================================
--- libavfilter/vf_overlay_vaapi.c
+++ libavfilter/vf_overlay_vaapi.c
@@ -312,8 +312,10 @@ static int overlay_vaapi_config_input_ov
         ctx->blend_alpha  = ctx->alpha;
     }
 
+    // VA_BLEND_PREMULTIPLIED_ALPHA may cause issues in
+    // per-pixel alpha case, disable it to align with MSDK.
     if (have_alpha_planar(inlink))
-        ctx->blend_flags |= VA_BLEND_PREMULTIPLIED_ALPHA;
+        ctx->blend_flags |= 0;
 
     return 0;
 }
