--- libavutil/hwcontext_vaapi.c
+++ libavutil/hwcontext_vaapi.c
@@ -1735,7 +1735,7 @@ static int vaapi_device_create(AVHWDeviceContext *ctx, const char *device,
                 if (priv->drm_fd < 0) {
                     av_log(ctx, AV_LOG_VERBOSE, "Cannot open "
                            "DRM render node for device %d.\n", n);
-                    break;
+                    continue;
                 }
 #if CONFIG_LIBDRM
                 info = drmGetVersion(priv->drm_fd);
