--- libavutil/hwcontext_vaapi.c
+++ libavutil/hwcontext_vaapi.c
@@ -18,15 +18,6 @@
 
 #include "config.h"
 
-#if HAVE_VAAPI_WIN32
-#   include <windows.h>
-#define COBJMACROS
-#   include <initguid.h>
-#   include <dxgi1_2.h>
-#   include "compat/w32dlfcn.h"
-#   include <va/va_win32.h>
-typedef HRESULT (WINAPI *PFN_CREATE_DXGI_FACTORY)(REFIID riid, void **ppFactory);
-#endif
 #if HAVE_VAAPI_X11
 #   include <va/va_x11.h>
 #endif
@@ -217,7 +208,7 @@ static int vaapi_get_image_format(AVHWDeviceContext *hwdev,
             return 0;
         }
     }
-    return AVERROR(ENOSYS);
+    return AVERROR(EINVAL);
 }
 
 static int vaapi_frames_get_constraints(AVHWDeviceContext *hwdev,
@@ -817,7 +808,7 @@ static int vaapi_map_frame(AVHWFramesContext *hwfc,
     err = vaapi_get_image_format(hwfc->device_ctx, dst->format, &image_format);
     if (err < 0) {
         // Requested format is not a valid output format.
-        return err;
+        return AVERROR(EINVAL);
     }
 
     map = av_malloc(sizeof(*map));
@@ -992,7 +983,7 @@ static int vaapi_map_to_memory(AVHWFramesContext *hwfc, AVFrame *dst,
     if (dst->format != AV_PIX_FMT_NONE) {
         err = vaapi_get_image_format(hwfc->device_ctx, dst->format, NULL);
         if (err < 0)
-            return err;
+            return AVERROR(ENOSYS);
     }
 
     err = vaapi_map_frame(hwfc, dst, src, flags);
@@ -1048,9 +1039,6 @@ static const struct {
 #if defined(VA_FOURCC_Y412) && defined(DRM_FORMAT_XVYU12_16161616)
     DRM_MAP(Y412, 1, DRM_FORMAT_XVYU12_16161616),
 #endif
-#if defined(VA_FOURCC_X2R10G10B10) && defined(DRM_FORMAT_XRGB2101010)
-    DRM_MAP(X2R10G10B10, 1, DRM_FORMAT_XRGB2101010),
-#endif
 };
 #undef DRM_MAP
 
@@ -1336,9 +1324,8 @@ static int vaapi_map_to_drm_esh(AVHWFramesContext *hwfc, AVFrame *dst,
 
         vas = vaSyncSurface(hwctx->display, surface_id);
         if (vas != VA_STATUS_SUCCESS) {
-            av_log(hwfc, AV_LOG_ERROR, "Failed to sync surface "
+            av_log(hwfc, AV_LOG_WARNING, "Failed to sync surface "
                    "%#x: %d (%s).\n", surface_id, vas, vaErrorStr(vas));
-            return AVERROR(EIO);
         }
     }
 
@@ -1675,7 +1662,7 @@ static int vaapi_device_create(AVHWDeviceContext *ctx, const char *device,
     VAAPIDevicePriv *priv;
     VADisplay display = NULL;
     const AVDictionaryEntry *ent;
-    int try_drm, try_x11, try_win32, try_all;
+    int try_drm, try_x11, try_all;
 
     priv = av_mallocz(sizeof(*priv));
     if (!priv)
@@ -1688,13 +1675,11 @@ static int vaapi_device_create(AVHWDeviceContext *ctx, const char *device,
 
     ent = av_dict_get(opts, "connection_type", NULL, 0);
     if (ent) {
-        try_all = try_drm = try_x11 = try_win32 = 0;
+        try_all = try_drm = try_x11 = 0;
         if (!strcmp(ent->value, "drm")) {
             try_drm = 1;
         } else if (!strcmp(ent->value, "x11")) {
             try_x11 = 1;
-        } else if (!strcmp(ent->value, "win32")) {
-            try_win32 = 1;
         } else {
             av_log(ctx, AV_LOG_ERROR, "Invalid connection type %s.\n",
                    ent->value);
@@ -1704,7 +1689,6 @@ static int vaapi_device_create(AVHWDeviceContext *ctx, const char *device,
         try_all = 1;
         try_drm = HAVE_VAAPI_DRM;
         try_x11 = HAVE_VAAPI_X11;
-        try_win32 = HAVE_VAAPI_WIN32;
     }
 
 #if HAVE_VAAPI_DRM
@@ -1812,68 +1796,6 @@ static int vaapi_device_create(AVHWDeviceContext *ctx, const char *device,
     }
 #endif
 
-#if HAVE_VAAPI_WIN32
-    if (!display && try_win32) {
-        // Try to create a display from the specified device, if any.
-        if (!device) {
-            display = vaGetDisplayWin32(NULL);
-        } else {
-            IDXGIFactory2 *pDXGIFactory = NULL;
-            IDXGIAdapter *pAdapter = NULL;
-#if !HAVE_UWP
-            HANDLE dxgi = dlopen("dxgi.dll", 0);
-            if (!dxgi) {
-                av_log(ctx, AV_LOG_ERROR, "Failed to load dxgi.dll\n");
-                return AVERROR_UNKNOWN;
-            }
-            PFN_CREATE_DXGI_FACTORY pfnCreateDXGIFactory =
-                (PFN_CREATE_DXGI_FACTORY)dlsym(dxgi, "CreateDXGIFactory");
-            if (!pfnCreateDXGIFactory) {
-                av_log(ctx, AV_LOG_ERROR, "CreateDXGIFactory load failed\n");
-                dlclose(dxgi);
-                return AVERROR_UNKNOWN;
-            }
-#else
-            // In UWP (which lacks LoadLibrary), CreateDXGIFactory isn't
-            // available, only CreateDXGIFactory1
-            PFN_CREATE_DXGI_FACTORY pfnCreateDXGIFactory =
-                (PFN_CREATE_DXGI_FACTORY)CreateDXGIFactory1;
-#endif
-            if (SUCCEEDED(pfnCreateDXGIFactory(&IID_IDXGIFactory2,
-                                              (void **)&pDXGIFactory))) {
-                int adapter = atoi(device);
-                if (SUCCEEDED(IDXGIFactory2_EnumAdapters(pDXGIFactory,
-                                                         adapter,
-                                                         &pAdapter))) {
-                    DXGI_ADAPTER_DESC desc;
-                    if (SUCCEEDED(IDXGIAdapter2_GetDesc(pAdapter, &desc))) {
-                        av_log(ctx, AV_LOG_INFO,
-                              "Using device %04x:%04x (%ls) - LUID %lu %ld.\n",
-                              desc.VendorId, desc.DeviceId, desc.Description,
-                              desc.AdapterLuid.LowPart,
-                              desc.AdapterLuid.HighPart);
-                        display = vaGetDisplayWin32(&desc.AdapterLuid);
-                    }
-                    IDXGIAdapter_Release(pAdapter);
-                }
-                IDXGIFactory2_Release(pDXGIFactory);
-            }
-#if !HAVE_UWP
-            dlclose(dxgi);
-#endif
-        }
-
-        if (!display) {
-            av_log(ctx, AV_LOG_ERROR, "Cannot open a VA display "
-                    "from Win32 display.\n");
-            return AVERROR_UNKNOWN;
-        }
-
-        av_log(ctx, AV_LOG_VERBOSE, "Opened VA display via "
-                "Win32 display.\n");
-    }
-#endif
-
     if (!display) {
         if (device)
             av_log(ctx, AV_LOG_ERROR, "No VA display found for "
