Index: jellyfin-ffmpeg/libavformat/assenc.c
===================================================================
--- libavformat/assenc.c
+++ libavformat/assenc.c
@@ -25,6 +25,7 @@
 #include "mux.h"
 
 #include "libavutil/opt.h"
+#include "libavutil/mem.h"
 
 typedef struct DialogueLine {
     int readorder;
@@ -56,6 +57,7 @@ static int write_header(AVFormatContext
     avpriv_set_pts_info(s->streams[0], 64, 1, 100);
     if (par->extradata_size > 0) {
         size_t header_size = par->extradata_size;
+        char *header_string = NULL;
         uint8_t *trailer = strstr(par->extradata, "\n[Events]");
 
         if (trailer)
@@ -70,9 +72,20 @@ static int write_header(AVFormatContext
                 ass->trailer = trailer;
         }
 
-        avio_write(s->pb, par->extradata, header_size);
-        if (par->extradata[header_size - 1] != '\n')
-            avio_write(s->pb, "\r\n", 2);
+        header_string = av_malloc(header_size + 1);
+        if (!header_string)
+            return AVERROR(ENOMEM);
+
+        memcpy(header_string, par->extradata, header_size);
+        header_string[header_size] = 0;
+
+        avio_printf(s->pb, "%s", header_string);
+
+        if (header_string[strlen(header_string) - 1] != '\n')
+            avio_printf(s->pb, "\r\n");
+
+        av_free(header_string);
+
         ass->ssa_mode = !strstr(par->extradata, "\n[V4+ Styles]");
         if (!strstr(par->extradata, "\n[Events]"))
             avio_printf(s->pb, "[Events]\r\nFormat: %s, Start, End, Style, Name, MarginL, MarginR, MarginV, Effect, Text\r\n",
