Index: jellyfin-ffmpeg/libavcodec/qsvenc_av1.c
===================================================================
--- libavcodec/qsvenc_av1.c
+++ libavcodec/qsvenc_av1.c
@@ -130,8 +130,8 @@ static const AVClass class = {
 
 static const FFCodecDefault qsv_enc_defaults[] = {
     { "b",         "1M"   },
-    { "g",         "-1"   },
-    { "bf",        "-1"   },
+    { "g",         "250"  },
+    { "bf",        "3"    },
     { "refs",      "0"    },
     { NULL },
 };
Index: jellyfin-ffmpeg/libavcodec/qsvenc_h264.c
===================================================================
--- libavcodec/qsvenc_h264.c
+++ libavcodec/qsvenc_h264.c
@@ -180,8 +180,8 @@ static const AVClass class = {
 static const FFCodecDefault qsv_enc_defaults[] = {
     { "b",         "1M"    },
     { "refs",      "0"     },
-    { "g",         "-1"    },
-    { "bf",        "-1"    },
+    { "g",         "250"   },
+    { "bf",        "3"     },
     { "qmin",      "-1"    },
     { "qmax",      "-1"    },
     { "trellis",   "-1"    },
Index: jellyfin-ffmpeg/libavcodec/qsvenc_hevc.c
===================================================================
--- libavcodec/qsvenc_hevc.c
+++ libavcodec/qsvenc_hevc.c
@@ -376,8 +376,8 @@ static const AVClass class = {
 static const FFCodecDefault qsv_enc_defaults[] = {
     { "b",         "1M"    },
     { "refs",      "0"     },
-    { "g",         "-1"    },
-    { "bf",        "-1"    },
+    { "g",         "250"   },
+    { "bf",        "4"     },
     { "qmin",      "-1"    },
     { "qmax",      "-1"    },
     { "trellis",   "-1"    },
