Index: jellyfin-ffmpeg/libavutil/hwcontext_vaapi.c
===================================================================
--- libavutil/hwcontext_vaapi.c
+++ libavutil/hwcontext_vaapi.c
@@ -150,6 +150,7 @@ static const VAAPIFormatDescriptor vaapi
     MAP(XRGB, RGB32,   0RGB, 0),
 #ifdef VA_FOURCC_X2R10G10B10
     MAP(X2R10G10B10, RGB32_10, X2RGB10, 0),
+    MAP(X2B10G10R10, RGB32_10, X2BGR10, 0),
 #endif
 #ifdef VA_FOURCC_Y410
     // libva doesn't include a fourcc for XV30 and the driver only declares
@@ -1016,9 +1017,11 @@ static const struct {
     DRM_MAP(NV12, 1, DRM_FORMAT_NV12),
 #if defined(VA_FOURCC_P010) && defined(DRM_FORMAT_R16)
     DRM_MAP(P010, 2, DRM_FORMAT_R16, DRM_FORMAT_RG1616),
+    DRM_MAP(P010, 2, DRM_FORMAT_R16, DRM_FORMAT_GR1616),
 #endif
 #if defined(VA_FOURCC_P012) && defined(DRM_FORMAT_R16)
     DRM_MAP(P012, 2, DRM_FORMAT_R16, DRM_FORMAT_RG1616),
+    DRM_MAP(P012, 2, DRM_FORMAT_R16, DRM_FORMAT_GR1616),
 #endif
     DRM_MAP(BGRA, 1, DRM_FORMAT_ARGB8888),
     DRM_MAP(BGRX, 1, DRM_FORMAT_XRGB8888),
@@ -1030,6 +1033,10 @@ static const struct {
 #endif
     DRM_MAP(ARGB, 1, DRM_FORMAT_BGRA8888),
     DRM_MAP(XRGB, 1, DRM_FORMAT_BGRX8888),
+#ifdef VA_FOURCC_X2R10G10B10
+    DRM_MAP(X2R10G10B10, 1, DRM_FORMAT_XRGB2101010),
+    DRM_MAP(X2B10G10R10, 1, DRM_FORMAT_XBGR2101010),
+#endif
 #if defined(VA_FOURCC_XYUV) && defined(DRM_FORMAT_XYUV8888)
     DRM_MAP(XYUV, 1, DRM_FORMAT_XYUV8888),
 #endif
@@ -1094,12 +1101,6 @@ static int vaapi_map_from_drm(AVHWFrames
 
     desc = (AVDRMFrameDescriptor*)src->data[0];
 
-    if (desc->nb_objects != 1) {
-        av_log(dst_fc, AV_LOG_ERROR, "VAAPI can only map frames "
-               "made from a single DRM object.\n");
-        return AVERROR(EINVAL);
-    }
-
     va_fourcc = 0;
     for (i = 0; i < FF_ARRAY_ELEMS(vaapi_drm_format_map); i++) {
         if (desc->nb_layers != vaapi_drm_format_map[i].nb_layer_formats)
@@ -1239,6 +1240,12 @@ static int vaapi_map_from_drm(AVHWFrames
                                buffer_attrs, FF_ARRAY_ELEMS(buffer_attrs));
     }
 #else
+    if (desc->nb_objects != 1) {
+        av_log(dst_fc, AV_LOG_ERROR, "VAAPI can only map frames "
+               "made from a single DRM object.\n");
+        return AVERROR(EINVAL);
+    }
+
     buffer_handle = desc->objects[0].fd;
     buffer_desc.pixel_format = va_fourcc;
     buffer_desc.width        = src_fc->width;
