Index: jellyfin-ffmpeg/libavfilter/vf_libplacebo.c
===================================================================
--- libavfilter/vf_libplacebo.c
+++ libavfilter/vf_libplacebo.c
@@ -98,7 +98,6 @@ typedef struct LibplaceboContext {
     float polar_cutoff;
     int disable_linear;
     int disable_builtin;
-    int force_icc_lut;
     int force_dither;
     int disable_fbos;
 
@@ -138,6 +137,7 @@ typedef struct LibplaceboContext {
     float desat_exp;
     int gamut_warning;
     int gamut_clipping;
+    int force_icc_lut;
 
      /* pl_dither_params */
     int dithering;
@@ -414,7 +414,9 @@ static int process_frames(AVFilterContex
             .minimum_peak = s->min_peak,
             .scene_threshold_low = s->scene_low,
             .scene_threshold_high = s->scene_high,
+#if PL_API_VER < 256
             .overshoot_margin = s->overshoot,
+#endif
         ),
 
         .color_map_params = pl_color_map_params(
@@ -446,7 +448,6 @@ static int process_frames(AVFilterContex
         .polar_cutoff = s->polar_cutoff,
         .disable_linear_scaling = s->disable_linear,
         .disable_builtin_scalers = s->disable_builtin,
-        .force_icc_lut = s->force_icc_lut,
         .force_dither = s->force_dither,
         .disable_fbos = s->disable_fbos,
     };
@@ -774,7 +775,7 @@ static const AVOption libplacebo_options
     { "polar_cutoff", "Polar LUT cutoff", OFFSET(polar_cutoff), AV_OPT_TYPE_FLOAT, {.dbl = 0}, 0.0, 1.0, DYNAMIC },
     { "disable_linear", "Disable linear scaling", OFFSET(disable_linear), AV_OPT_TYPE_BOOL, {.i64 = 0}, 0, 1, DYNAMIC },
     { "disable_builtin", "Disable built-in scalers", OFFSET(disable_builtin), AV_OPT_TYPE_BOOL, {.i64 = 0}, 0, 1, DYNAMIC },
-    { "force_icc_lut", "Force the use of a full ICC 3DLUT for color mapping", OFFSET(force_icc_lut), AV_OPT_TYPE_BOOL, {.i64 = 0}, 0, 1, DYNAMIC },
+    { "force_icc_lut", "Deprecated, does nothing", OFFSET(force_icc_lut), AV_OPT_TYPE_BOOL, {.i64 = 0}, 0, 1, DYNAMIC | AV_OPT_FLAG_DEPRECATED },
     { "force_dither", "Force dithering", OFFSET(force_dither), AV_OPT_TYPE_BOOL, {.i64 = 0}, 0, 1, DYNAMIC },
     { "disable_fbos", "Force-disable FBOs", OFFSET(disable_fbos), AV_OPT_TYPE_BOOL, {.i64 = 0}, 0, 1, DYNAMIC },
     { NULL },
Index: jellyfin-ffmpeg/libavutil/hwcontext_vulkan.c
===================================================================
--- libavutil/hwcontext_vulkan.c
+++ libavutil/hwcontext_vulkan.c
@@ -1394,6 +1394,7 @@ static int vulkan_device_create_internal
         goto end;
     }
     p->device_features_1_2.timelineSemaphore = 1;
+    p->device_features_1_2.hostQueryReset = dev_features_1_2.hostQueryReset;
 
     /* Setup queue family */
     if ((err = setup_queue_families(ctx, &dev_info)))
