--- libavfilter/qsvvpp.c
+++ libavfilter/qsvvpp.c
@@ -463,11 +463,13 @@ static QSVFrame *submit_frame(QSVVPPCont
             !qsv_frame->frame->interlaced_frame ? MFX_PICSTRUCT_PROGRESSIVE :
             (qsv_frame->frame->top_field_first ? MFX_PICSTRUCT_FIELD_TFF :
                                                  MFX_PICSTRUCT_FIELD_BFF);
-    if (qsv_frame->frame->repeat_pict == 1)
+    if (qsv_frame->frame->repeat_pict == 1) {
         qsv_frame->surface.Info.PicStruct |= MFX_PICSTRUCT_FIELD_REPEATED;
-    else if (qsv_frame->frame->repeat_pict == 2)
+        qsv_frame->surface.Info.PicStruct |= qsv_frame->frame->top_field_first ? MFX_PICSTRUCT_FIELD_TFF :
+                                                                                 MFX_PICSTRUCT_FIELD_BFF;
+    } else if (qsv_frame->frame->repeat_pict == 2) {
         qsv_frame->surface.Info.PicStruct |= MFX_PICSTRUCT_FRAME_DOUBLING;
-    else if (qsv_frame->frame->repeat_pict == 4)
+    } else if (qsv_frame->frame->repeat_pict == 4)
         qsv_frame->surface.Info.PicStruct |= MFX_PICSTRUCT_FRAME_TRIPLING;
 
     return qsv_frame;
