--- gitweb/gitweb.perl.orig	2015-04-30 11:55:48.319824040 +0200
+++ gitweb/gitweb.perl	2015-04-30 11:56:13.535823034 +0200
@@ -42,7 +42,7 @@
 
 	# Base URL for relative URLs in gitweb ($logo, $favicon, ...),
 	# needed and used only for URLs with nonempty PATH_INFO
-	our $base_url = $my_url;
+	our $base_url = $my_url . "/";
 
 	# When the script is used as DirectoryIndex, the URL does not contain the name
 	# of the script file itself, and $cgi->url() fails to strip PATH_INFO, so we
