diff -uprN ../ffmpeg-4.3-032/libavformat/wavenc.c ./libavformat/wavenc.c
--- ../ffmpeg-4.3-032/libavformat/wavenc.c	2020-06-15 14:54:24.000000000 -0400
+++ ./libavformat/wavenc.c	2020-06-16 20:10:39.073227980 -0400
@@ -46,6 +46,8 @@
 #include "internal.h"
 #include "riff.h"
 
+#include "synoconfig.h"
+
 #define RF64_AUTO   (-1)
 #define RF64_NEVER  0
 #define RF64_ALWAYS 1
@@ -84,6 +86,9 @@ typedef struct WAVMuxContext {
     int peak_block_pos;
     int peak_ppv;
     int peak_bps;
+#ifdef SYNO_WRITE_RIFF_INFO_TAG_TO_WAV
+    int write_info_tag;
+#endif
 } WAVMuxContext;
 
 #if CONFIG_WAV_MUXER
@@ -355,7 +360,13 @@ static int wav_write_header(AVFormatCont
 
     if (wav->write_peak != PEAK_ONLY) {
         /* info header */
+#ifdef SYNO_WRITE_RIFF_INFO_TAG_TO_WAV
+        if (wav->write_info_tag) {
+            ff_riff_write_info(s);
+        }
+#else
         ff_riff_write_info(s);
+#endif
 
         /* data header */
         wav->data = ff_start_tag(pb, "data");
@@ -491,6 +502,9 @@ static const AVOption options[] = {
     { "peak_block_size", "Number of audio samples used to generate each peak frame.",   OFFSET(peak_block_size), AV_OPT_TYPE_INT, { .i64 = 256 }, 0, 65536, ENC },
     { "peak_format",     "The format of the peak envelope data (1: uint8, 2: uint16).", OFFSET(peak_format), AV_OPT_TYPE_INT,     { .i64 = PEAK_FORMAT_UINT16 }, PEAK_FORMAT_UINT8, PEAK_FORMAT_UINT16, ENC },
     { "peak_ppv",        "Number of peak points per peak value (1 or 2).",              OFFSET(peak_ppv), AV_OPT_TYPE_INT, { .i64 = 2 }, 1, 2, ENC },
+#ifdef SYNO_WRITE_RIFF_INFO_TAG_TO_WAV
+    { "write_info_tag",      "Write RIFF info tag.",     OFFSET(write_info_tag),            AV_OPT_TYPE_INT, { .i64 = 1  }, 0, 1, ENC },
+#endif
     { NULL },
 };
 
diff -uprN ../ffmpeg-4.3-032/synoconfig.h ./synoconfig.h
--- ../ffmpeg-4.3-032/synoconfig.h	2020-06-16 20:10:09.000000000 -0400
+++ ./synoconfig.h	2020-06-16 20:10:39.074228048 -0400
@@ -61,3 +61,8 @@
 #if defined(SYNO_VIDEOSTATION)
 #define SYNO_FIX_MP4_REMUX_H264PARSER
 #endif
+
+/* Optional for writing RIFF info tag to wav
+ * See Media Server #386
+ */
+#define SYNO_WRITE_RIFF_INFO_TAG_TO_WAV
