Index: FFmpeg/libavcodec/videotoolboxenc.c
===================================================================
--- libavcodec/videotoolboxenc.c
+++ libavcodec/videotoolboxenc.c
@@ -1035,9 +1035,10 @@ static int create_cv_pixel_buffer_info(A
     CFNumberRef width_num = NULL;
     CFNumberRef height_num = NULL;
     CFMutableDictionaryRef pixel_buffer_info = NULL;
+    enum AVPixelFormat pix_fmt = avctx->pix_fmt == AV_PIX_FMT_VIDEOTOOLBOX ? avctx->sw_pix_fmt : avctx->pix_fmt;
     int cv_color_format;
     int status = get_cv_pixel_format(avctx,
-                                     avctx->pix_fmt,
+                                     pix_fmt,
                                      avctx->color_range,
                                      &cv_color_format,
                                      NULL);
@@ -1692,11 +1693,9 @@ static int vtenc_configure_encoder(AVCod
                              kCFBooleanTrue);
     }
 
-    if (avctx->pix_fmt != AV_PIX_FMT_VIDEOTOOLBOX) {
-        status = create_cv_pixel_buffer_info(avctx, &pixel_buffer_info);
-        if (status)
-            goto init_cleanup;
-    }
+    status = create_cv_pixel_buffer_info(avctx, &pixel_buffer_info);
+    if (status)
+        goto init_cleanup;
 
     vtctx->dts_delta = vtctx->has_b_frames ? -1 : 0;
 
