Index: FFmpeg/libswscale/swscale.c
===================================================================
--- libswscale/swscale.c
+++ libswscale/swscale.c
@@ -1430,7 +1430,12 @@ int sws_frame_setup(SwsContext *ctx, con
 
     /* For now, if a single frame has a context, then both need a context */
     if (!!src->hw_frames_ctx != !!dst->hw_frames_ctx) {
-        return AVERROR(ENOTSUP);
+        const AVPixFmtDescriptor *src_desc = av_pix_fmt_desc_get(src->format);
+        const AVPixFmtDescriptor *dst_desc = av_pix_fmt_desc_get(dst->format);
+        const int is_mapped_hwframe_to_sw = !(src_desc->flags & AV_PIX_FMT_FLAG_HWACCEL) &&
+                                            !(dst_desc->flags & AV_PIX_FMT_FLAG_HWACCEL);
+        if (!is_mapped_hwframe_to_sw)
+            return AVERROR(ENOTSUP);
     } else if (!!src->hw_frames_ctx) {
         /* Both hardware frames must already be allocated */
         if (!src->data[0] || !dst->data[0])
