Index: FFmpeg/libavcodec/vaapi_encode_mjpeg.c
===================================================================
--- libavcodec/vaapi_encode_mjpeg.c
+++ libavcodec/vaapi_encode_mjpeg.c
@@ -583,7 +583,7 @@ const FFCodec ff_mjpeg_vaapi_encoder = {
                       FF_CODEC_CAP_INIT_CLEANUP,
     .defaults       = vaapi_encode_mjpeg_defaults,
     CODEC_PIXFMTS(AV_PIX_FMT_VAAPI),
-    .color_ranges   = AVCOL_RANGE_MPEG, /* FIXME: implement tagging */
+    .color_ranges   = AVCOL_RANGE_MPEG | AVCOL_RANGE_JPEG, /* Depends on driver impl */
     .hw_configs     = ff_vaapi_encode_hw_configs,
     .p.wrapper_name = "vaapi",
 };
