Index: FFmpeg/libavcodec/qsvdec.c
===================================================================
--- libavcodec/qsvdec.c
+++ libavcodec/qsvdec.c
@@ -708,17 +708,35 @@ static int qsv_export_hdr_side_data(AVCo
     // The SDK reuses this flag for HDR SEI parsing
     if (mdcv->InsertPayloadToggle) {
         AVMasteringDisplayMetadata *mastering;
-        const int mapping[3] = {2, 0, 1};
-        const int chroma_den = 50000;
-        const int luma_den = 10000;
-        int i;
+        int mapping[3] = { 0, 1, 2 };
+        int chroma_den;
+        int max_luma_den;
+        int min_luma_den;
+
+        switch (avctx->codec_id) {
+        case AV_CODEC_ID_HEVC:
+            mapping[0] = 2;
+            mapping[1] = 0;
+            mapping[2] = 1;
+            chroma_den = 50000;
+            max_luma_den = 10000;
+            min_luma_den = 10000;
+            break;
+        case AV_CODEC_ID_AV1:
+            chroma_den = 1 << 16;
+            max_luma_den = 1 << 8;
+            min_luma_den = 1 << 14;
+            break;
+        default:
+            return AVERROR(ENOSYS);
+        }
 
         ret = ff_decode_mastering_display_new(avctx, frame, &mastering);
         if (ret < 0)
             return ret;
 
         if (mastering) {
-            for (i = 0; i < 3; i++) {
+            for (int i = 0; i < 3; i++) {
                 const int j = mapping[i];
                 mastering->display_primaries[i][0] = av_make_q(mdcv->DisplayPrimariesX[j], chroma_den);
                 mastering->display_primaries[i][1] = av_make_q(mdcv->DisplayPrimariesY[j], chroma_den);
@@ -727,8 +745,8 @@ static int qsv_export_hdr_side_data(AVCo
             mastering->white_point[0] = av_make_q(mdcv->WhitePointX, chroma_den);
             mastering->white_point[1] = av_make_q(mdcv->WhitePointY, chroma_den);
 
-            mastering->max_luminance = av_make_q(mdcv->MaxDisplayMasteringLuminance, luma_den);
-            mastering->min_luminance = av_make_q(mdcv->MinDisplayMasteringLuminance, luma_den);
+            mastering->max_luminance = av_make_q(mdcv->MaxDisplayMasteringLuminance, max_luma_den);
+            mastering->min_luminance = av_make_q(mdcv->MinDisplayMasteringLuminance, min_luma_den);
 
             mastering->has_luminance = 1;
             mastering->has_primaries = 1;
@@ -751,46 +769,6 @@ static int qsv_export_hdr_side_data(AVCo
 
     return 0;
 }
-
-static int qsv_export_hdr_side_data_av1(AVCodecContext *avctx, mfxExtMasteringDisplayColourVolume *mdcv,
-                                        mfxExtContentLightLevelInfo *clli, AVFrame *frame)
-{
-    if (mdcv->InsertPayloadToggle) {
-        AVMasteringDisplayMetadata *mastering = av_mastering_display_metadata_create_side_data(frame);
-        const int chroma_den   = 1 << 16;
-        const int max_luma_den = 1 << 8;
-        const int min_luma_den = 1 << 14;
-
-        if (!mastering)
-            return AVERROR(ENOMEM);
-
-        for (int i = 0; i < 3; i++) {
-            mastering->display_primaries[i][0] = av_make_q(mdcv->DisplayPrimariesX[i], chroma_den);
-            mastering->display_primaries[i][1] = av_make_q(mdcv->DisplayPrimariesY[i], chroma_den);
-        }
-
-        mastering->white_point[0] = av_make_q(mdcv->WhitePointX, chroma_den);
-        mastering->white_point[1] = av_make_q(mdcv->WhitePointY, chroma_den);
-
-        mastering->max_luminance = av_make_q(mdcv->MaxDisplayMasteringLuminance, max_luma_den);
-        mastering->min_luminance = av_make_q(mdcv->MinDisplayMasteringLuminance, min_luma_den);
-
-        mastering->has_luminance = 1;
-        mastering->has_primaries = 1;
-    }
-
-    if (clli->InsertPayloadToggle) {
-        AVContentLightMetadata *light = av_content_light_metadata_create_side_data(frame);
-        if (!light)
-            return AVERROR(ENOMEM);
-
-        light->MaxCLL  = clli->MaxContentLightLevel;
-        light->MaxFALL = clli->MaxPicAverageLightLevel;
-    }
-
-    return 0;
-}
-
 #endif
 
 static int h264_decode_fpa(H2645SEIFramePacking *fpa, AVFrame *frame)
@@ -1062,15 +1040,9 @@ static int qsv_decode(AVCodecContext *av
 #endif
 
 #if QSV_VERSION_ATLEAST(1, 35)
-        if (QSV_RUNTIME_VERSION_ATLEAST(q->ver, 1, 35) && avctx->codec_id == AV_CODEC_ID_HEVC) {
+        if ((QSV_RUNTIME_VERSION_ATLEAST(q->ver, 1, 35) && avctx->codec_id == AV_CODEC_ID_HEVC) ||
+            (QSV_RUNTIME_VERSION_ATLEAST(q->ver, 2, 9)  && avctx->codec_id == AV_CODEC_ID_AV1)) {
             ret = qsv_export_hdr_side_data(avctx, &aframe.frame->mdcv, &aframe.frame->clli, frame);
-
-            if (ret < 0)
-                return ret;
-        }
-
-        if (QSV_RUNTIME_VERSION_ATLEAST(q->ver, 2, 9) && avctx->codec_id == AV_CODEC_ID_AV1) {
-            ret = qsv_export_hdr_side_data_av1(avctx, &aframe.frame->mdcv, &aframe.frame->clli, frame);
             if (ret < 0)
                 return ret;
         }
