# avoid build errors by disabling parallel make
# 

--- autoconf/Makefile.in.orig	2018-03-04 23:16:36.000000000 +0000
+++ autoconf/Makefile.in	2022-06-25 13:23:46.988110010 +0000
@@ -107,3 +107,5 @@
 #
 .PHONY: all help static elf install clean check install-elf install-static makefiles install-pkgconfig
 #
+
+.NOTPARALLEL:
--- slsh/Makefile.in.orig	2018-03-04 23:16:36.000000000 +0000
+++ slsh/Makefile.in	2022-06-25 13:23:23.287540065 +0000
@@ -200,3 +200,4 @@
   install_docs install-static static install_common \
   install_slsh install_static_slsh install_slshrc
 
+.NOTPARALLEL:
--- src/Makefile.in.orig	2018-03-04 23:16:36.000000000 +0000
+++ src/Makefile.in	2022-06-25 13:23:01.777510398 +0000
@@ -260,3 +260,5 @@
 #---------------------------------------------------------------------------
 # Object Rules : These are created from the configure script (hopefully)
 #---------------------------------------------------------------------------
+
+.NOTPARALLEL:
--- modules/Makefile.in.orig	2018-03-04 23:16:36.000000000 +0000
+++ modules/Makefile.in	2022-06-25 13:23:36.322155457 +0000
@@ -219,3 +219,5 @@
 	-/bin/rm -f Makefile $(OBJDIR)/Makefile $(CONFIG_H)
 
 .PHONY: clean distclean all install runtests
+
+.NOTPARALLEL:
