--- libavformat/movenc.c	2016-03-29 04:25:29.000000000 +0200
+++ libavformat/movenc2.c	2016-05-29 23:58:28.358207860 +0200
@@ -4918,10 +4918,13 @@
 
         if (first[st->codec->codec_type] < 0)
             first[st->codec->codec_type] = i;
-        if (st->disposition & AV_DISPOSITION_DEFAULT) {
-            mov->tracks[i].flags |= MOV_TRACK_ENABLED;
-            enabled[st->codec->codec_type]++;
-        }
+        // note whether each media type has any enabled streams
+        if (st->disposition & AV_DISPOSITION_DEFAULT)
+            enabled[st->codec->codec_type] = 1;
+
+        // clear the enables for each stream and track
+        st->disposition &= ~AV_DISPOSITION_DEFAULT;
+        mov->tracks[i].flags &= ~MOV_TRACK_ENABLED;
     }
 
     for (i = 0; i < AVMEDIA_TYPE_NB; i++) {
@@ -4929,13 +4932,25 @@
         case AVMEDIA_TYPE_VIDEO:
         case AVMEDIA_TYPE_AUDIO:
         case AVMEDIA_TYPE_SUBTITLE:
-            if (enabled[i] > 1)
-                mov->per_stream_grouping = 1;
             if (!enabled[i] && first[i] >= 0)
-                mov->tracks[first[i]].flags |= MOV_TRACK_ENABLED;
+            enabled[i] = 1;
             break;
         }
     }
+    // translate the first enabled stream of each media type
+    // into its stream being default and its output track being enabled
+    for (i = 0; i < AVMEDIA_TYPE_NB; i++) {
+        switch (i) {
+        case AVMEDIA_TYPE_VIDEO:
+        case AVMEDIA_TYPE_AUDIO:
+        case AVMEDIA_TYPE_SUBTITLE:
+            if (enabled[i]) {
+                s->streams[first[i]]->disposition |= AV_DISPOSITION_DEFAULT;
+                 mov->tracks[first[i]].flags |= MOV_TRACK_ENABLED;
+            }
+             break;
+         }
+    }
 }
 
 static void mov_free(AVFormatContext *s)
