--- setup.py.orig	2014-02-12 17:44:57.000000000 +0100
+++ setup.py	2014-12-12 18:46:22.893491541 +0100
@@ -74,26 +74,19 @@
 if sys.platform == 'win32':
     libraries.append('ws2_32')
 
-if sys.byteorder == 'big':
-    macros = [('__BIG_ENDIAN__', '1')]
-else:
-    macros = [('__LITTLE_ENDIAN__', '1')]
-
 ext_modules = []
 if not hasattr(sys, 'pypy_version_info'):
     ext_modules.append(Extension('msgpack._packer',
                                  sources=['msgpack/_packer.cpp'],
                                  libraries=libraries,
                                  include_dirs=['.'],
-                                 define_macros=macros,
                                  ))
     ext_modules.append(Extension('msgpack._unpacker',
                                  sources=['msgpack/_unpacker.cpp'],
                                  libraries=libraries,
                                  include_dirs=['.'],
-                                 define_macros=macros,
                                  ))
-del libraries, macros
+del libraries
 
 
 desc = 'MessagePack (de)serializer.'
