--- Makefile.orig	2024-11-26 00:55:17.000000000 +0000
+++ Makefile	2025-01-03 23:05:51.677704322 +0000
@@ -36,7 +36,7 @@
 LDFLAGS?=-s
 #CFLAGS?=-ggdb
 CFLAGS?=-O3
-CFLAGS+=-std=c11 -Wpedantic -Wall -Wextra -Wstrict-prototypes -Wshadow -Wconversion
+#CFLAGS+=-std=c11 -Wpedantic -Wall -Wextra -Wstrict-prototypes -Wshadow -Wconversion
 # _LARGEFILE64_SOURCE may be considered obsolete
 CPPFLAGS+=-DLARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64
 
