--- configure.orig	2010-07-09 13:38:46.000000000 +0100
+++ configure	2013-11-15 23:54:27.000000000 +0000
@@ -26565,7 +26565,7 @@
 if test "${ac_cv_sys_restartable_syscalls+set}" = set; then
   $as_echo_n "(cached) " >&6
 else
-  if test "$cross_compiling" = yes; then
+  if test "$cross_compiling" = maybe; then
   { { $as_echo "$as_me:$LINENO: error: in \`$ac_pwd':" >&5
 $as_echo "$as_me: error: in \`$ac_pwd':" >&2;}
 { { $as_echo "$as_me:$LINENO: error: cannot run test program while cross compiling
@@ -27036,7 +27036,7 @@
 if test "${mysql_cv_gcc_atomic_builtins+set}" = set; then
   $as_echo_n "(cached) " >&6
 else
-  if test "$cross_compiling" = yes; then
+  if test "$cross_compiling" = maybe; then
   mysql_cv_gcc_atomic_builtins=no
 else
   cat >conftest.$ac_ext <<_ACEOF
@@ -48287,7 +48287,7 @@
   { $as_echo "$as_me:$LINENO: checking whether GCC atomic builtins are available" >&5
 $as_echo_n "checking whether GCC atomic builtins are available... " >&6; }
   # either define HAVE_IB_GCC_ATOMIC_BUILTINS or not
-  if test "$cross_compiling" = yes; then
+  if test "$cross_compiling" = maybe; then
   { { $as_echo "$as_me:$LINENO: error: in \`$ac_pwd':" >&5
 $as_echo "$as_me: error: in \`$ac_pwd':" >&2;}
 { { $as_echo "$as_me:$LINENO: error: cannot run test program while cross compiling
@@ -48393,7 +48393,7 @@
   { $as_echo "$as_me:$LINENO: checking whether pthread_t can be used by GCC atomic builtins" >&5
 $as_echo_n "checking whether pthread_t can be used by GCC atomic builtins... " >&6; }
   # either define HAVE_IB_ATOMIC_PTHREAD_T_GCC or not
-  if test "$cross_compiling" = yes; then
+  if test "$cross_compiling" = maybe; then
   { { $as_echo "$as_me:$LINENO: error: in \`$ac_pwd':" >&5
 $as_echo "$as_me: error: in \`$ac_pwd':" >&2;}
 { { $as_echo "$as_me:$LINENO: error: cannot run test program while cross compiling
@@ -48595,7 +48595,7 @@
   { $as_echo "$as_me:$LINENO: checking whether pthread_t can be used by Solaris libc atomic functions" >&5
 $as_echo_n "checking whether pthread_t can be used by Solaris libc atomic functions... " >&6; }
   # either define HAVE_IB_ATOMIC_PTHREAD_T_SOLARIS or not
-  if test "$cross_compiling" = yes; then
+  if test "$cross_compiling" = maybe; then
   { { $as_echo "$as_me:$LINENO: error: in \`$ac_pwd':" >&5
 $as_echo "$as_me: error: in \`$ac_pwd':" >&2;}
 { { $as_echo "$as_me:$LINENO: error: cannot run test program while cross compiling
@@ -51555,7 +51555,7 @@
 if test -n "$CONFIG_FILES"; then
 
 
-ac_cr=''
+ac_cr=''
 ac_cs_awk_cr=`$AWK 'BEGIN { print "a\rb" }' </dev/null 2>/dev/null`
 if test "$ac_cs_awk_cr" = "a${ac_cr}b"; then
   ac_cs_awk_cr='\\r'
