--- rabbitmq-server.orig	2021-02-15 19:14:15.000000000 +0100
+++ ./rabbitmq-server	2021-02-17 15:52:41.936138655 +0100
@@ -16,6 +16,10 @@
 ##  Copyright (c) 2007-2021 VMware, Inc. or its affiliates.  All rights reserved.
 ##

+export HOME=/var/packages/rabbitmq/home
+export RABBITMQ_HOME=/var/packages/rabbitmq/target
+export RABBITMQ_CONF_ENV_FILE=${RABBITMQ_HOME}/etc/rabbitmq/rabbitmq-env.conf
+
 set -e

 # Get default settings with user overrides for (RABBITMQ_)<var_name>
