--- deps/uv/src/unix/async.c.orig	2013-04-22 18:34:20.000000000 +0200
+++ deps/uv/src/unix/async.c	2013-04-22 18:34:43.000000000 +0200
@@ -104,8 +104,6 @@
                          : "m"  (*pending));
     return val != 0;
   }
-#elif defined(__GNUC__) && (__GNUC__ > 4 || __GNUC__ == 4 && __GNUC_MINOR__ > 0)
-  return __sync_val_compare_and_swap(pending, 0, 1) != 0;
 #else
   ACCESS_ONCE(int, *pending) = 1;
   return 0;
