--- deps/openssl/openssl/crypto/mem_clr.c.orig	2014-08-19 17:19:52.000000000 +0200
+++ deps/openssl/openssl/crypto/mem_clr.c	2014-08-20 21:44:34.877847400 +0200
@@ -61,6 +61,7 @@
 
 unsigned char cleanse_ctr = 0;
 
+#if    !defined(__arm__)
 void OPENSSL_cleanse(void *ptr, size_t len)
 	{
 	unsigned char *p = ptr;
@@ -75,3 +76,4 @@
 		ctr += (63 + (size_t)p);
 	cleanse_ctr = (unsigned char)ctr;
 	}
+#endif
