--- deps/openssl/openssl.gyp.orig	2014-08-19 17:19:52.000000000 +0200
+++ deps/openssl/openssl.gyp	2014-08-20 21:39:55.742348191 +0200
@@ -939,7 +939,10 @@
           ],
         }],
         ['target_arch=="arm"', {
-          'sources': ['openssl/crypto/armcap.c'],
+          'sources': [
+            'openssl/crypto/armcap.c',
+            'openssl/crypto/armv4cpuid.S',
+          ],
         }],
       ],
       'include_dirs': [
