diff -ur src.orig/crypt.c src/crypt.c
--- src.orig/crypt.c	2011-10-11 12:44:58.000000000 +0200
+++ src/crypt.c	2012-02-13 21:24:40.000000000 +0100
@@ -44,12 +44,16 @@
 
 #ifdef USE_POLARSSL
 #include <polarssl/havege.h>
+#if defined POLARSSL_API_V1 && !defined(POLARSSL_API_V1_0)
+#define RAND_bytes(_dst_, _size_) havege_random(&hs, _dst_, _size_)
+#else
 #define RAND_bytes(_dst_, _size_) do { \
 	int i; \
 	for (i = 0; i < _size_; i++) { \
 	_dst_[i] = havege_rand(&hs); \
 	} \
  } while (0);
+#endif
 
 extern havege_state hs;
 #endif
diff -ur src.orig/crypt.h src/crypt.h
--- src.orig/crypt.h	2011-10-11 12:44:58.000000000 +0200
+++ src/crypt.h	2012-02-13 21:24:26.000000000 +0100
@@ -38,6 +38,19 @@
 #ifdef USE_POLARSSL
 #include <polarssl/havege.h>
 #include <polarssl/aes.h>
+#include <polarssl/version.h>
+#ifndef POLARSSL_VERSION_MAJOR
+	#define POLARSSL_API_V0
+#else
+#if (POLARSSL_VERSION_MAJOR == 0)
+	#define POLARSSL_API_V0
+#else
+#define POLARSSL_API_V1
+#if (POLARSSL_VERSION_MINOR == 0)
+	#define POLARSSL_API_V1_0
+#endif
+#endif
+#endif
 #define AES_BLOCK_SIZE 16
 #else
 #include <openssl/rand.h>
diff -ur src.orig/ssl.c src/ssl.c
--- src.orig/ssl.c	2011-10-11 12:44:58.000000000 +0200
+++ src/ssl.c	2012-02-13 21:24:26.000000000 +0100
@@ -187,7 +187,11 @@
 	ssl_set_endpoint(ssl, SSL_IS_SERVER);	
 	ssl_set_authmode(ssl, SSL_VERIFY_NONE);
 
+#if defined POLARSSL_API_V1 && !defined(POLARSSL_API_V1_0)
+	ssl_set_rng(ssl, havege_random, &hs);
+#else
 	ssl_set_rng(ssl, havege_rand, &hs);
+#endif
 	ssl_set_dbg(ssl, pssl_debug, NULL);
 	ssl_set_bio(ssl, net_recv, fd, net_send, fd);
 
diff -ur src.orig/ssl.h src/ssl.h
--- src.orig/ssl.h	2011-10-11 12:44:58.000000000 +0200
+++ src/ssl.h	2012-02-13 21:24:26.000000000 +0100
@@ -46,7 +46,10 @@
 #if (POLARSSL_VERSION_MAJOR == 0)
 	#define POLARSSL_API_V0
 #else
-	#define POLARSSL_API_V1
+#define POLARSSL_API_V1
+#if (POLARSSL_VERSION_MINOR == 0)
+	#define POLARSSL_API_V1_0
+#endif
 #endif
 #endif
 
