--- duply.orig	2016-03-24 12:23:12.306773418 +0100
+++ duply	2016-03-24 12:24:09.090775423 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/env bash
+#!/usr/local/duplicity/bin/bash
 #
 ################################################################################
 #  duply (grown out of ftplicity), is a shell front end to duplicity that      #
@@ -453,7 +453,7 @@
 DEFAULT_GPG='gpg'
 DEFAULT_GPG_KEY='_KEY_ID_'
 DEFAULT_GPG_PW='_GPG_PASSWORD_'
-DEFAULT_PYTHON='python'
+DEFAULT_PYTHON='/usr/local/duplicity/env/bin/python'
 
 # function definitions ##########################
 
