--- src/input/mpegts/mpegts_mux.c.org	2014-04-08 12:15:53.000000000 +0200
+++ src/input/mpegts/mpegts_mux.c	2014-04-08 12:13:51.000000000 +0200
@@ -557,10 +557,12 @@
     TAILQ_REMOVE(&mn->mn_initial_scan_current_queue, mm, mm_initial_scan_link);
     mm->mm_initial_scan_status = MM_SCAN_PENDING;
     gtimer_disarm(&mm->mm_initial_scan_timeout);
-    if (mm->mm_initial_scan_done)
+    if (mm->mm_initial_scan_done) {
       TAILQ_INSERT_TAIL(&mn->mn_initial_scan_pending_queue, mm, mm_initial_scan_link);
-    else
+    }
+    else {
       TAILQ_INSERT_HEAD(&mn->mn_initial_scan_pending_queue, mm, mm_initial_scan_link);
+    }
     mpegts_network_schedule_initial_scan(mn);
   }
 
@@ -610,12 +612,14 @@
   assert(mm->mm_initial_scan_status == MM_SCAN_DONE);
 
   mm->mm_initial_scan_status = MM_SCAN_PENDING;
-  if (mm->mm_initial_scan_done)
+  if (mm->mm_initial_scan_done) {
     TAILQ_INSERT_TAIL(&mn->mn_initial_scan_pending_queue, mm,
                       mm_initial_scan_link);
-  else
+  }
+  else {
     TAILQ_INSERT_HEAD(&mn->mn_initial_scan_pending_queue, mm,
                       mm_initial_scan_link);
+  }
   mn->mn_initial_scan_num++;
   
   mn->mn_display_name(mn, buf1, sizeof(buf1));
