--- Makefile.orig	2014-08-13 14:01:08.000000000 +0200
+++ Makefile	2014-08-13 14:01:25.000000000 +0200
@@ -27,8 +27,8 @@
 # Common compiler flags
 #
 
-CFLAGS  += -g -O2 -Wunused-result
-CFLAGS  += -Wall -Werror -Wwrite-strings -Wno-deprecated-declarations
+CFLAGS  += -g -O2
+CFLAGS  += -Wall -Wwrite-strings -Wno-deprecated-declarations
 CFLAGS  += -Wmissing-prototypes
 CFLAGS  += -fms-extensions -funsigned-char -fno-strict-aliasing
 CFLAGS  += -D_FILE_OFFSET_BITS=64
