--- meson.build-orig	2026-03-09 01:52:34.000000000 +0000
+++ meson.build	2026-03-25 00:37:04.094074254 +0000
@@ -21,8 +21,8 @@ cy = meson.get_compiler('cython')
 
 # Check compiler is recent enough (see the SciPy Toolchain Roadmap for details)
 if cc.get_id() == 'gcc'
-  if not cc.version().version_compare('>=9.3')
-    error('NumPy requires GCC >= 9.3')
+  if not cc.version().version_compare('>=8.5')
+    error('NumPy requires GCC >= 8.5')
   endif
 elif cc.get_id() == 'msvc'
   if not cc.version().version_compare('>=19.20')
