diff -uprN ../linuxtv-gitea2766f182b3a4e03543be2ded0845fca4d4fa80-ORIG/v4l/config-mycompat.h ./v4l/config-mycompat.h
--- ../linuxtv-gitea2766f182b3a4e03543be2ded0845fca4d4fa80-ORIG/v4l/config-mycompat.h	1970-01-01 00:00:00.000000000 +0000
+++ ./v4l/config-mycompat.h	2020-07-27 19:04:26.148264906 +0000
@@ -0,0 +1,4 @@
+
+#define GPIOD_FLAGS_BIT_OPEN_DRAIN      BIT(3)
+
+#define GPIOD_OUT_HIGH_OPEN_DRAIN (GPIOD_OUT_HIGH | GPIOD_FLAGS_BIT_OPEN_DRAIN)
