--- meson.build-orig	2026-08-01 00:00:00.000000000 +0000
+++ meson.build	2026-08-01 00:00:00.000000000 +0000
@@ -49,8 +49,8 @@ cython = find_program(cy.cmd_array()[0])
 
 # Check compiler is recent enough (see "Toolchain Roadmap" for details)
 if cc.get_id() == 'gcc'
-  if not cc.version().version_compare('>=10.3')
-    error('SciPy requires GCC >= 10.3')
+  if not cc.version().version_compare('>=8.5')
+    error('SciPy requires GCC >= 8.5')
   endif
 elif cc.get_id() == 'clang' or cc.get_id() == 'clang-cl'
