rename ncurses library from ncursesw6 to ncursesw
to match library created by cross/ncursesw (version 6.1)

--- ./configure.org	2018-04-10 13:43:15.000000000 +0000
+++ ./configure	2020-01-21 19:29:26.825017141 +0000
@@ -5618,13 +5618,13 @@
    if test ! "x$htop_config_script_libs" = x; then
       LDFLAGS="$htop_config_script_libs $LDFLAGS"
       CFLAGS="$htop_config_script_cflags $CFLAGS"
-      { $as_echo "$as_me:${as_lineno-$LINENO}: checking for addnwstr in -lncursesw6" >&5
-$as_echo_n "checking for addnwstr in -lncursesw6... " >&6; }
+      { $as_echo "$as_me:${as_lineno-$LINENO}: checking for addnwstr in -lncursesw" >&5
+$as_echo_n "checking for addnwstr in -lncursesw... " >&6; }
 if ${ac_cv_lib_ncursesw6_addnwstr+:} false; then :
   $as_echo_n "(cached) " >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lncursesw6  $LIBS"
+LIBS="-lncursesw  $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -5876,13 +5876,13 @@
    fi
    if test "x$htop_script_success" = xno; then
 
-   { $as_echo "$as_me:${as_lineno-$LINENO}: checking for addnwstr in -lncursesw6" >&5
-$as_echo_n "checking for addnwstr in -lncursesw6... " >&6; }
+   { $as_echo "$as_me:${as_lineno-$LINENO}: checking for addnwstr in -lncursesw" >&5
+$as_echo_n "checking for addnwstr in -lncursesw... " >&6; }
 if ${ac_cv_lib_ncursesw6_addnwstr+:} false; then :
   $as_echo_n "(cached) " >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lncursesw6  $LIBS"
+LIBS="-lncursesw  $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -5917,7 +5917,7 @@
 
 $as_echo "#define HAVE_LIBNCURSESW 1" >>confdefs.h
 
-      LIBS="-lncursesw6 $LIBS "
+      LIBS="-lncursesw $LIBS "
 
 else
 
