--- json_tokener.c.orig	2014-04-11 04:35:45.000000000 +0200
+++ json_tokener.c	2014-11-19 15:08:24.483987405 +0100
@@ -352,12 +352,10 @@

     case json_tokener_state_inf: /* aka starts with 'i' */
       {
-	int size;
 	int size_inf;
 	int is_negative = 0;

 	printbuf_memappend_fast(tok->pb, &c, 1);
-	size = json_min(tok->st_pos+1, json_null_str_len);
 	size_inf = json_min(tok->st_pos+1, json_inf_str_len);
 	char *infbuf = tok->pb->buf;
 	if (*infbuf == '-')
