# fix netinet/in*.h conflicts for Synology 3.x Kernels
# 
--- src/flowlabel.h.orig	2023-02-16 22:23:12.000000000 +0000
+++ src/flowlabel.h	2023-06-26 06:27:57.549262516 +0000
@@ -31,7 +31,7 @@
 #include <linux/types.h>
 #include <linux/version.h>
 
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,7,0)
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,11,0)
 #include <linux/in6.h>
 #else
 #ifndef __ANDROID__
@@ -39,7 +39,7 @@
    It is just a stripped copy of the Linux kernel header "linux/in6.h"
    "Flow label" things are still not defined in "netinet/in*.h" headers,
    but we cannot use "linux/in6.h" immediately because it currently
-   conflicts with "netinet/in.h" . (in kernel versions < 3.7.0)
+   conflicts with "netinet/in.h" . (in synology kernel versions < 3.11.0)
 */
 struct in6_flowlabel_req
 {
