--- setup.py.orig	2009-11-15 17:06:10.000000000 +0100
+++ setup.py	2011-08-08 23:28:45.000000000 +0200
@@ -5,7 +5,7 @@
 # Usage: python setup.py install
 #

-import glob, os, re, struct, string, sys
+import glob, os, re, string, sys

 # make it possible to run the setup script from another directory
 try:
@@ -34,8 +34,8 @@
 # TIFF_ROOT = libinclude("/opt/tiff")

 TCL_ROOT = None
-JPEG_ROOT = None
-ZLIB_ROOT = None
+JPEG_ROOT = libinclude(os.environ['STAGING_INSTALL_PREFIX'])
+ZLIB_ROOT = libinclude(os.environ['STAGING_INSTALL_PREFIX'])
 TIFF_ROOT = None
 FREETYPE_ROOT = None
 LCMS_ROOT = None
@@ -147,7 +147,6 @@
             add_directory(library_dirs, "/opt/local/lib")
             add_directory(include_dirs, "/opt/local/include")

-        add_directory(library_dirs, "/usr/local/lib")
         # FIXME: check /opt/stuff directories here?

         prefix = sysconfig.get_config_var("prefix")
@@ -208,11 +208,9 @@
                 add_directory(include_dirs, tcl_dir)
 
         # standard locations
-        add_directory(library_dirs, "/usr/local/lib")
-        add_directory(include_dirs, "/usr/local/include")
 
-        add_directory(library_dirs, "/usr/lib")
-        add_directory(include_dirs, "/usr/include")
+        add_directory(library_dirs, os.environ['STAGING_INSTALL_PREFIX'] + "/lib")
+        add_directory(include_dirs, os.environ['STAGING_INSTALL_PREFIX'] + "/include")
 
         #
         # insert new dirs *before* default libs, to avoid conflicts
@@ -299,8 +299,6 @@
             defs.append(("HAVE_LIBZ", None))
         if sys.platform == "win32":
             libs.extend(["kernel32", "user32", "gdi32"])
-        if struct.unpack("h", "\0\1")[0] == 1:
-            defs.append(("WORDS_BIGENDIAN", None))

         exts = [(Extension(
             "_imaging", files, libraries=libs, define_macros=defs
