--- setup.py.orig	2014-09-05 10:26:43.120557588 +0200
+++ setup.py	2014-09-05 10:27:10.408559135 +0200
@@ -85,7 +85,6 @@
     """
 
     def finalize_options(self):
-        self.distribution.ext_modules = get_ext_modules()
         build.finalize_options(self)
 
 
@@ -97,7 +96,6 @@
     """
 
     def finalize_options(self):
-        self.distribution.ext_modules = get_ext_modules()
         install.finalize_options(self)
 
 
