# Patch udev-builtin-input_id.c for older kernels
# 
--- src/udev/udev-builtin-input_id.c.orig	2013-05-09 15:45:50.000000000 +0200
+++ src/udev/udev-builtin-input_id.c	2022-12-04 09:38:23.094516200 +0100
@@ -30,6 +30,16 @@
 
 #include "udev.h"
 
+/*
+ * define max key to check the higher block in test_key()
+ * BTN_TRIGGER_HAPPY in <linux/input.h> is not defined for older kernels
+ */
+#ifndef BTN_TRIGGER_HAPPY
+#define __MAX_KEY (KEY_NUMERIC_POUND+1)
+#else
+#define __MAX_KEY BTN_TRIGGER_HAPPY
+#endif
+
 /* we must use this kernel-compatible implementation */
 #define BITS_PER_LONG (sizeof(unsigned long) * 8)
 #define NBITS(x) ((((x)-1)/BITS_PER_LONG)+1)
@@ -163,7 +173,7 @@
         }
         /* If there are no keys in the lower block, check the higher block */
         if (!found) {
-                for (i = KEY_OK; i < BTN_TRIGGER_HAPPY; ++i) {
+                for (i = KEY_OK; i < __MAX_KEY; ++i) {
                         if (test_bit (i, bitmask_key)) {
                                 log_debug("test_key: Found key %x in high block\n", i);
                                 found = 1;
