--- util/Makefile.in.org	2015-09-24 02:23:18.169476100 -1000
+++ util/Makefile.in	2015-09-24 02:28:31.817476100 -1000
@@ -650,7 +650,7 @@
 
 info-am:
 
-install-data-am: install-data-local
+install-data-am:
 
 install-dvi: install-dvi-am
 
@@ -730,10 +730,8 @@
 	fi
 
 install-exec-local:
-	$(MKDIR_P) $(DESTDIR)$(MOUNT_FUSE_PATH)
-	$(INSTALL_PROGRAM) $(builddir)/mount.fuse $(DESTDIR)$(MOUNT_FUSE_PATH)/mount.fuse
-	$(MKDIR_P) $(DESTDIR)$(INIT_D_PATH)
-	$(INSTALL_SCRIPT) $(srcdir)/init_script $(DESTDIR)$(INIT_D_PATH)/fuse
+	$(MKDIR_P) $(sbindir)
+	$(INSTALL_PROGRAM) $(builddir)/mount.fuse $(sbindir)/mount.fuse
 	@if test -x /usr/sbin/update-rc.d; then \
 		echo "/usr/sbin/update-rc.d fuse start 34 S . start 41 0 6 . || true"; \
 		/usr/sbin/update-rc.d fuse start 34 S . start 41 0 6 . || true; \
