diff -rU 4 /home/bob/src/spksrc-allanc/spk/python3/work/acl-2.2.52/chacl/Makefile /home/bob/src/spksrc-allanc/spk/python3/acl-2.2.52-fixed/chacl/Makefile
--- chacl/Makefile	2013-04-30 15:20:43.000000000 +0000
+++ chacl/Makefile	2018-05-24 06:59:51.352327115 +0000
@@ -30,6 +30,10 @@
 include $(BUILDRULES)
 
+# -D is added redundantly because acl-2.2.52 is really old, needs an autoreconf,
+# but that's techdebt for later; now, this -D is harmless but causes the current
+# install binary to actually install the file rather than silently fail
+
 install: default
 	$(INSTALL) -m 755 -d $(PKG_BIN_DIR)
-	$(LTINSTALL) -m 755 $(LTCOMMAND) $(PKG_BIN_DIR)
+	$(LTINSTALL) -o $(shell id -un) -m 755 $(LTCOMMAND) $(PKG_BIN_DIR)
 install-dev install-lib:
diff -rU 4 /home/bob/src/spksrc-allanc/spk/python3/work/acl-2.2.52/debian/Makefile /home/bob/src/spksrc-allanc/spk/python3/acl-2.2.52-fixed/debian/Makefile
--- debian/Makefile	2013-04-30 15:20:43.000000000 +0000
+++ debian/Makefile	2018-05-24 06:59:51.360327714 +0000
@@ -27,25 +27,29 @@
 default:
 
 include $(BUILDRULES)
 
+# -D is added redundantly because acl-2.2.52 is really old, needs an autoreconf,
+# but that's techdebt for later; now, this -D is harmless but causes the current
+# install binary to actually install the file rather than silently fail
+
 install: default
 ifeq ($(PKG_DISTRIBUTION), debian)
 	$(INSTALL) -m 755 -d $(PKG_DOC_DIR)
-	$(INSTALL) -m 644 changelog $(PKG_DOC_DIR)/changelog.Debian
+	$(INSTALL) -D -m 644 changelog $(PKG_DOC_DIR)/changelog.Debian
 endif
 
 install-dev: default
 ifeq ($(PKG_DISTRIBUTION), debian)
 	$(INSTALL) -m 755 -d $(PKG_DOC_DIR)
 	$(INSTALL) -m 755 -d $(DEV_DOC_DIR)
-	$(INSTALL) -m 644 copyright $(DEV_DOC_DIR)
-	$(INSTALL) -m 644 changelog $(DEV_DOC_DIR)/changelog.Debian
+	$(INSTALL) -D -m 644 copyright $(DEV_DOC_DIR)
+	$(INSTALL) -D -m 644 changelog $(DEV_DOC_DIR)/changelog.Debian
 endif
 
 install-lib: default
 ifeq ($(PKG_DISTRIBUTION), debian)
 	$(INSTALL) -m 755 -d $(PKG_DOC_DIR)
 	$(INSTALL) -m 755 -d $(LIB_DOC_DIR)
-	$(INSTALL) -m 644 copyright $(LIB_DOC_DIR)
-	$(INSTALL) -m 644 changelog $(LIB_DOC_DIR)/changelog.Debian
+	$(INSTALL) -D -m 644 copyright $(LIB_DOC_DIR)
+	$(INSTALL) -D -m 644 changelog $(LIB_DOC_DIR)/changelog.Debian
 endif
diff -rU 4 /home/bob/src/spksrc-allanc/spk/python3/work/acl-2.2.52/getfacl/Makefile /home/bob/src/spksrc-allanc/spk/python3/acl-2.2.52-fixed/getfacl/Makefile
--- getfacl/Makefile	2013-04-30 15:20:43.000000000 +0000
+++ getfacl/Makefile	2018-05-24 06:59:51.356327527 +0000
@@ -29,8 +29,12 @@
 default: $(LTCOMMAND)
 
 include $(BUILDRULES)
 
+# -D is added redundantly because acl-2.2.52 is really old, needs an autoreconf,
+# but that's techdebt for later; now, this -D is harmless but causes the current
+# install binary to actually install the file rather than silently fail
+
 install: default
 	$(INSTALL) -m 755 -d $(PKG_BIN_DIR)
-	$(LTINSTALL) -m 755 $(LTCOMMAND) $(PKG_BIN_DIR)
+	$(LTINSTALL) -D -m 755 $(LTCOMMAND) $(PKG_BIN_DIR)
 install-dev install-lib:
diff -rU 4 /home/bob/src/spksrc-allanc/spk/python3/work/acl-2.2.52/setfacl/Makefile /home/bob/src/spksrc-allanc/spk/python3/acl-2.2.52-fixed/setfacl/Makefile
--- setfacl/Makefile	2013-04-30 15:20:43.000000000 +0000
+++ setfacl/Makefile	2018-05-24 06:59:51.356327527 +0000
@@ -29,8 +29,12 @@
 default: $(LTCOMMAND)
 
 include $(BUILDRULES)
 
+# -D is added redundantly because acl-2.2.52 is really old, needs an autoreconf,
+# but that's techdebt for later; now, this -D is harmless but causes the current
+# install binary to actually install the file rather than silently fail
+
 install: default
 	$(INSTALL) -m 755 -d $(PKG_BIN_DIR)
-	$(LTINSTALL) -m 755 $(LTCOMMAND) $(PKG_BIN_DIR)
+	$(LTINSTALL) -D -m 755 $(LTCOMMAND) $(PKG_BIN_DIR)
 install-dev install-lib:
