# Update shebang for installed python
--- bin/duplicity
+++ bin/duplicity
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/local/duplicity/env/bin/python
 # -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
 #
 # duplicity -- Encrypted bandwidth efficient backup
--- bin/rdiffdir
+++ bin/rdiffdir
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/local/duplicity/env/bin/python
 # rdiffdir -- Extend rdiff functionality to directories
 # Version 0.8.08 released December 08, 2019
 #
