--- setup.py.org	2014-05-09 15:27:41.000000000 +0200
+++ setup.py	2014-10-02 07:28:43.000000000 +0200
@@ -25,6 +25,7 @@
 from setuptools.command.test import test
 from setuptools.command.install import install
 from setuptools.command.sdist import sdist
+from setuptools.command.install_scripts import install_scripts
 
 version_string = "0.6.24"
 
@@ -124,6 +136,26 @@
         os.system("mv duplicity-" + version + ".tar.gz " + self.dist_dir)
 
 
+class InstallScriptsCommand(install_scripts):
+    def run(self):
+        install_scripts.run(self)
+
+        for outfile in self.outfiles:
+            fp = open(outfile, 'rb')
+            data = fp.read()
+            fp.close()
+
+            # skip binary files
+            if '\0' in data:
+                continue
+
+            shebang = '#!python'
+            data = data.replace('#!/usr/local/duplicity/bin/python2.7', shebang.encode('string_escape'))
+            fp = open(outfile, 'wb')
+            fp.write(data)
+            fp.close()
+
+
 setup(name="duplicity",
       version=version_string,
       description="Encrypted backup using rsync algorithm",
@@ -151,5 +183,6 @@
       test_suite = 'testing',
       cmdclass={'test': TestCommand,
                 'install': InstallCommand,
-                'sdist': SDistCommand},
+                'sdist': SDistCommand,
+                'install_scripts': InstallScriptsCommand},
       )
