--- setup.py.orig	2016-03-24 11:29:14.846659092 +0100
+++ setup.py	2016-03-24 11:29:01.422658618 +0100
@@ -26,6 +26,7 @@
 from setuptools.command.test import test
 from setuptools.command.install import install
 from setuptools.command.sdist import sdist
+from setuptools.command.install_scripts import install_scripts
 from distutils.command.build_scripts import build_scripts
 
 version_string = "0.7.06"
@@ -126,6 +127,24 @@
         os.system("mkdir -p " + self.dist_dir)
         os.system("mv duplicity-" + version + ".tar.gz " + self.dist_dir)
 
+class InstallScriptsCommand(install_scripts):
+    def run(self):
+        install_scripts.run(self)
+
+        for outfile in self.outfiles:
+            fp = open(outfile, 'rb')
+            data = fp.read()
+            fp.close()
+
+            # skip binary files
+            if '\0' in data:
+                continue
+
+            shebang = '#!python'
+            data = data.replace('#!/usr/local/duplicity/bin/python2.7', shebang.encode('string_escape'))
+            fp = open(outfile, 'wb')
+            fp.write(data)
+            fp.close()
 
 # don't touch my shebang
 class BSCommand (build_scripts):
@@ -198,5 +217,6 @@
       cmdclass={'test': TestCommand,
                 'install': InstallCommand,
                 'sdist': SDistCommand,
-                'build_scripts': BSCommand},
+                'build_scripts': BSCommand,
+                'install_scripts': InstallScriptsCommand},                
       )
