--- mcs/tools/xbuild/targets/Microsoft.Portable.Common.targets	1969-12-31 17:00:00.000000000 -0700
+++ mcs/tools/xbuild/targets/Microsoft.Portable.Common.targets.new	2014-05-01 14:00:02.668857231 -0600
@@ -0,0 +1,4 @@
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+	<Import Project="..\Microsoft.Portable.Core.props" />
+	<Import Project="..\Microsoft.Portable.Core.targets" />
+</Project>
