# Disable the builtin check for newer versions.
--- sabnzbd/misc.py
+++ sabnzbd/misc.py
@@ -390,6 +390,8 @@
     they are already using an alpha/beta/rc.
     RC's are valued higher than Beta's, which are valued higher than Alpha's.
     """
+    logging.info("Version check disabled by SynoCommunity packaging")
+    return

     if not cfg.version_check():
         return
--- interfaces/Config/templates/config_general.tmpl
+++ interfaces/Config/templates/config_general.tmpl
@@ -169,15 +169,6 @@
                     <input type="checkbox" name="auto_browser" id="auto_browser" value="1" <!--#if int($auto_browser) > 0 then 'checked="checked"' else ""#--> />
                     <span class="desc">$T('explain-auto_browser')</span>
                 </div>
-                <div class="field-pair">
-                    <label class="config" for="check_new_rel">$T('opt-check_new_rel')</label>
-                    <select name="check_new_rel" id="check_new_rel">
-                        <option value="0" <!--#if $check_new_rel == 0 then 'selected="selected"' else ""#--> >$T('off')</option>
-                        <option value="1" <!--#if $check_new_rel == 1 then 'selected="selected"' else ""#--> >$T('on')</option>
-                        <option value="2" <!--#if $check_new_rel == 2 then 'selected="selected"' else ""#--> >$T('also-test')</option>
-                    </select>
-                    <span class="desc">$T('explain-check_new_rel')</span>
-                </div>
                 <div class="field-pair advanced-settings <!--#if int($certificate_validation) == 0 then "disabled" else ""#-->">
                     <label class="config" for="enable_https_verification">$T('opt-enable_https_verification')</label>
                     <input type="checkbox" name="enable_https_verification" id="enable_https_verification" value="1" <!--#if int($enable_https_verification) > 0 then 'checked="checked"' else ""#--> <!--#if int($certificate_validation) == 0 then "disabled=\"disabled\"" else ""#--> />
--- sabnzbd/scheduler.py
+++ sabnzbd/scheduler.py
@@ -192,15 +192,6 @@
             self.scheduler.add_interval_task(sabnzbd.RSSReader.run, "RSS", delay * 60, interval * 60)
             self.scheduler.add_single_task(sabnzbd.RSSReader.run, "RSS", 15)

-        if cfg.version_check():
-            # Check for new release, once per week on random time
-            m = random.randint(0, 59)
-            h = random.randint(0, 23)
-            d = (random.randint(1, 7),)
-
-            logging.info("Scheduling VersionCheck on day %s at %s:%s", d[0], h, m)
-            self.scheduler.add_daytime_task(sabnzbd.misc.check_latest_version, "VerCheck", d, None, (h, m))
-
         action, hour, minute = sabnzbd.BPSMeter.get_quota()
         if action:
             logging.info("Setting schedule for quota check daily at %s:%s", hour, minute)
