--- sabnzbd/misc.py
+++ sabnzbd/misc.py
@@ -820,9 +820,11 @@
         else:
             # Check for other forms of virtualization
             try:
-                if virt := run_command(["systemd-detect-virt"]).strip():
-                    if virt != "none":
-                        platform_tags.append(virt)
+                # Check if the command exists by searching in the PATH
+                if run_command(["which", "systemd-detect-virt"]).strip():
+                    if virt := run_command(["systemd-detect-virt"]).strip():
+                        if virt != "none":
+                            platform_tags.append(virt)
             except:
                 pass
 
