--- sabnzbd/newsunpack.py.orig	2013-09-02 22:06:57.000000000 +0200
+++ sabnzbd/newsunpack.py	2013-09-08 16:32:24.664751012 +0200
@@ -966,7 +966,7 @@
 
     # Append the wildcard for this set
     wildcard = '%s*' % os.path.join(os.path.split(parfile)[0], setname)
-    if single or len(globber(wildcard, None)) < 2:
+    if len(globber(wildcard, None)) < 2:
         # Support bizarre naming conventions
         wildcard = os.path.join(os.path.split(parfile)[0], '*')
     if sabnzbd.WIN32 or sabnzbd.DARWIN:
